{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqUtils import GC\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import statistics\n",
    "import operator\n",
    "from tqdm.notebook import tqdm\n",
    "from functions import *\n",
    "import random\n",
    "from Bio import Restriction\n",
    "import copy\n",
    "from itertools import combinations\n",
    "from itertools import permutations\n",
    "from itertools import product\n",
    "import urllib.request as urllib2\n",
    "from dna_features_viewer import BiopythonTranslator\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create a codon table for e.coli:\n",
    "all_name_cds_ecoli = {str(record.description).split(\"gene=\")[1].split(\"]\",1)[0]:str(record.seq)\n",
    "                      for record in SeqIO.parse(\"../ref_files_m13/GCF_000005845.2_ASM584v2_cds_from_genomic.fna\", \"fasta\")}\n",
    "codon_table_ecoli = create_codon_table(list(all_name_cds_ecoli.values()), 11)\n",
    "\n",
    "# Make a reversed dict codon usage of whole genome\n",
    "rev_ecoli_codon_usage = reverse_codon_table(codon_table_ecoli)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{(0, 378): 'AATGCTACTACTATTAGTAGAATTGATGCCACCTTTTCAGCTCGCGCCCCAAATGAAAATATAGCTAAACAGGTTATTGACCATTTGCGAAATGTATCTAATGGTCAAACTAAATCTACTCGTTCGCAGAATTGGGAATCAACTGTTATATGGAATGAAACTTCCAGACACCGTACTTTAGTTGCATATTTAAAACATGTTGAGCTACAGCATTATATTCAGCAATTAAGCTCTAAGCCATCCGCAAAAATGACCTCTTATCAAAAGGAGCAATTAAAGGTACTCTCTAATCCTGACCTGTTGGAGTTTGCTTCCGGTCTGGTTCGCTTTGAAGCTCGAATTAAAACGCGATATTTGAAGTCTTTCGGGCTTCCTCTT'}, {(426, 474): 'AAAGACCTGATTTTTGATTTATGGTCATTCTCGTTTTCTGAACTGTTT'}, {(495, 768): 'ATGAATATTTATGACGATTCCGCAGTATTGGACGCTATCCAGTCTAAACATTTTACTATTACCCCCTCTGGCAAAACTTCTTTTGCAAAAGCCTCTCGCTATTTTGGTTTTTATCGTCGTCTGGTAAACGAGGGTTATGATAGTGTTGCTCTTACTATGCCTCGTAATTCCTTTTGGCGTTATGTATCTGCATTAGTTGAATGTGGTATTCCTAAATCTCAACTGATGAATCTTTCTACCTGTAATAATGTTGTTCCGTTAGTTCGTTTTATT'}, {(816, 822): 'CTTAAA'}, {(842, 1085): 'ATGATTAAAGTTGAAATTAAACCATCTCAAGCCCAATTTACTACTCGTTCTGGTGTTTCTCGTCAGGGCAAGCCTTATTCACTGAATGAGCAGCTTTGTTACGTTGATTTGGGTAATGAATATCCGGTTCTTGTCAAGATTACTCTTGATGAAGGTCAGCCAGCCTATGCGCCTGGTCTGTACACCGTTCATCTGTCCTCTTTCAAAGTTGGTCAGTTCGGTTCCCTTATGATTGACCGTCTG'}, {(1107, 1152): 'ATGGAGCAGGTCGCGGATTTCGACACAATTTATCAGGCGATGATA'}, {(1211, 1280): 'GTTTTAGTGTATTCTTTTGCCTCTTTCGTTTTAGGTTGGTGCCTTCGTAGTGGCATTACGTATTTTACC'}, {(1306, 1498): 'AAGTCTTTAGTCCTCAAAGCCTCTGTAGCCGTTGCTACCCTCGTTCCGATGCTGTCTTTCGCTGCTGAGGGTGACGATCCCGCAAAAGCGGCCTTTAACTCCCTGCAAGCCTCAGCGACCGAATATATCGGTTATGCGTGGGCGATGGTTGTTGTCATTGTCGGCGCAACTATCGGTATCAAGCTGTTTAAG'}, {(1577, 2714): 'GTGAAAAAATTATTATTCGCAATTCCTTTAGTTGTTCCTTTCTATTCTCACTCCGCTGAAACTGTTGAAAGTTGTTTAGCAAAATCCCATACAGAAAATTCATTTACTAACGTCTGGAAAGACGACAAAACTTTAGATCGTTACGCTAACTATGAGGGCTGTCTGTGGAATGCTACAGGCGTTGTAGTTTGTACTGGTGACGAAACTCAGTGTTACGGTACATGGGTTCCTATTGGGCTTGCTATCCCTGAAAATGAGGGTGGTGGCTCTGAGGGTGGCGGTTCTGAGGGTGGCGGTTCTGAGGGTGGCGGTACTAAACCTCCTGAGTACGGTGATACACCTATTCCGGGCTATACTTATATCAACCCTCTCGACGGCACTTATCCGCCTGGTACTGAGCAAAACCCCGCTAATCCTAATCCTTCTCTTGAGGAGTCTCAGCCTCTTAATACTTTCATGTTTCAGAATAATAGGTTCCGAAATAGGCAGGGGGCATTAACTGTTTATACGGGCACTGTTACTCAAGGCACTGACCCCGTTAAAACTTATTACCAGTACACTCCTGTATCATCAAAAGCCATGTATGACGCTTACTGGAACGGTAAATTCAGAGACTGCGCTTTCCATTCTGGCTTTAATGAGGATTTATTTGTTTGTGAATATCAAGGCCAATCGTCTGACCTGCCTCAACCTCCTGTCAATGCTGGCGGCGGCTCTGGTGGTGGTTCTGGTGGCGGCTCTGAGGGTGGTGGCTCTGAGGGTGGCGGTTCTGAGGGTGGCGGCTCTGAGGGAGGCGGTTCCGGTGGTGGCTCTGGTTCCGGTGATTTTGATTATGAAAAGATGGCAAACGCTAATAAGGGGGCTATGACCGAAAATGCCGATGAAAACGCGCTACAGTCTGACGCTAAAGGCAAACTTGATTCTGTCGCTACTGATTACGGTGCTGCTATCGATGGTTTCATTGGTGACGTTTCCGGCCTTGCTAATGGTAATGGTGCTACTGGTGATTTTGCTGGCTCTAATTCCCAAATGGCTCAAGTCGGTGACGGTGATAATTCACCTTTAATGAATAATTTCCGTCAATATTTACCTTCCCTCCCTCAATCGGTTGAATGTCGCCCTTTTGTCTTTGGCGCT'}, {(2765, 2834): 'GGTGTCTTTGCGTTTCTTTTATATGTTGCCACCTTTATGTATGTATTTTCTACGTTTGCTAACATACTG'}, {(2854, 3082): 'ATGCCAGTTCTTTTGGGTATTCCGTTATTATTGCGTTTCCTCGGTTTCCTTCTGGTAACTTTGTTCGGCTATCTGCTTACTTTTCTTAAAAAGGGCTTCGGTAAGATAGCTATTGCTATTTCATTGTTTCTTGCTCTTATTATTGGGCTTAACTCAATTCTTGTGGGTTATCTCTCTGATATTAGCGCTCAATTACCCTCTGACTTTGTTCAGGGTGTTCAGTTAATT'}, {(3133, 3172): 'GCTATTTTCATTTTTGACGTTAAACAAAAAATCGTTTCT'}, {(3194, 3218): 'ATGGCTGTTTATTTTGTAACTGGC'}, {(3275, 3317): 'GCTGGGTGCAAAATAGCAACTAATCTTGATTTAAGGCTTCAA'}, {(3341, 4058): 'GCTAAAACGCCTCGCGTTCTTAGAATACCGGATAAGCCTTCTATATCTGATTTGCTTGCTATTGGGCGCGGTAATGATTCCTACGATGAAAATAAAAACGGCTTGCTTGTTCTCGATGAGTGCGGTACTTGGTTTAATACCCGTTCTTGGAATGATAAGGAAAGACAGCCGATTATTGATTGGTTTCTACATGCTCGTAAATTAGGATGGGATATTATTTTTCTTGTTCAGGACTTATCTATTGTTGATAAACAGGCGCGTTCTGCATTAGCTGAACATGTTGTTTATTGTCGTCGTCTGGACAGAATTACTTTACCTTTTGTCGGTACTTTATATTCTCTTATTACTGGCTCGAAAATGCCTCTGCCTAAATTACATGTTGGCGTTGTTAAATATGGCGATTCTCAATTAAGCCCTACTGTTGAGCGTTGGCTTTATACTGGTAAGAATTTGTATAACGCATATGATACTAAACAGGCTTTTTCTAGTAATTATGATTCCGGTGTTTATTCTTATTTAACGCCTTATTTATCACACGGTCGGTATTTCAAACCATTAAATTTAGGTCAGAAGATGAAATTAACTAAAATATATTTGAAAAAGTTTTCTCGCGTTCTTTGTCTTGCGATTGGATTTGCATCAGCATTTACATATAGTTATATAACCCAACCTAAGCCGGAGGTTAAAAAGGTAGTCTCTCAGACCTATGATTTTGAT'}, {(4109, 4196): 'TTCAAGGATTCTAAGGGAAAATTAATTAATAGCGACGATTTACAGAAGCAAGGTTATTCACTCACATATATTGATTTATGTACTGTT'}, {(4242, 4629): 'TTTGTTTTCTTGATGTTTGTTTCATCATCTTCTTTTGCTCAGGTAATTGAAATGAATAATTCGCCTCTGCGCGATTTTGTAACTTGGTATTCAAAGCAATCAGGCGAATCCGTTATTGTTTCTCCCGATGTAAAAGGTACTGTTACTGTATATTCATCTGACGTTAAACCTGAAAATCTACGCAATTTCTTTATTTCTGTTTTACGTGCAAATAATTTTGATATGGTAGGTTCTAACCCTTCCATTATTCAGAAGTATAATCCAAACAATCAGGATTATATTGATGAATTGCCATCATCTGATAATCAGGAATATGATGATAATTCCGCTCCTTCTGGTGGTTTCTTTGTTCCGCAAAATGATAATGTTACTCAAACTTTTAAAATT'}, {(4683, 5484): 'AATACTTCTAAATCCTCAAATGTATTATCTATTGACGGCTCTAATCTATTAGTTGTTAGTGCTCCTAAAGATATTTTAGATAACCTTCCTCAATTCCTTTCAACTGTTGATTTGCCAACTGACCAGATATTGATTGAGGGTTTGATATTTGAGGTTCAGCAAGGTGATGCTTTAGATTTTTCATTTGCTGCTGGCTCTCAGCGTGGCACTGTTGCAGGCGGTGTTAATACTGACCGCCTCACCTCTGTTTTATCTTCTGCTGGTGGTTCGTTCGGTATTTTTAATGGCGATGTTTTAGGGCTATCAGTTCGCGCATTAAAGACTAATAGCCATTCAAAAATATTGTCTGTGCCACGTATTCTTACGCTTTCAGGTCAGAAGGGTTCTATCTCTGTTGGCCAGAATGTCCCTTTTATTACTGGTCGTGTGACTGGTGAATCTGCCAATGTAAATAATCCATTTCAGACGATTGAGCGTCAAAATGTAGGTATTTCCATGAGCGTTTTTCCTGTTGCAATGGCTGGCGGTAATATTGTTCTGGATATTACCAGCAAGGCCGATAGTTTGAGTTCTTCTACTCAGGCAAGTGATGTTATTACTAATCAAAGAAGTATTGCTACAACGGTTAATTTGCGTGATGGACAGACTCTTTTACTCGGTGGCCTCACTGATTATAAAAACACTTCTCAGGATTCTGGCGTACCGTTCCTGTCTAAAATCCCTTTAATCGGCCTCCTGTTTAGCTCCCGCTCTGATTCTAACGAGGAAAGCACGTTATACGTGCTCGTCAAAGCAACCATA'}, {(5867, 6683): 'TTAGAAAAACTCATCGAGCATCAAATGAAACTGCAATTTATTCATATCAGGATTATCAATACCATATTTTTGAAAAAGCCGTTTCTGTAATGAAGGAGAAAACTCACCGAGGCAGTTCCATAGGATGGCAAGATCCTGGTATCGGTCTGCGATTCCGACTCGTCCAACATCAATACAACCTATTAATTTCCCCTCGTCAAAAATAAGGTTATCAAGTGAGAAATCACCATGAGTGACGACTGAATCCGGTGAGAATGGCAAAAGCTTATGCATTTCTTTCCAGACTTGTTCAACAGGCCAGCCATTACGCTCGTCATCAAAATCACTCGCATCAACCAAACCGTTATTCATTCGTGATTGCGCCTGAGCGAGACGAAATACGCGATCGCTGTTAAAAGGACAATTACAAACAGGAATCGAATGCAACCGGCGCAGGAACACTGCCAGCGCATCAACAATATTTTCACCTGAATCAGGATATTCTTCTAATACCTGGAATGCTGTTTTCCCGGGGATCGCAGTGGTGAGTAACCATGCATCATCAGGAGTACGGATAAAATGCTTGATGGTCGGAAGAGGCATAAATTCCGTCAGCCAGTTTAGTCTGACCATCTCATCTGTAACATCATTGGCAACGCTACCTTTGCCATGTTTCAGAAACAACTCTGGCGCATCGGGCTTCCCATACAATCGATAGATTGTCGCACCTGATTGCCCGACATTATCGCGAGCCCATTTATACCCATATAAATCAGCATCCATGTTGGAATTTAATCGCGGCCTCGAGCAAGACGTTTCCCGTTGAATATGGCTCAT'}, {(6880, 7072): 'ATGATTGACATGCTAGTTTTACGATTACCGTTCATCGATTCTCTTGTTTGCTCCAGACTCTCAGGCAATGACCTGATAGCCTTTGTAGATCTCTCAAAAATAGCTACCCTCTCCGGCATTAATTTATCAGCTAGAACGGTTGAATATCATATTGATGGTGATTTGACTGTCTCCGGCCTTTCTCACCCTTTT'}, {(7126, 7282): 'AAAAATTTTTATCCTTGCGTTGAAATAAAGGCTTCTCCCGCAAAAGTATTACAGGGTCATAATGTTTTTGGTACAACCGATTTAGCTTTATGCTCTGAGGCTTTATTGCTTAATTTTGCTAATTCTTTGCCTTGCCTGTATGATTTATTGGATGTT'}]\n",
      "(0, 378)\n",
      "(426, 474)\n",
      "(495, 768)\n",
      "(816, 822)\n",
      "(842, 1085)\n",
      "(1107, 1152)\n",
      "(1211, 1280)\n",
      "(1306, 1498)\n",
      "(1577, 2714)\n",
      "(2765, 2834)\n",
      "(2854, 3082)\n",
      "(3133, 3172)\n",
      "(3194, 3218)\n",
      "(3275, 3317)\n",
      "(3341, 4058)\n",
      "(4109, 4196)\n",
      "(4242, 4629)\n",
      "(4683, 5484)\n",
      "(5867, 6683)\n",
      "(6880, 7072)\n",
      "(7126, 7282)\n"
     ]
    }
   ],
   "source": [
    "def recode_perc_change(perc,cds,codon_table):\n",
    "    #recode the cds by changing the % provided as input based on the codon table provided as input.\n",
    "    #output: recoded cds (string)\n",
    "    new_cds = \"\"\n",
    "    for pos in range(int(len(cds)/3)):    #25% de chance de le changer\n",
    "        codon_in_cds = cds[pos*3:pos*3+3].upper()\n",
    "        #set seed at each position in the CDS\n",
    "        random.seed(pos)\n",
    "        if pick_this_pos(perc) and codon_in_cds !=\"ATG\" and codon_in_cds!=\"TGG\" and pos*3+3 != len(cds):\n",
    "            #if no TGG (cannot be changed), no start codon (cannot be changed) or no stop codon:\n",
    "            other_codons = {}\n",
    "            for aa, codon_freq in codon_table.items():\n",
    "                if codon_in_cds in codon_freq:\n",
    "                    for codon,freq in codon_freq.items():\n",
    "                        if codon_in_cds != codon:\n",
    "                            other_codons[codon] = freq\n",
    "            new_codon_dict = {codon:freq/sum(list(codon_freq.values())) for codon,freq in other_codons.items()}\n",
    "            replacement_codon = random.choices(population=list(new_codon_dict.keys()),weights=list(new_codon_dict.values()),k=1)\n",
    "            new_cds += ''.join(replacement_codon).upper()\n",
    "        else:\n",
    "            new_cds += codon_in_cds\n",
    "    #new_cds is recoded\n",
    "    return new_cds\n",
    "\n",
    "\n",
    "def merge_contiguous_regions(list_of_tuples):\n",
    "    #the regions do not ovif regions == forbiden_regions[-1]: \n",
    "    #last forbidden region, exceptionnal case:erlap, but have common borders. This function merge common borders into one forbidden region.\n",
    "    #input: list of tuples\n",
    "    #output: list of tuples merged\n",
    "    merge_list = []\n",
    "    max_range = [-1]\n",
    "    max_list_of_tuples = max([max(i) for i in list_of_tuples])\n",
    "    min_range = [max_list_of_tuples+1]\n",
    "    for nb,ref in enumerate(list_of_tuples):\n",
    "        if nb+1 < len(list_of_tuples):    #to make sure we do not compared with something outside the list!\n",
    "            comp = list_of_tuples[nb+1]\n",
    "            if comp[0] <= ref[1] and comp[1] >= ref[0]: #if there is any overlap:\n",
    "                min_temp = min([ref[0],comp[0]])\n",
    "                max_temp = max([ref[1],comp[1]])\n",
    "                min_range.append(min_temp)\n",
    "                max_range.append(max_temp)\n",
    "                max_range = [max(max_range)]\n",
    "                min_range = [min(min_range)]\n",
    "                if nb+1 == len(list_of_tuples)-1:\n",
    "                    merge_list.append((min_range[0],max_range[0]))\n",
    "            else:\n",
    "                if nb > 0:\n",
    "                    if max_range == [-1] and min_range == [max_list_of_tuples+1]:\n",
    "                        merge_list.append((ref[0],ref[1]))\n",
    "                    else:\n",
    "                        merge_list.append((min_range[0],max_range[0]))\n",
    "                    if nb+1 == len(list_of_tuples)-1:  #if last\n",
    "                         merge_list.append((comp[0],comp[1]))\n",
    "                    max_range = [-1]\n",
    "                    min_range = [max_list_of_tuples+1]\n",
    "                else:\n",
    "                    merge_list.append((ref[0],ref[1]))\n",
    "    return merge_list\n",
    "def pick_this_pos(percent):\n",
    "    #return true if the random number (between 0-100) is lower than the given percent.\n",
    "    return random.randrange(100) < percent\n",
    "\n",
    "def same_ORF(interval, ref_pos_gene):\n",
    "    #this function tells if coordinates of a part of a cds to be recoded overlaps with an actual gene (other than himself)\n",
    "    #if the pos[0]-interval[0] < 3, on l'exclu de toute maniere\n",
    "    return bool((ref_pos_gene-interval)%3 == 0)\n",
    "\n",
    "# def overlap_situation(cds_parts,CDSs):\n",
    "#     no_overlap_to_recode = {}\n",
    "#     for gene_name, parts in cds_parts.items():\n",
    "#         no_overlap_to_recode[gene_name] = {}\n",
    "#         for gene_seq, interval in parts.items():\n",
    "#             for m,(comp_gene_name,pos) in enumerate(CDSs.items()):\n",
    "#                 if interval[0] < interval[1]:#gene II case: gene overlaping with the last position and the first position of the file.\n",
    "#                     if pos[0] < interval[1] and interval[0] < pos[1]: #if overlap between region and compared region\n",
    "#                         if gene_name == comp_gene_name: #if we are looking at interval in same gene\n",
    "#                                 no_overlap_to_recode[gene_name][gene_seq] = interval\n",
    "#                             else: #if overlaps, but not in the same gene:\n",
    "#                                 if same_ORF(pos[1],pos[m+1][0]): #if the overlap is in the same ORF with the next gene\n",
    "                                    \n",
    "#                         else: #if comp_gene_name != gene_name, meaning an interval (seq to recode) overlaps with an actual CDSs in the construct other than himself\n",
    "#                             if same_ORF(interval[1],pos[0]):#if this genome section is in the same orf compared to the others, append it. Otherwise, exclude it.\n",
    "#                                 print(interval,pos)\n",
    "#                                 no_overlap_to_recode[gene_name][gene_seq] = (interval[1],pos[1])\n",
    "#                 else:\n",
    "#                     print(gene_name)\n",
    "#     return no_overlap_to_recode\n",
    "                    \n",
    "def find_forbidden_regions(path_to_gb_file):\n",
    "    #we have to exlcude promoter sequence, terminator, rbs and the oriR in the restricted regions dic:\n",
    "    #data structure : {start:end, start1:end1, start2:end2, start3:end3...}\n",
    "    forbidden_regions = []\n",
    "    CDSs = {}\n",
    "    #CDSs in the form of :{I:(0,1000), II:(980,1500),...}: it is the position of every gene!\n",
    "    for gb_record in SeqIO.parse(open(path_to_gb_file, \"r\"), \"genbank\"):\n",
    "        dna_seq = gb_record.seq\n",
    "        for feat in gb_record.features:\n",
    "            if \"prom\" in str(feat.qualifiers) or \"rbs\" in str(feat.qualifiers) or \"term\" in str(feat.qualifiers) or \"origin\" in str(feat.qualifiers):\n",
    "                initial_loc_forbid = str(feat.location)\n",
    "                start = int(initial_loc_forbid.split(\":\")[0][1:])\n",
    "                end = int(initial_loc_forbid.split(\":\")[1].split(\"]\")[0])\n",
    "                forbidden_regions.append((start,end))\n",
    "            if feat.type == \"CDS\" and \"join\":\n",
    "                if \"join\" not in str(feat.location):\n",
    "                    #join is here for gene II, because it is a circular construction (overlaps with the beginning and the end of the file)\n",
    "                    initial_loc_cds = str(feat.location)\n",
    "                    start = int(initial_loc_cds.split(\":\")[0][1:])\n",
    "                    end = int(initial_loc_cds.split(\":\")[1].split(\"]\")[0])\n",
    "                    CDSs[feat.qualifiers[\"label\"][0]] = (start,end)\n",
    "                else:\n",
    "                    initial_loc_cds_start = str(feat.location)[5:].split(\",\")[0]\n",
    "                    initial_loc_cds_end = str(feat.location).split(\",\")[1]\n",
    "                    start = int(initial_loc_cds_start.split(\":\")[0][1:])\n",
    "                    end = int(initial_loc_cds_end.split(\":\")[1].split(\"]\")[0])\n",
    "                    CDSs[feat.qualifiers[\"label\"][0]] = (start,end)\n",
    "    \n",
    "    return forbidden_regions,CDSs,dna_seq\n",
    "    \n",
    "def get_changing_cds_parts(coor,mforbiden_regions,dna_seq):\n",
    "    #coor is absolute coordinates of the fasta file\n",
    "    #returns all dna_string that could be recoded (excludes forbidden sequences (term_seq,prom_seq, rbs, rep origin...) and absolute pos in a dict \n",
    "    #{seq1:(start1,end1), seq2:(start2,end2),...}\n",
    "    cds_parts = {}\n",
    "    notpickreg = []\n",
    "    for i,regions in enumerate(mforbiden_regions):\n",
    "        #find all changeable part of CDSs, or exclude forbiden sequences:\n",
    "        if regions[0] > coor[0] and regions[0] < coor[1]: #Normal case: if, elif else\n",
    "            if not notpickreg: #if notpickreg empty\n",
    "                cds_parts[str(dna_seq[coor[0]:regions[0]])] = (coor[0],regions[0])\n",
    "                notpickreg = [regions[0],regions[1]]\n",
    "            else:\n",
    "                cds_parts[str(dna_seq[notpickreg[1]:regions[0]])] = (notpickreg[1],regions[0])\n",
    "                notpickreg = [regions[0],regions[1]]\n",
    "        elif regions[0] <= coor[0] and regions[1] >= coor[0]:\n",
    "            notpickreg = [regions[0],regions[1]]\n",
    "            if regions[1] >= coor[1]:\n",
    "                return []\n",
    "        else:\n",
    "            if regions[0] > coor[1]:  #if region[0] exceeds the limit of the cds\n",
    "                if not notpickreg: #meaning if no forbidden region in cds:\n",
    "                    cds_parts[str(dna_seq[coor[0]:coor[1]])] = (coor[0],coor[1])\n",
    "                else:\n",
    "                    if notpickreg[1] <= coor[1]:\n",
    "                        cds_parts[str(dna_seq[notpickreg[1]:coor[1]])] = (notpickreg[1],coor[1])\n",
    "                break\n",
    "        if regions == forbiden_regions[-1]: #last forbidden region, exceptionnal case:\n",
    "            if not notpickreg:\n",
    "                cds_parts[str(dna_seq[coor[0]:regions[0]])] = (coor[0],regions[0])\n",
    "            else:\n",
    "                cds_parts[str(dna_seq[notpickreg[1]:coor[1]])] = (notpickreg[1],coor[1])\n",
    "    return cds_parts\n",
    "\n",
    "def find_start_of_gene(last_pos,CDSs):\n",
    "    #find the corresponding gene with a position in the genome, returns always the first occurence of overlapping in the CDSs.\n",
    "    #Since uniq_all_locations_sorted is sorted, I do not see how this function may not work. \n",
    "    #If two in frame genes are compared in same_ORF(), they will both pointing in the same gene, and so be in frame. \n",
    "    for gene_name,pos in CDSs.items():\n",
    "        if pos[0] < pos[1]:\n",
    "            if last_pos >= int(pos[0]) and last_pos <= int(pos[1]):\n",
    "                return (gene_name,pos)\n",
    "                break\n",
    "        else: #for the gene that overlaps the last and the first coordinate in file:\n",
    "            if (last_pos <= int(pos[0]) and last_pos <= int(pos[1])) or (last_pos >= int(pos[0]) and last_pos >= int(pos[1])):\n",
    "                return (gene_name,pos)\n",
    "                break\n",
    "\n",
    "# main:\n",
    "forbiden_regions,CDSs,dna_seq = find_forbidden_regions(path_to_gb_file = \"../ref_files_m13/m13mp18-Kan_annotated.gb\")\n",
    "\n",
    "#add overlaping region in forbiden_regions:\n",
    "merge_forbiden_regions = merge_contiguous_regions(forbiden_regions)  #merge continuous region, but doesnt care about overlaping region.\n",
    "max_len_dna = len(str(dna_seq))\n",
    "#print(CDSs)\n",
    "\n",
    "cds_parts = {}\n",
    "for gene_name, coor in CDSs.items():\n",
    "    gene = dna_seq[coor[0]:coor[1]]\n",
    "    if coor[0] < coor[1]:\n",
    "        cds_parts[gene_name] = get_changing_cds_parts(coor,merge_forbiden_regions,dna_seq=dna_seq)\n",
    "    else:\n",
    "        cds_parts_end = get_changing_cds_parts((coor[0],len(dna_seq)),merge_forbiden_regions,dna_seq=dna_seq)\n",
    "        cds_parts_start = get_changing_cds_parts((0, coor[1]),merge_forbiden_regions,dna_seq=dna_seq)\n",
    "        cds_parts[gene_name] = {**cds_parts_end, **cds_parts_start}\n",
    "        \n",
    "#trouver les portions de genome qui s'overlappent et les exlcure!\n",
    "no_overlap_to_recode = {}\n",
    "all_locations = [interval for gene_name, parts in cds_parts.items() for gene_seq, interval in parts.items()]\n",
    "uniq_all_locations_sorted = sorted(list(set(all_locations)), key=lambda tup: (tup[0],tup[1])) #sorted by first value (and second) in tuple unique values.\n",
    "#print(uniq_all_locations_sorted)\n",
    "parts_to_recode = []\n",
    "wrong_frame_lastpos = 0 #the last position in the last cds not in frame:  \n",
    "for i,positions in enumerate(uniq_all_locations_sorted):\n",
    "    #print(positions)\n",
    "    append = True\n",
    "    if i+1 < len(uniq_all_locations_sorted):\n",
    "        new_coors = {}\n",
    "        next_coors = uniq_all_locations_sorted[i+1]\n",
    "        #print(\"next_coors : \",next_coors)\n",
    "        while next_coors[0] <= positions[1] and positions[0] <= next_coors[1]: #while there are overlapping regions.\n",
    "            #print(find_start_of_gene(int(positions[1]),CDSs)[0],find_start_of_gene(int(positions[1]),CDSs)[1][0],positions[1],next_coors[0])\n",
    "            #if next_coors[0] > or < find_start_of_gene(int(positions[1]),CDSs), the condition still works :)\n",
    "            #the start of reference gene compared together:\n",
    "            if same_ORF(find_start_of_gene(int(positions[1]),CDSs)[1][0],next_coors[0]):\n",
    "                if positions[1] == next_coors[1]: #should be always true, but just for safety:\n",
    "                    new_coors[(max([wrong_frame_lastpos,positions[0]]),positions[1])] = str(dna_seq)[max([wrong_frame_lastpos,positions[0]]):positions[1]] #not append directly, other cdss might overlap with a different ORF\n",
    "                else:\n",
    "                    print(\"two overlaping CDSs in the same ORF have different STOP codSon position. Please revise annotations\")\n",
    "                    print(\"the program exit in an unexcepted way.\")\n",
    "                    break\n",
    "            else: #we need to find the in frame coors! So a little bit more complicated...\n",
    "                if positions[0] < next_coors[0]: # we do not want to add empty sequences in dictionary:\n",
    "                    parts_to_recode.append({(max([wrong_frame_lastpos,positions[0]]),next_coors[0]) : str(dna_seq[max([wrong_frame_lastpos,positions[0]]):next_coors[0]]) })\n",
    "                if positions[1] < next_coors[1]: #meaning all the end of gene_ref has overlap with an altORF, so we  \n",
    "                    append = False\n",
    "                    wrong_frame_lastpos = positions[1]\n",
    "                    break\n",
    "                else:\n",
    "                    last_pos = next_coors[1]\n",
    "                wrong_frame_lastpos = positions[1]\n",
    "            #print(parts_to_recode)\n",
    "            try: #changing position to the next element of the list:\n",
    "                next_coors = uniq_all_locations_sorted[uniq_all_locations_sorted.index(next_coors[0],next_coors[1])+1]\n",
    "            except: #if we are at the end of the list\n",
    "                break\n",
    "\n",
    "    if append: #if something to append in the list of dicts:\n",
    "        if new_coors:\n",
    "            parts_to_recode.append(new_coors)\n",
    "        else: #if no specified coordinates, meaning no overlap in frame, simply add the given region:\n",
    "            parts_to_recode.append({(max([wrong_frame_lastpos,positions[0]]),positions[1]) : str(dna_seq[max([wrong_frame_lastpos,positions[0]]):positions[1]]) })\n",
    "    new_coors = {}\n",
    "    #print(parts_to_recode)\n",
    "    \n",
    "#remove repetitive parts of genome to recode. parts of genome are found based on each gene position. \n",
    "#If 2 genes overlap, it is possible a part of the genome is included\n",
    "all_pos = [tuple(list(dic.keys())[0]) for dic in parts_to_recode]\n",
    "removed_items = []\n",
    "for i in all_pos:\n",
    "    for j in all_pos:\n",
    "        if i != j and i[0] <= j[0] and i[1] >= j[1]:\n",
    "            if j not in removed_items:\n",
    "                parts_to_recode = [dic for dic in parts_to_recode if j not in dic] #remove j from the list of dicts\n",
    "#print(\"je suis le vrai dico: \", parts_to_recode)\n",
    "\n",
    "in_frame_parts_to_recode = []\n",
    "#get only in frame region and adjust the coordinates accordingly:\n",
    "#######if there are gene on different strand, this code is not correct#######\n",
    "\n",
    "for recode_regions in parts_to_recode:\n",
    "    init_pos,end_pos = (next(iter(recode_regions))[0],next(iter(recode_regions))[1]) #next(iter(recode_regions)) is the first and only key for each dict (aka the position of each region).\n",
    "    start_ref_gene = find_start_of_gene(end_pos,CDSs)[1][0]\n",
    "    end_ref_gene = find_start_of_gene(end_pos,CDSs)[1][1]\n",
    "    #note: this if was really difficult to write. I am NOT sure it is the most general answer to find the correct frame for eache region.\n",
    "    if start_ref_gene > end_ref_gene and init_pos < end_ref_gene: #exceptionnal case\n",
    "        len_dna_seq = len(dna_seq)\n",
    "        init_pos_to_zero = (3-(init_pos)%3)%3   #est-ce que ma position est en dans le meme cadre avec le 0?\n",
    "        end_pos_to_zero = (end_pos)%3\n",
    "        start_ref_gene_end_fileinframe = (3-(len_dna_seq-start_ref_gene)%3)%3 #est-ce que la fin du fichier est in frame avec le debut du gene qui overlap avec le debut et la fin du fichier?\n",
    "        start_ref_gene_end_fileinframe = (len_dna_seq-start_ref_gene)%3\n",
    "        inframe_init_pos = init_pos + init_pos_to_zero + start_ref_gene_end_fileinframe\n",
    "        inframe_end_pos = end_pos - start_ref_gene_end_fileinframe - end_pos_to_zero  #NOT GOOD\n",
    "        in_frame_parts_to_recode.append({(inframe_init_pos,inframe_end_pos) : str(dna_seq)[inframe_init_pos:inframe_end_pos]})\n",
    "        \n",
    "    else:  #for general case, find the gene with the first position in the list. Adjust coordinates to keep only in frame\n",
    "        inframe_init_pos = init_pos+(3-(init_pos-start_ref_gene)%3)%3\n",
    "        inframe_end_pos = end_pos-(3-(start_ref_gene-end_pos)%3)%3\n",
    "        in_frame_parts_to_recode.append({(inframe_init_pos,inframe_end_pos) : str(dna_seq)[inframe_init_pos:inframe_end_pos]})\n",
    "print(in_frame_parts_to_recode)\n",
    "\n",
    "#check if all the regions are now %3:\n",
    "for dics in in_frame_parts_to_recode:\n",
    "    if len(next(iter(dics.values())))%3 != 0:\n",
    "        print(\"REGIONS IN FRAME ARE NOT ALL %3, SUCH AS: \")\n",
    "        print(position,regions_in_frame)\n",
    "        print(\"PLEASE FIX THIS ISSUE, IT WILL OTHERWISE CAUSE MAJOR PROBLEM!\")\n",
    "\n",
    "#recode every in frame regions, recode at 25% each CDSs:\n",
    "recoded_in_frame_parts = [{next(iter(dics.keys())):recode_perc_change(25,next(iter(dics.values())),rev_ecoli_codon_usage)}\n",
    "                          for dics in in_frame_parts_to_recode]\n",
    "coor_min = 0\n",
    "coor_max = 0\n",
    "recoded_genome_parts = []\n",
    "for i,dics in enumerate(recoded_in_frame_parts):\n",
    "    position = next(iter(dics.keys()))\n",
    "    rec_seq = next(iter(dics.values()))\n",
    "    if position[0] != 0:\n",
    "        recoded_genome_parts.append(str(dna_seq)[0:position[0]])\n",
    "    else:\n",
    "        continue\n",
    "    if i+1 < len(recoded_in_frame_parts):\n",
    "        next_position = next(iter(recoded_in_frame_parts[i+1].keys()))\n",
    "        if position[1] < next_position[0]:\n",
    "            recoded_genome_parts.append(rec_seq)\n",
    "            recoded_genome_parts.append(str(dna_seq)[position[1]:next_position[0]])\n",
    "    else:\n",
    "        if position[1] != \n",
    "    \n",
    "    \n",
    "              \n",
    "### gtg pour le codon start: jamais recoder!\n",
    "### gene antisens: attention!\n",
    "\n",
    "        \n",
    "        \n",
    "        \n",
    "        #delta_inframe_init_pos = (find_start_of_gene(init_pos,CDSs)[1][0] - init_pos)%3 #delta inframe is the adjustement to make each start of each region in frame \n",
    "#il faut trouver une condition pour retrouver les cadres de lecture pour le gène pII.\n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATGAACATATATGACGATTCCGCAGTATTGGACGCTATCCAGTCAAAACATTTTACAATTACCCCGAGCGGCAAAACTTCTTTTGCGAAAGCCAGCCGGTATTTTGGTTTTTATCGTCGTCTGGTAAACGAAGGTTATGACAGTGTTGCCCTTACTATGCCTCGCAACTCCTTCTGGCGTTATGTATCTGCATTAGTTGAGTGCGGTATTCCCAAATCCCAACTGATGAATCTTAGTACTTGTAATAACGTTGTTCCGTTAGTGCGTTTCATT'"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recode_perc_change(25,\"ATGAATATTTATGACGATTCCGCAGTATTGGACGCTATCCAGTCTAAACATTTTACTATTACCCCCTCTGGCAAAACTTCTTTTGCAAAAGCCTCTCGCTATTTTGGTTTTTATCGTCGTCTGGTAAACGAGGGTTATGATAGTGTTGCTCTTACTATGCCTCGTAATTCCTTTTGGCGTTATGTATCTGCATTAGTTGAATGTGGTATTCCTAAATCTCAACTGATGAATCTTTCTACCTGTAATAATGTTGTTCCGTTAGTTCGTTTTATT\",rev_ecoli_codon_usage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#draft of may things!\n",
    "        \n",
    "        #append something\n",
    "        \n",
    "                \n",
    "            # else:\n",
    "            #     new_coors = {}\n",
    "            #     wrong_frame = True\n",
    "            #     wrong_frame_lastpos = next_coors[1]\n",
    "            #     if next_coors[1] < positions[1]: #in the case the next cds (next_coors) is not in frame and shorter than the cds it is compared with:\n",
    "            #         #appending the part before next_coors[0]:\n",
    "            #         if not cds_smaller:\n",
    "            #             parts_to_recode.append({str(dna_seq[positions[0]:next_coors[0]]):(positions[0],next_coors[0])})\n",
    "            #             last_start = next_coors[0]\n",
    "            #             last_end = next_coors[1]\n",
    "            #             cds_smaller = True\n",
    "            #             try:\n",
    "            #                 next_coors = all_locations[all_locations.index(next_coors[0],next_coors[1])+1]\n",
    "            #             except: #if we are at the end of the list, need to append the recoded parts not overlaping with the little cds:                                                                        \n",
    "            #                 #appending the last part of the recodable cds, excluding the cds not in frame:\n",
    "            #                 new_coors[dna_seq[next_coors[1]:positions[1]]] = (next_coors[1],positions[1])\n",
    "            #                 parts_to_recode.append({str(dna_seq[next_coors[1]:positions[1]]):(next_coors[1],positions[1])})\n",
    "            #         else: #if multiple smaller CDSs\n",
    "            #             parts_to_recode.append({str(dna_seq[last_start:next_coors[0]]):(positions[0],next_coors[0])})\n",
    "            #             last_start = next_coors[0]\n",
    "            #             try:\n",
    "            #                 next_coors = all_locations[all_locations.index(next_coors[0],next_coors[1])+1]\n",
    "            #             except: #if we are at the end of the list, need to append the recoded parts not overlaping with the little cds:                                                                        \n",
    "            #                 #appending the last part of the recodable cds, excluding the cds not in frame:\n",
    "            #                 new_coors[dna_seq[next_coors[1]:positions[1]]] = (next_coors[1],positions[1])\n",
    "            #             break\n",
    "            #     else:\n",
    "            #         in_frame_startpos = last_end+((3-last_end-positions[0])%3)\n",
    "            #         in_frame_end_pos = ###ACHANGER POUR LE IN FRAME ET REMPLACER LE NEXT_COORD[0] PLUS BAS!\n",
    "            #         if cds_smaller and in_frame_startpos =< next_coors[0]+3: #these two might overlap! Get at least 1 codon !\n",
    "            #             in_frame_pos = (3-last_end-positions[0])%3\n",
    "            #             parts_to_recode.append({str(dna_seq[in_frame_pos:next_coors[0]]):(positions[0],next_coors[0])})\n",
    "            #         break\n",
    "            # try:\n",
    "            #     next_coors = all_locations[all_locations.index(next_coors[0],next_coors[1])+1]\n",
    "            # except: #if we are at the end of the list\n",
    "            #     break\n",
    "                \n",
    "    # elif wrong_frame: #do not append in dict if overlaping region not in frame\n",
    "    #     if cds_smaller:\n",
    "    #         parts_to_recode.append({str(dna_seq[next_coors[1]:positions[1]]):(next_coors[1],positions[1])})\n",
    "    #     continue\n",
    "    # elif new_coors:\n",
    "    #     parts_to_recode.append(new_coors)\n",
    "    # elif not new_coors:\n",
    "    #     parts_to_recode.append({str(dna_seq[positions[0]:positions[1]]):(positions[0],positions[1])})\n",
    "    # print(parts_to_recode)\n",
    "    \n",
    "# CDSs_no_overlaps = {}\n",
    "# whose_overlaps = {}   #gene key overlaps with gene value.\n",
    "# for i, (gene_name, coors) in enumerate(CDSs.items()):\n",
    "#     CDSs_no_overlaps[gene_name] = {}\n",
    "#     if i+1 < len(CDSs):\n",
    "#         next_coors = list(CDSs.values())[i+1]\n",
    "#         new_coors = {}\n",
    "#         while next_coors[0] <= coors[1] and coors[0] <= next_coors[1] : # if overlap between 2 cdss\n",
    "#             whose_overlaps[list(CDSs.keys())[i+1]] = gene_name\n",
    "#             if gene_name in whose_overlaps: #meaning gene key starts not at the begining of itself, rather after the known end of previously fixed overlapé\n",
    "#                 max_start_of_gene = max([CDSs[gene][1] for gene in whose_overlaps if gene == gene_name])\n",
    "#             else:\n",
    "#                 max_start_of_gene = coors[0]\n",
    "#             if same_ORF(coors[1],next_coors[0]): #if overlaping cds are in frame, it is possible to recode\n",
    "#                 new_coors[str(dna_seq[max_start_of_gene:next_coors[0]])] = (max_start_of_gene,coors[1])\n",
    "#             else: #if there is just one cds not in frame, even if there is other overlaping cds, we cannot recode this part\n",
    "#                 new_coors[str(dna_seq[max_start_of_gene:next_coors[0]])] = (max_start_of_gene,next_coors[0])\n",
    "#                 break\n",
    "#             next_coors = list(CDSs.values())[list(CDSs.values()).index((next_coors[0],next_coors[1]))+1]\n",
    "#             print(next_coors,gene_name)\n",
    "#         if not new_coors:\n",
    "#             CDSs_no_overlaps[gene_name] = {str(dna_seq[coors[0]:coors[1]]):(coors[0],coors[1])}\n",
    "            \n",
    "# print(CDSs_no_overlaps)\n",
    "\n",
    "            \n",
    "            \n",
    "\n",
    "\n",
    "\n",
    "# cds_part_no_prob_overlap = overlap_situation(cds_parts,CDSs)\n",
    "# print(cds_part_no_prob_overlap)\n",
    "# #print(\"j suis quoi???\",cds_part_no_prob_overlap)\n",
    "\n",
    "# recoded_genome_parts = {}\n",
    "# for gene_name, parts in cds_part_no_prob_overlap.items():\n",
    "#     recoded_genome_parts[gene_name] = {}\n",
    "#     for part,coor in parts.items():\n",
    "#         #Search for orf coordinates and recoded the cds:\n",
    "#         start_part = coor[0]\n",
    "#         end_part = coor[1]\n",
    "#         start_codon = start_part+(start_part%3)\n",
    "#         end_codon = end_part-(end_part%3)\n",
    "#         if start_part < end_part:\n",
    "#             in_frame_part = dna_seq[start_codon:end_codon]\n",
    "#             recoded_part = str(recode_perc_change(50,in_frame_part,rev_ecoli_codon_usage))\n",
    "#             if recoded_part != \"\":\n",
    "#                 recoded_genome_parts[gene_name][recoded_part] = (start_codon,end_codon)\n",
    "#         else:\n",
    "#             print(\"minou\")\n",
    "#             in_frame_part_end = dna_seq[start_codon:max_len_dna]\n",
    "#             in_frame_part_start = dna_seq[0:end_codon]\n",
    "#             recoded_part_end = str(recode_perc_change(50,in_frame_part_end,rev_ecoli_codon_usage))\n",
    "#             recoded_part_start = str(recode_perc_change(50,in_frame_part_start,rev_ecoli_codon_usage))\n",
    "#             if recoded_part != \"\":\n",
    "#                 recoded_genome_parts[gene_name][in_frame_part_end] = (start_codon,max_len_dna)\n",
    "#                 recoded_genome_parts[gene_name][recoded_part_start] = (0,end_codon)       \n",
    "# print(recoded_genome_parts)\n",
    "\n",
    "\n",
    "#construct the final m13 vector:\n",
    "# all_recoded_start_coor = [str(list(recoded_parts.values())).split(\",\")[0][2:] for recoded_parts in recoded_genome_parts.values()]\n",
    "# all_recoded_start_coor = [(recoded_parts.values()) for recoded_parts in recoded_genome_parts.values()]\n",
    "# print(all_recoded_start_coor)\n",
    "# all_recoded_end_coor = [coor[1] for seq,coor in parts.items() for recoded_parts in recoded_genome_parts.values()]\n",
    "# flat_recoded_cds = {coor:seq for seq,coor in partsATGA.items() for parts in recoded_genome_parts.values()}\n",
    "# print(flat_recoded_cds)\n",
    "\n",
    "# max_len_dna = len(str(dna_seq))[1,2,3,4,5,6,7,8].index(2)+1\n",
    "# m13_WTZ = \"\"\n",
    "# last_pos = 0\n",
    "# for i in range(max_len_dna):\n",
    "#     if i in all_recoded_start_coor:\n",
    "#         m13_WTZ += dna_seq[last_pos:i]\n",
    "#         last_pos = i\n",
    "#     elif i in all_recoded_end_coor:\n",
    "#         m13_WTZ += dna_seq[last_pos:i]\n",
    "#         last_pos = i\n",
    "#     else:\n",
    "#         continue\n",
    "# print(len(m13_WTZ))\n",
    "        \n",
    "        \n",
    "#     print(\"old\",cds_parts, \"\\n\")    \n",
    "#     #le gene en chevauche-t-il un autre? Si oui, couper la partie correspondante dans la liste de gene\n",
    "#     cds_parts = overlap_situation(coor,cds_parts) \n",
    "#     print(\"new\",cds_parts,\"\\n\")\n",
    "#     print(old_cds_parts==cds_parts)    \n",
    "#     cds_recoded = {}\n",
    "#     traceback = []\n",
    "        \n",
    "                 \n",
    "        \n",
    "        \n",
    "        #when forbiden_regions[0] < coor[0]\n",
    "    #find the reading frame in each part of changeable CDSs\n",
    "    \n",
    "    #change the reading frame in cds\n",
    "    \n",
    "    #merge all substring into a new file"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
