{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqUtils import GC\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.pyplot import figure\n",
    "import statistics\n",
    "import operator\n",
    "from tqdm.notebook import tqdm\n",
    "from functions import *\n",
    "import random\n",
    "from Bio import Restriction\n",
    "import copy\n",
    "from itertools import combinations\n",
    "from itertools import permutations\n",
    "from itertools import product\n",
    "import urllib.request as urllib2\n",
    "from dna_features_viewer import BiopythonTranslator\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'K': {'AAA': 0.767, 'AAG': 0.233}, 'N': {'AAT': 0.448, 'AAC': 0.552}, 'I': {'ATA': 0.07, 'ATT': 0.509, 'ATC': 0.421}, 'M': {'ATG': 1.0}, 'T': {'ACA': 0.129, 'ACT': 0.165, 'ACC': 0.438, 'ACG': 0.268}, 'R': {'AGA': 0.036, 'AGG': 0.02, 'CGA': 0.063, 'CGT': 0.383, 'CGC': 0.402, 'CGG': 0.097}, 'S': {'AGT': 0.15, 'AGC': 0.278, 'TCA': 0.122, 'TCT': 0.146, 'TCC': 0.149, 'TCG': 0.154}, '*': {'TAA': 0.64, 'TAG': 0.07, 'TGA': 0.289}, 'Y': {'TAT': 0.568, 'TAC': 0.432}, 'L': {'TTA': 0.13, 'TTG': 0.128, 'CTA': 0.036, 'CTT': 0.103, 'CTC': 0.105, 'CTG': 0.498}, 'F': {'TTT': 0.573, 'TTC': 0.427}, 'C': {'TGT': 0.442, 'TGC': 0.558}, 'W': {'TGG': 1.0}, 'Q': {'CAA': 0.347, 'CAG': 0.653}, 'H': {'CAT': 0.569, 'CAC': 0.431}, 'P': {'CCA': 0.19, 'CCT': 0.157, 'CCC': 0.123, 'CCG': 0.529}, 'E': {'GAA': 0.69, 'GAG': 0.31}, 'D': {'GAT': 0.626, 'GAC': 0.374}, 'V': {'GTA': 0.154, 'GTT': 0.258, 'GTC': 0.216, 'GTG': 0.372}, 'A': {'GCA': 0.212, 'GCT': 0.161, 'GCC': 0.27, 'GCG': 0.357}, 'G': {'GGA': 0.107, 'GGT': 0.338, 'GGC': 0.406, 'GGG': 0.149}}\n"
     ]
    }
   ],
   "source": [
    "#Create a codon table for e.coli:\n",
    "all_name_cds_ecoli = {str(record.description).split(\"gene=\")[1].split(\"]\",1)[0]:str(record.seq)\n",
    "                      for record in SeqIO.parse(\"../ref_files_m13/GCF_000005845.2_ASM584v2_cds_from_genomic.fna\", \"fasta\")}\n",
    "codon_table_ecoli = create_codon_table(list(all_name_cds_ecoli.values()), 11)\n",
    "\n",
    "# Make a reversed dict codon usage of whole genome\n",
    "rev_ecoli_codon_usage = reverse_codon_table(codon_table_ecoli)\n",
    "print(rev_ecoli_codon_usage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output file has been created\n"
     ]
    }
   ],
   "source": [
    "def recode_son(perc,fastafilename, path_to_tsv_file,outputdirname):\n",
    "    \n",
    "    def merge_contiguous_regions(list_of_tuples):\n",
    "\n",
    "        \"\"\"\n",
    "        ````\n",
    "        input: list of tuples\n",
    "\n",
    "        output: list of tuples merged (no overlaping intervals)\n",
    "\n",
    "        short description: Merge overlaping annotations in one forbiden region.\n",
    "\n",
    "        Complete description: compare tup[0] and tup[1] and find if there is overlap. If it does, it appends the min and the max. \n",
    "        If there is still overlap between the current overlap and the next tup (tup[2]) in list, min/max values are updated. If comp is last, append min_range[0]/max_range[0]\n",
    "        Elif there is overlap with the previous min/max range: continue (except if we are at the end of the list, in this case append minmax range).\n",
    "        Else (no overlap): if there is still no overlap in the list (max_range and min_range not updated): append ref values as interval.\n",
    "                          else (there was overlap, but there is no more): append last updated min and max\n",
    "                          if comp is last (no overlaps): append comp[0],comp[1]\n",
    "        ````\n",
    "        \"\"\"\n",
    "\n",
    "        merge_list = []\n",
    "        max_range = [-1]\n",
    "        max_list_of_tuples = max([max(i) for i in list_of_tuples])\n",
    "        min_range = [max_list_of_tuples+1]\n",
    "        for nb,ref in enumerate(list_of_tuples):\n",
    "            if nb+1 < len(list_of_tuples):    #to make sure we do not compared with something outside the list!\n",
    "                comp = list_of_tuples[nb+1]\n",
    "                if comp[0] <= ref[1] and comp[1] >= ref[0]: #if there is overlap with the next element in list:\n",
    "                    min_temp = min([ref[0],comp[0]])\n",
    "                    max_temp = max([ref[1],comp[1]])\n",
    "                    min_range.append(min_temp)\n",
    "                    max_range.append(max_temp)\n",
    "                    max_range = [max(max_range)]\n",
    "                    min_range = [min(min_range)]\n",
    "                    if nb+1 == len(list_of_tuples)-1: #if last, but overlap\n",
    "                        merge_list.append((min_range[0],max_range[0]))\n",
    "                elif comp[0] <= max_range[0] and comp[1] >= min_range[0]: #check if there is overlap with a previous border:\n",
    "                    if nb+1 == len(list_of_tuples)-1: #if last,append the last min_range/max_range.\n",
    "                        merge_list.append((min_range[0],max_range[0]))\n",
    "                    continue           \n",
    "                else:    #if no overlap\n",
    "                    if max_range == [-1] and min_range == [max_list_of_tuples+1]:    #no overlap between element in the list yet.\n",
    "                        merge_list.append((ref[0],ref[1]))\n",
    "                        max_range = [-1]\n",
    "                        min_range = [max_list_of_tuples+1]\n",
    "\n",
    "                    else: #there was overlap once, but there is no more: append min_range and max_range\n",
    "                        merge_list.append((min_range[0],max_range[0]))\n",
    "                        max_range = [-1]\n",
    "                        min_range = [max_list_of_tuples+1]\n",
    "\n",
    "                    if nb+1 == len(list_of_tuples)-1:  #if last but no overlap\n",
    "                         merge_list.append((comp[0],comp[1]))\n",
    "        return merge_list\n",
    "\n",
    "    def same_ORF(interval, ref_pos_gene):\n",
    "        #this function tells if coordinates of a part of a cds to be recoded overlaps with an actual gene (other than itself)\n",
    "        #Note: if the pos[0]-interval[0] < 3, the region is excluded by default.\n",
    "        return bool((ref_pos_gene-interval)%3 == 0)\n",
    "\n",
    "    def recode_perc_change(perc,cds,codon_table,pos_modifier):\n",
    "\n",
    "        \"\"\"\n",
    "        ````\n",
    "        Input: \n",
    "        perc: % of codon we want to change. type: int\n",
    "        cds: sequence to recode. type : str\n",
    "        codon_table: {a.a :{codon:freq,...},...}, type=dict of dict\n",
    "        pos_modifier: modify the seed at each time we go through this function. \n",
    "                      By doing so, we avoid chosing the same codon at the same position in two different cds to recode.\n",
    "\n",
    "        Output: recoded cds (type=str), number of changed codon (int) in all genome.\n",
    "\n",
    "        short description: Recode a coding sequence at a given percentage based on the codon table provided.\n",
    "        Complete description: Recode at perc or at max_perc: if the perc of recoding is higher than the maximum possible (because of codon without synonymous options)\n",
    "                            aka atg and tgg), recode at max_perc. Find the number of codons to recode in a cds and choose randomly in the list of codons.\n",
    "                            The synonymous codon (recoded) is found with the codon abundancies in e.coli genome, the initial codon (to be recoded) is removed from the dictionary.\n",
    "\n",
    "        ````\n",
    "        \"\"\"\n",
    "        ref_pos_modifier = pos_modifier\n",
    "        len_cds = len(cds)\n",
    "        nb_of_codons = int(len_cds/3)\n",
    "        nb_codons_to_change = round((perc/100)*nb_of_codons)\n",
    "        codons_pos = [(cds[pos*3:pos*3+3].upper(),pos) for pos in range(nb_of_codons)]    #[(codon0,pos0),(codon1,pos1),...]\n",
    "        do_not_recode = [\"ATG\",\"TGG\"]\n",
    "        nb_of_codon_imposs_to_recode = len([i for i in codons_pos if i[0] in do_not_recode])\n",
    "        max_possible_nb_codons_to_change = nb_of_codons - nb_of_codon_imposs_to_recode\n",
    "        perc_max = round((max_possible_nb_codons_to_change/nb_of_codons)*100,3)\n",
    "        if perc_max < perc:\n",
    "            print(f\"It is impossible to recode at {perc}% with those forbiden codons {do_not_recode} this cds part {cds}.\")\n",
    "            print(f\"At {perc}% of recoding, {nb_codons_to_change} codons have to be recoded, but there are {max_possible_nb_codons_to_change} codons that can be changed...\")\n",
    "            print(f\"The max %of recoding for this cds part is : {perc_max}\")\n",
    "            perc = perc_max\n",
    "            nb_codons_to_change = max_possible_nb_codons_to_change\n",
    "\n",
    "        codons_pos_without_forbiden = [i for i in codons_pos if i[0] not in do_not_recode]\n",
    "        random.seed(pos_modifier)\n",
    "        codons_to_change = random.sample(codons_pos_without_forbiden, nb_codons_to_change)\n",
    "\n",
    "        codons_recoded = []\n",
    "        for i,codon_pos in enumerate(codons_to_change):\n",
    "            codon_to_recode = codon_pos[0]\n",
    "            pos = codon_pos[1]\n",
    "            other_codons = {}\n",
    "            for aa, codon_freq in codon_table.items():\n",
    "                if codon_to_recode in codon_freq:\n",
    "                    #select the dict for the codon to recode, and evaluate the freq of synonymous codon without codon_in_cds.\n",
    "                    for codon,freq in codon_freq.items():\n",
    "                        if codon_to_recode != codon:\n",
    "                            other_codons[codon] = freq\n",
    "            #no need to ponderate frequency now that the codon_in_cds is removed because relative weights are ponderate by default with cumulative wieght (sum of weights)\n",
    "            #https://docs.python.org/dev/library/random.html#random.choices\n",
    "            #https://stackoverflow.com/questions/55440986/cumulative-weights-in-random-choices\n",
    "            random.seed(ref_pos_modifier+i)    #set the seed\n",
    "            #print(codon_to_recode,list(other_codons.keys()))\n",
    "            recoded_codon = random.choices(population=list(other_codons.keys()),weights=list(other_codons.values()),k=1)\n",
    "            #print(recoded_codon[0])\n",
    "            codons_recoded.append((pos,recoded_codon[0]))\n",
    "\n",
    "        if len(codons_recoded) != len(list(set(codons_recoded))):\n",
    "            print(\"BIIG MISTAKE IN THE CODE. PLS REVISE IT !!!! (in recode_perc_change function)\")\n",
    "\n",
    "        dct_codons_recoded = dict(codons_recoded)\n",
    "        #create a string with the list of codon recoded:\n",
    "        recoded_cds = \"\"\n",
    "        for pos in range(int(len_cds/3)):\n",
    "            if pos in dct_codons_recoded:   #key: position, value: recoded_codon\n",
    "                recoded_cds += dct_codons_recoded[pos]\n",
    "            else:\n",
    "                recoded_cds+= codons_pos[pos][0]\n",
    "        return recoded_cds,nb_codons_to_change,perc\n",
    "\n",
    "\n",
    "    def get_part_in_frame_parts_to_recode(perc,cds_parts,len_dna_seq):       \n",
    "        \"\"\"\n",
    "        ````\n",
    "        input: cds_parts and len_dna_seq (length of cds to recode)\n",
    "\n",
    "        output: list of every dna element to join => getting the wild type zone \n",
    "\n",
    "        short description: %3 each cds part depending on their position, recode, assemble dna parts (recoded and wt) in a list to form complete recoded construction. \n",
    "\n",
    "        Complete description:\n",
    "        First_part: A): Not in frame overlaps cds_part are excluded and B): redundant parts in cds_part are excluded too. input: cds_part, output: parts_to_recode NOT IN FRAME (type=dict, key=coordinates, value=dna_seq (str))\n",
    "        Second part: In frame cds_parts only: adjust the coordinates accordingly. NOTE: gene orientation is not considered here (always + considered)\n",
    "        Third part: Recode portion of CDS and assemble all strings (recoded and wt) to create the recoded genome.\n",
    "        ````\n",
    "        \"\"\"\n",
    "\n",
    "        #First part:\n",
    "        #A):\n",
    "        no_overlap_to_recode = {}\n",
    "        all_locations = [interval for gene_name, parts in cds_parts.items() for gene_seq, interval in parts.items()]\n",
    "        #cds_parts may contain redundant sequences (when 2 cds overlap, for instance). No need to recode the same cds_part more than once.\n",
    "        uniq_all_locations_sorted = sorted(list(set(all_locations)), key=lambda tup: (tup[0],tup[1])) #sorted by first value (and second) in tuple unique values.\n",
    "        parts_to_recode = []\n",
    "        wrong_frame_lastpos = 0 #the last position in the last cds not in frame:  \n",
    "        for i,positions in enumerate(uniq_all_locations_sorted):\n",
    "            append = True\n",
    "            if i+1 < len(uniq_all_locations_sorted):\n",
    "                new_coors = {}\n",
    "                next_coors = uniq_all_locations_sorted[i+1]\n",
    "                #print(\"next_coors : \",next_coors)\n",
    "                while next_coors[0] <= positions[1] and positions[0] <= next_coors[1]: #while there are overlaping regions.\n",
    "                    #if next_coors[0] > or < find_start_of_gene(int(positions[1]),CDSs), the condition still works :)\n",
    "                    #see if start of gene (found with position) is in frame with the next element in uniq_all_locations_sorted (sorted list of tuples).\n",
    "                    if same_ORF(find_start_of_gene(int(positions[1]),CDSs)[1][0],next_coors[0]):  \n",
    "                        if positions[1] == next_coors[1]: #should always be true, but just for safety see else print error message for more info:\n",
    "                            new_coors[(max([wrong_frame_lastpos,positions[0]]),positions[1])] = str(dna_seq)[max([wrong_frame_lastpos,positions[0]]):positions[1]] #not append directly, other cdss might overlap with a different ORF\n",
    "                        else:\n",
    "                            print(\"two overlaping CDSs in the same ORF have different STOP codon position. Please revise annotations\")\n",
    "                            print(\"the program exit in an unexcepted way.\")\n",
    "                            break\n",
    "                    else: \n",
    "                        if positions[0] < next_coors[0]: # we do not want to add empty sequences in dictionary:\n",
    "                            parts_to_recode.append({(max([wrong_frame_lastpos,positions[0]]),next_coors[0]) : str(dna_seq[max([wrong_frame_lastpos,positions[0]]):next_coors[0]]) })\n",
    "                        if positions[1] < next_coors[1]: #meaning all the end of gene_ref has overlap with an other ORF\n",
    "                            append = False\n",
    "                            wrong_frame_lastpos = positions[1]\n",
    "                            break\n",
    "                        else:\n",
    "                            last_pos = next_coors[1]\n",
    "                        wrong_frame_lastpos = positions[1]\n",
    "                    try: #changing position to the next element of the list:\n",
    "                        next_coors = uniq_all_locations_sorted[uniq_all_locations_sorted.index(next_coors[0],next_coors[1])+1]\n",
    "                    except: #if we are at the end of the list\n",
    "                        break\n",
    "\n",
    "            if append: #if something to append in the list of dicts:\n",
    "                if new_coors:\n",
    "                    parts_to_recode.append(new_coors)\n",
    "                else: #if no specified coordinates, meaning no overlap in frame, simply add the given region:\n",
    "                    parts_to_recode.append({(max([wrong_frame_lastpos,positions[0]]),positions[1]) : str(dna_seq[max([wrong_frame_lastpos,positions[0]]):positions[1]]) })\n",
    "            new_coors = {}\n",
    "\n",
    "\n",
    "        #B):\n",
    "        #remove repetitive parts of genome to recode. parts of genome are found based on each gene position. \n",
    "        #If 2 genes overlap, it is possible a part of the genome is included twice:\n",
    "        all_pos = [tuple(list(dic.keys())[0]) for dic in parts_to_recode]\n",
    "        removed_items = []\n",
    "        for i in all_pos:\n",
    "            for j in all_pos:\n",
    "                if i != j and i[0] <= j[0] and i[1] >= j[1]:\n",
    "                    if j not in removed_items:\n",
    "                        parts_to_recode = [dic for dic in parts_to_recode if j not in dic] #remove j from the list of dicts\n",
    "\n",
    "\n",
    "        #Second part:\n",
    "        in_frame_parts_to_recode = []\n",
    "\n",
    "        #######if there are genes on different strands, this code is not correct#######\n",
    "        for recode_regions in parts_to_recode:\n",
    "            init_pos,end_pos = (next(iter(recode_regions))[0],next(iter(recode_regions))[1]) #next(iter(recode_regions)) is the first and only key for each dict (aka the position of each region).\n",
    "            start_ref_gene = find_start_of_gene(end_pos,CDSs)[1][0]\n",
    "            end_ref_gene = find_start_of_gene(end_pos,CDSs)[1][1]\n",
    "            #note: this if was really difficult to write. I am NOT sure it is the most general answer to find the correct frame for eache region.\n",
    "            if start_ref_gene > end_ref_gene and init_pos < end_ref_gene: #exceptionnal case\n",
    "                init_pos_to_zero = (3-(init_pos)%3)%3   #est-ce que ma position est dans le meme cadre avec le 0?\n",
    "                end_pos_to_zero = (end_pos)%3\n",
    "                start_ref_gene_end_fileinframe = (3-(len_dna_seq-start_ref_gene)%3)%3 #est-ce que la fin du fichier est in frame avec le debut du gene qui overlap avec le debut et la fin du fichier?\n",
    "                start_ref_gene_end_fileinframe = (len_dna_seq-start_ref_gene)%3\n",
    "                inframe_init_pos = init_pos + init_pos_to_zero + start_ref_gene_end_fileinframe\n",
    "                inframe_end_pos = end_pos - start_ref_gene_end_fileinframe - end_pos_to_zero  #NOT GOOD\n",
    "                in_frame_parts_to_recode.append({\"pos\":(inframe_init_pos,inframe_end_pos), \"dna_seq\": str(dna_seq)[inframe_init_pos:inframe_end_pos]})\n",
    "\n",
    "            else:  #for general case, find the gene with the first position in the list. Adjust coordinates to keep only in frame\n",
    "                inframe_init_pos = init_pos+(3-(init_pos-start_ref_gene)%3)%3\n",
    "                inframe_end_pos = end_pos-(3-(start_ref_gene-end_pos)%3)%3\n",
    "                in_frame_parts_to_recode.append({\"pos\":(inframe_init_pos,inframe_end_pos),\"dna_seq\": str(dna_seq)[inframe_init_pos:inframe_end_pos]})\n",
    "\n",
    "\n",
    "        #check if all the regions are now %3:\n",
    "        for dics in in_frame_parts_to_recode:\n",
    "            if len(dics[\"dna_seq\"])%3 != 0:\n",
    "                print(\"REGIONS IN FRAME ARE NOT ALL %3, SUCH AS: \")\n",
    "                print(dics[\"dna_seq\"])\n",
    "                print(\"PLEASE FIX THIS ISSUE, IT WILL OTHERWISE CAUSE MAJOR PROBLEM!\")\n",
    "\n",
    "        #Third part\n",
    "        #recode every in frame regions, recode at perc% each CDSs:\n",
    "        recoded_in_frame_parts = []\n",
    "        total_changed_codons = 0\n",
    "        for i,dics in enumerate(in_frame_parts_to_recode):\n",
    "            pos = dics[\"pos\"]\n",
    "            cds = dics[\"dna_seq\"]\n",
    "            recoded_part, nb_changed_codons, annotations = recode_perc_change(perc,cds,rev_ecoli_codon_usage,i)\n",
    "            total_changed_codons += nb_changed_codons\n",
    "            recoded_in_frame_parts.append({\"pos\":pos, \"dna_recoded\":recoded_part, \"annotations\":annotations})\n",
    "\n",
    "        coor_min = 0\n",
    "        coor_max = 0\n",
    "        recoded_genome_parts = []\n",
    "        len_recoded_in_frame_parts = len(recoded_in_frame_parts)\n",
    "        #this last part is to construct the genome. Assemble wt parts and recoded parts.\n",
    "        for i,dics in enumerate(recoded_in_frame_parts):\n",
    "            position = dics[\"pos\"]\n",
    "            rec_seq = dics[\"dna_recoded\"]\n",
    "            annotation = dics[\"annotations\"]\n",
    "            if i == 0:    #the first element is the beginning of the fasta, not sure if this is the cleanest way to assemble the string, but it should work e\n",
    "                if position[0] != 0:\n",
    "                    recoded_genome_parts.append(str(dna_seq)[0:position[0]])\n",
    "                    recoded_genome_parts.append(rec_seq)\n",
    "                    try: #if the list contains only one element, id like the code to work fine:\n",
    "                        next_position = recoded_in_frame_parts[i+1][\"pos\"]\n",
    "                        recoded_genome_parts.append(str(dna_seq)[position[1]:next_position[0]])\n",
    "                    except: #if we are at the end of the list, meaning only one region has to be recoded.\n",
    "                        #should be pretty rare when I think of it\n",
    "                        break\n",
    "                else:\n",
    "                    recoded_genome_parts.append(rec_seq)\n",
    "                    try: #if the list contains only one element, id like the code to work fine:\n",
    "                        next_position = recoded_in_frame_parts[i+1][\"pos\"]\n",
    "                        recoded_genome_parts.append(str(dna_seq)[position[1]:next_position[0]])\n",
    "                    except: #if we are at the end of the list, meaning only one region has to be recoded.\n",
    "                        break\n",
    "            elif i+1 < len_recoded_in_frame_parts: #jusqu'au dernier element de la liste:\n",
    "                next_position = recoded_in_frame_parts[i+1][\"pos\"]\n",
    "                if position[1] < next_position[0]:\n",
    "                    recoded_genome_parts.append(rec_seq)\n",
    "                    recoded_genome_parts.append(str(dna_seq)[position[1]:next_position[0]])\n",
    "                if recoded_in_frame_parts[i+1] == recoded_in_frame_parts[-1]: #if last element of the list of dicts:\n",
    "                    recoded_genome_parts.append(recoded_in_frame_parts[i+1][\"dna_recoded\"])\n",
    "                    if next_position[1] < len(dna_seq):\n",
    "                        recoded_genome_parts.append(str(dna_seq)[next_position[1]:])\n",
    "\n",
    "        return recoded_genome_parts, recoded_in_frame_parts, total_changed_codons\n",
    "\n",
    "    def find_start_of_gene(last_pos,CDSs):\n",
    "        #find the corresponding gene with a given position in the genome. Returns always the first hit in CDSs, meaning the gene with the smallest start position.\n",
    "        #Since uniq_all_locations_sorted is sorted, this function works. Otherwise, it would not.\n",
    "        #If two in frame genes are compared in same_ORF(), they will both pointing in the same gene and so in frame.\n",
    "        for gene_name,pos in CDSs.items():\n",
    "            if pos[0] < pos[1]:\n",
    "                if last_pos >= int(pos[0]) and last_pos <= int(pos[1]):\n",
    "                    return (gene_name,pos)\n",
    "                    break\n",
    "            else: #for the gene that overlaps the last and the first coordinate in fasta file:\n",
    "                if (last_pos <= int(pos[0]) and last_pos <= int(pos[1])) or (last_pos >= int(pos[0]) and last_pos >= int(pos[1])):\n",
    "                    return (gene_name,pos)\n",
    "                    break\n",
    "\n",
    "    def find_forbidden_regions(path_to_fasta_file, path_to_tsv_file):\n",
    "        \"\"\"\n",
    "        ```\n",
    "        Entree: \n",
    "        fichier fasta et d'annotation (tsv). \n",
    "\n",
    "        Sortie : \n",
    "        dna_seq : type=str (upper_case): sequence d'ADN;\n",
    "        forbiden_sequences: type=list(tuples) : Toutes positions (start,end) annotees comme n'etant pas des genes ou des CDSs dans une liste de tuples.\n",
    "        CDSs: type=dict ({'pX': (495, 831), 'pV': (842, 1106), 'pVII': (1107, 1209)...}): nom du gene comme cle, position du gene (min,max)\n",
    "\n",
    "        Description: Creation initiale d'objets utiles pour la construction d'un genome recode. Toutes sequences n'etant pas un gene ou un cds est consideree\n",
    "        comme etant une region a eviter pour le recodage. La coordonee start a ete ajustee (le tsv de snapgene est en 1 base, pas python).  \n",
    "        ````\n",
    "        \"\"\"\n",
    "\n",
    "        forbidden_regions = []\n",
    "        CDSs = {}\n",
    "        #CDSs in the form of :{I:(0,1000), II:(980,1500),...}: it is the position of every gene!\n",
    "        # get the fasta file\n",
    "        for fasta in SeqIO.parse(open(path_to_fasta_file, \"r\"), \"fasta\"):\n",
    "            dna_seq = str(fasta.seq)\n",
    "        ann_df = pd.read_csv(path_to_tsv_file, sep='\\t', header=None)\n",
    "        ann_df.columns = [\"Features_name\",\"start..end\",\"length\",\"strand\",\"feat_type\"]\n",
    "        ann_df[['start', 'end']] = ann_df['start..end'].str.split('.',1,expand=True)\n",
    "        ann_df['end'] = ann_df['end'].str.replace('.',\"\")\n",
    "        ann_df[\"start\"] = pd.to_numeric(ann_df[\"start\"], errors='coerce') -1\n",
    "        ann_df[\"end\"] = pd.to_numeric(ann_df[\"end\"], errors='coerce')\n",
    "        CDSs = {row[\"Features_name\"].rstrip():(int(row[\"start\"]),int(row[\"end\"])) for index, row in ann_df.iterrows() if row[\"feat_type\"].rstrip() == \"CDS\"}\n",
    "        forbiden_regions = [ (row[\"start\"],row[\"end\"]) for index, row in ann_df.iterrows() if row[\"feat_type\"].rstrip() != \"gene\" and row[\"feat_type\"].rstrip() != \"CDS\"]\n",
    "        return   forbiden_regions,CDSs,dna_seq.upper()\n",
    "\n",
    "    def get_changing_cds_parts(coor,mforbiden_regions,dna_seq):\n",
    "\n",
    "        \"\"\"\n",
    "        ````\n",
    "        Input:\n",
    "        coor: type=tuple: coordinate (min,max)\n",
    "        mforbiden_regions: type=list of tuple: all forbiden region (to keep wild-type) (this list must be sorted in ascending order without overlaps between coordinates)\n",
    "        dna_seq: type=string (uppercase): dna sequence to recode\n",
    "\n",
    "        output:\n",
    "        cds_parts: type=dict : for all CDSs: {seq1:(start1,end1), seq2:(start2,end2),...}\n",
    "\n",
    "        Description: Append cds part only in dictionary when we are sure a sequence do not contain a forbiden region. \n",
    "        Coor is a cds part position. regions is a forbiden region.\n",
    "        Some exception cases do not required to be considered, because they do not provide infos on cds parts to append.\n",
    "        what conditions based on forbiden region positions mean we know for sure this part could be recoded (and so given as output)? Here what I came with:\n",
    "        ````\n",
    "        \"\"\"\n",
    "\n",
    "        cds_parts = {}\n",
    "        notpickreg = []   #region to exclude when appending a string in cds_parts\n",
    "        for i,regions in enumerate(mforbiden_regions):\n",
    "            #find all changeable part of CDSs, or exclude forbiden sequences:\n",
    "            if regions[0] > coor[0] and regions[0] < coor[1]:\n",
    "                if not notpickreg: #if notpickreg empty\n",
    "                    cds_parts[str(dna_seq[coor[0]:regions[0]])] = (coor[0],regions[0])\n",
    "                    notpickreg = [regions[0],regions[1]]\n",
    "                else:\n",
    "                    cds_parts[str(dna_seq[notpickreg[1]:regions[0]])] = (notpickreg[1],regions[0])\n",
    "                    notpickreg = [regions[0],regions[1]]\n",
    "            elif regions[0] <= coor[0] and regions[1] >= coor[1]:\n",
    "                #the cds is completly overlapped by forbiden region.\n",
    "                return {}\n",
    "            else:\n",
    "                if regions[0] > coor[1]:  #if region[0] exceeds the limit of the cds,if == do not append\n",
    "                    if not notpickreg: #meaning if no previous forbidden region in cds:\n",
    "                        cds_parts[str(dna_seq[coor[0]:coor[1]])] = (coor[0],coor[1])\n",
    "                    else:\n",
    "                        if notpickreg[1] < coor[1]:\n",
    "                            cds_parts[str(dna_seq[notpickreg[1]:coor[1]])] = (notpickreg[1],coor[1])\n",
    "                    break\n",
    "            if regions == forbiden_regions[-1]: #last forbidden region, exceptionnal case:\n",
    "                if not notpickreg:\n",
    "                    cds_parts[str(dna_seq[coor[0]:regions[0]])] = (coor[0],regions[0])\n",
    "                else:\n",
    "                    cds_parts[str(dna_seq[notpickreg[1]:coor[1]])] = (notpickreg[1],coor[1])\n",
    "        return cds_parts\n",
    "\n",
    "    def get_cds_parts(CDSs,merge_forbiden_regions):\n",
    "\n",
    "        \"\"\"\n",
    "        ```\n",
    "        input: \n",
    "        CDSs (for coordinates), merge_forbiden_regions (to split cds at the right place) \n",
    "\n",
    "        output : \n",
    "        type: dict of dict: containing gene_name, sequence and genomic position.\n",
    "        ex: {'pX': {'ATGAAT...': (842, 1087)}, 'pVII': {'ATGGAG...': (1107, 1153), 'ATGA': (1205, 1209)}, \n",
    "        'pIX': {'ATGAGT...': (1205, 1280), 'ATGA': (1300, 1304)}...}\n",
    "\n",
    "        Description: Find and cut at right place the fasta file to collect the cds parts to recode.\n",
    "        ````\n",
    "        \"\"\"\n",
    "        cds_parts = {}\n",
    "        for gene_name, coor in CDSs.items():\n",
    "            gene = dna_seq[coor[0]:coor[1]]\n",
    "            if coor[0] < coor[1]:\n",
    "                cds_parts[gene_name] = get_changing_cds_parts(coor,merge_forbiden_regions,dna_seq=dna_seq)\n",
    "            else:\n",
    "                cds_parts_end = get_changing_cds_parts((coor[0],len(dna_seq)),merge_forbiden_regions,dna_seq=dna_seq)\n",
    "                cds_parts_start = get_changing_cds_parts((0, coor[1]),merge_forbiden_regions,dna_seq=dna_seq)\n",
    "                cds_parts[gene_name] = {**cds_parts_end, **cds_parts_start}\n",
    "        return cds_parts\n",
    "\n",
    "    ##################################\n",
    "    # MAIN:\n",
    "    ##################################\n",
    "\n",
    "\n",
    "    #Open fasta file and annotation file provided with snap gene, creating first object for m13 rewritten genome:\n",
    "    forbiden_regions,CDSs,dna_seq = find_forbidden_regions(path_to_fasta_file = fastafilename, \n",
    "                                                           path_to_tsv_file = path_to_tsv_file)\n",
    "\n",
    "    #check if forbiden_regions has overlaping region between end and start of the file:\n",
    "    for pos in forbiden_regions:\n",
    "        if pos[0] > pos[1]:\n",
    "            print(\"path_to_tsv_file variable contained a forbidden region that overlaps the end and the start of the fasta file. Please split as follow: (x,end) and (0,y).\")\n",
    "            print(\"NO OTHER ACTIONS ARE TAKEN IF U SEE THIS MESSAGE, BUT ANNOTATIONS AND RECODING COULD BE WRONG. PLEASE CHECK IF THIS IS THE CASE.\")\n",
    "\n",
    "    #del kanR gene in CDSs, because we'd like to keep it wild-type for the moment. (hardcoded)\n",
    "    if \"kanR\" in CDSs:\n",
    "        del CDSs[\"kanR\"]\n",
    "\n",
    "    #merge overlaping region in forbiden_regions:\n",
    "    merge_forbiden_regions = merge_contiguous_regions(forbiden_regions)  #merge continuous region, but doesnt care about overlaping region.\n",
    "\n",
    "    #Get cds parts to recode. Cut cds sequences based on merge_forbiden_regions. cds_parts type:dict of dict:\n",
    "    cds_parts = get_cds_parts(CDSs,merge_forbiden_regions)\n",
    "\n",
    "\n",
    "    len_dna_seq = len(str(dna_seq))           \n",
    "    all_parts, recoded_in_frame_cdsparts, total_changed_codons = get_part_in_frame_parts_to_recode(perc,cds_parts,len_dna_seq)\n",
    "\n",
    "    recoded_dna_seq = ''.join(all_parts)    #create the genome by joining every element in the list.\n",
    "\n",
    "    #SECTION ECRITURE:\n",
    "    recoded_dna_seq_titled = f\">recoded_{perc}%_WTZ_m13genome\\n\"+recoded_dna_seq    #Titre\n",
    "    #write the output:\n",
    "    with open(outputdirname, 'w') as f_in:\n",
    "        f_in.write(recoded_dna_seq_titled)\n",
    "        print(\"output file has been created\")\n",
    "\n",
    "\n",
    "recode_son(25,\"../ref_files_m13/m13mp18-Kan_annotated.fa\",\"../ref_files_m13/Features_from_m13mp18-Kan_annotated_nonatg.txt\",\n",
    "           '../ref_files_m13/trial_recoded_25perc_m13mp18-Kan_v2.fa')\n",
    "#commentaires:\n",
    "### gtg pour le codon start: jamais recoder! #has to be annotated ok (just put it as annotation)\n",
    "### gene antisens: attention!\n",
    "###: si juste un cds a recoder, ton code marches tu??\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  True\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que les positions entre la version wt et wtz sont pareilles :  True\n",
      "Est-ce que les sequences en adn sont pareilles?:  False\n",
      "Est-ce que toutes les parties de cds encodent pour la meme prot: True\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAs4AAAG6CAYAAAAYt8BmAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAAAxOAAAMTgF/d4wjAAA8P0lEQVR4nO3dd5hkVZn48e9LFkYGkCzqgBgAIwYUEQczRtaE6Kq4P3d1Dauga0BRRtRllUVMq65pzAkDroJrwAEVBQUjiMQhSYYhSRB4f3+cU07N7aru0z3dXT0z38/z1FPd56b3nrrhrVvnnhuZiSRJkqTxrTXqACRJkqRVgYmzJEmS1MDEWZIkSWpg4ixJkiQ1MHGWJEmSGpg4S5IkSQ1MnCVJkqQGs5I4R8SSiBhZh9ERsTgiMiIW9JUtqGWLRxVXjWOkdTNdIuJeEfGtiLi01uuyUcc0VRGxcUR8MCKWRsRtdX0eNOq4ZsOw/WLQPjQXRMS6EbEoIs6KiFtqjPuMOi5NLCIW1s/rkFHHIkXE1hHx2Yi4KCJur9vmJqOOazqNOu9ZXfb55sS5rmz/65aIuCIiTo2IT0bE3hGx9kwEWROYpTMx75k2VxOO6VQ/928DTwG+CywCDmuY7oUR8YeIuCEifh8Rzx8y3lYRcWVEHD6dcY/jvcBrgD8A/0FZn0tnadmanNcDbwf+AhxO+azOmO0gIuI5EfGhiPhpRFxX9/kvNEy3e0QcExFXR8RNdT943XjH0oh4Wv3CfW3dd06KiJdMsJyXRMTJdfxr6/RPG2f8tSPigBrPTTW+YyJi94nWSVpFLQZeBBwPvItyLLm5ZcKI2DUivlST7lsi4rKIOD4iXjxk/J0j4msRcXlE3BwRf64XAO40zjImfazQzFhnCtMsqu9rA5sAu1A2tv8H/DoiXpiZZ3ameTGw4VSDnAZvoSRyF48whmFGXTfTYXtgZ+ATmfkvLRNExDOALwAnAR8D9ga+HBHXZ+b3OqN/BLgaOHj6Qh7X04AzM/Pps7S8VcFc3YeeBtwAPCEzbx1hHG8DHlhjuQi470QTRMQzgW9QTs5fpWzjTwfeDzwKeO6AaV4NfAi4irL/3Ao8B1gcEffPzDcMmOZwyheMi4BPAOsBzwf+NyJek5kf7owfwFfqfP8MfBjYDNgXOCEinp2ZR0+0ftKqIiLWA54A/CgzXzjJaV8NfAC4Bvge5Ri5GXA/ysWkz3XG3w04DlgXOAq4EHgs5QLA4yLicZl5S2eaSR8rNIMys+kFZBl94LCtgK/VcS4Atmydb+OylwJLp3meC2q8i6dzvgOWs7guZ8FMLmeUL2DPuo6HTGKaY4EzgXXq//MpB55jOuM9B7gDePQsrs8dwJJR1+uIPstZ2S+mMd5zp/vYMMU49gLuBQSwsNbhF8YZf2PgcuAW4KF95RsAJ9bpnz/gs7mZkjQv6CvfFDi7TvPIzjS71/KzgU0787qqzm9BZ5r96jQ/BzboK39Yjfdy4M7TUGe9ejpk1J+frzX7Bdx9Ksc94In1fPF/g/YJYN3O/2sDp9dlPaOvfC1KEp3AmzvTTPpYMU68Iz2+ry77/GRWeGji3PfB/6SOd2Rn2JLutPUE85L6wV9RD+AX1g1w304lD3ot7sS2BNga+CTlG9/twP51+GI6yWv/BkS5OvRtyre4G4GfAU8csI6H1GkWtmyQ48S+dLy66avPVwC/olzFurH+/a/AWkM+nyXA5sD/AJfUHe004KVT2MAfQvmG29thzwf+G9hm0HYx4DXujkE5eHylU3YScFrf/5tRmkh8eKU2ctiGctV6KeUK3RXAN4GHDNpOB7yWNCxjaX1tDBxR//5bfz3U7WwxZTu/FbgM+BJwnyHz3BB4E/Br4Pq6HfwJ+CCw1VTWsW/8O9c4L6Lse2cABwI7dLfjxn1oAeUq5ZV1fr8GnjZk2fOBI1uXPWQevXiG7lt1vOcBJwDXAjdRmt+8BVh/Kp9h4/a2kIkT53+q43x2wLDH1mHHd8rfWcsXtc6PcrUrGXAMGDa/Wl8J7DVgmqHzG2ddtwI+Vbf3m4DfUo79vXo6pDP+klq+DnAQcBblGHQh8J/AegOWsQ/lCvyZlGPljcApwL8x4Hg5TqwTnpc6429HuSJ/bo3xKuA7wMOmqS6WdrfpvmGHMPx81HysoW/fBl5O2UdurtP8DzB/yPK3oxyLzqrrcjVwMnDwytQT5dh0MPBH4DrKse8cypXWgcezIfHdq26vF9c6+Ev9/14D6njcHGOcZfyubmt3aYxp4L5dh/WOf0uBWJljxTjLX8DyY3Zr3rN/nWZ/4KmUfeNGyoWuo7r1OcHyB27nk/m8uts+5eLaycBf67p8BbjrkOU/DPhB3aauA34EPJJx9qVBr6k01RgoM++IiHfVFdkvIg7IGukQ76acwM6jXK2+lnLyfxjlZ4evUjagRcDr6jRH9k3/2878NgN+SUkuvkn5FnhZQ+jbA7+gHCw+XmPYFzg2Il6QmV9tmMcwiygH9AdSfspZVsuXDR59BZ8HXkA58H2S8qH+AyV53QMY9HPSJpSrRLdSNuj1KXX56Yi4IzM/2xJ0bfv4DcpJ5ChK0vwQStL+zIjYIzPP61vHBZSD//GUkx5978NcADwoItaq287GwL0pn0XPBykH5De3xD1kXbanHBC2pfw89mXgbpR6eWr92fm7dfTFNe53UNZ5cS1f2ri49eoyNqPsnNdRtm8i4smU7XJd4H8pVwC3A55V49grM0/ti3tTyhfRB1J+Lv805XO9J/DSOq/LprCORMT6wI8p+9rvgC9Stp2Dgcc0rmu/e1AOXOdSttvez/pHR8TjM/MnfcveoMa4K/Cbuuz5wFuBR09imd+mfC6vq/8fWd+X9S3rPZRjzJWUpOEGSpOg9wBPiogn5tjmHUM/w2n22Pr+/QHDTqCcBHaPiPVz+c+2401zbGecluUcS/nMH0vZ5nufz+51+T8dMs2L6jSfGTB8BRGxOeVEuwNlG/0Z5Rj7MUr9judLlG3iWMrn8BTgjcCWlH2g32GUY/5JlBPv/BrjByjb+YsmirVqOS/11m3Xug6bURLrb1IuXOwD/Cwi/iEzj+kbf2XqotlkjzV93gs8qU7zA8qvKP8M7Ehnu4qIh1LWeTPK9vpNyhf9nSlJyKF94zbXU20m9H3KNvgLyrnvthr/XpRt8pSGOngYJSm6MyVBP52SKP4j5Rz2+Mz8VR39SMo57LWU4+G3a/lvJ1jG/YAH1PGvjoi9KOfJrNP+JDPv6Ew2dH/MzHMj4kzKeXAHypeFcadh+LFiIlPJe55FOX5+i3KefBDwbGCviNg9M//cuOwxJvl59Xsl8Iw6zfHAbnU9HhgRD+qvj4jYk7Idrk3ZBs8B7k85zx43qYAn8U0hGeeKcx1nfcoVmgS27ytf0p2W8o3zImDDAfPZfMA3woHfuvtjo3w7WWfA8MUMv1qWwPs64z+0rsc1wMaDvuWM901uomV3hg+qm95PpacC8/rKN6JcyUvgBUPq4JPA2n3lO1MOPKc3fs7z6mdzO53mEZSrnwn8oFO+kMk31diH5T8Hv49ydSGBp9fhT63/P751nkOW8391Pm/tlO9e6+Wq/jruq8slk1zO0jrdj4CNOsM2rdvSlcDOnWH3oyR0p3bKv1Tn91E6V8zqZzR/qutIuYqXlC9Ha/WVb0/5xt60HbPiPvSOzvhPquXdpjcH1/Ivs+JVlbtRrvA1Xenp1PuYYwPlKkJSvqBt3Ve+DiUxSOCg1s9wkttCb38Y74rzr+o4w34R6O0PO/WV9epn4NWtuh0l9ZhKOV4kcP2Q8Tevwy/rK9ullv1hyDQPrcNPaqyL/6njv3/AfHrnikM6w5bU8lOAzfrKN6Ikgbf3f6Z12D0HLHst4LN1Xrs1xtt0Xqrb0dmUK7OP6Yy3LSV5v4S+XzamWBcDt+867BA65yOmdqxZ3Lev3L2zjr1fHx7eV74e5YvFmPNQHb7dVOuJksgk8K0hn+em3fIB4wXll7kEXtgZtm8tP4MVj30LmPyx56V1mo9TkrbsvH4P7NiZ5ut12LOHzPO7dfjefWWTPlaME/OCvvha8579+6Z5Wmea19byHzfW2cLudj7Fz6u37V8H3L8zTe/c+bzOtnNWt27rsFf0rd/CpvWYxEaSTJA41/EuZeyOtqQ7LeUAdR4DfjIdMM+lTJw438KQttWMf9JfxuC2Sb1pXjLgwxpTuUxv4vzDOs2gn00eV4cdN6AObuzf4PuG9XbqeYNi6Iz7wjrulwYMW4flB8z+A+yYnaFxm3oJpSnJjZRvvv9Yy+dTTl6frP8/u+44t9dt4V8a579djet8Om3N6vDP1+EvHlCXSya5LkvrdA8cMKx3cHnVkGnfX4fvXP/fsq7rX5gggZvKOlIOILczONnobeMTbsd92/xS+r6s9Q0/H7iyU9ZLfMbsD5SrzpM9eS1lcOL8iTqvMdsK5YrO7cC5rZ/hJLeF3v4wXuJ8Zh1nxyHDf16HP7Kv7NZaNubiQB1+cR2+Tf1/2/r/RUPGX7cOv6WvrNcm+mdDprlXHf7nhnpYl7JvX8eAn/v7tqlDOuVLGPKlmfIL15gT+Dgx7FrHf3vj+E3nJeCZDEg++ob39vmnrGRdDNy+67BDGJs4T+pY01n2ywaM30sOX91X9uxadnRDfU62nnqJ85jzT+uLcrNcAicOGf7TOnzPvrIFTP7Y07uQdBvlWPcUSjOve7H8uPtn+poWUa54Dty26/Av1uH79ZVN+lgxTsy99VxGe96zP0OSY8rV2979FfdoWP7C7nY+xc+rt+2/a8D4e9Vhh/eV7cGAvKkOW6t+TivsS+O9pq2pRp+o7znBeF+kdPl1ekR8jZLc/SIzr53icpdm5uVTmO7UzLx+QPkSSmL3YMpVi9m0K/UGtQHDjqec9B88YNhZmXndgPIL6/umlCsOEy0bBvx0kZm3RcQJlJ3vwZQrFFOWpenIoLr9r/r++voz39cpV0dfRfm56OMRcXGO7X2jq1dHP83Mvw0Yfhzlp6AH07nzeYpuplxl6HpkfX9gDO6/8t71fSfKT1QPo+zMJ2TmjRMsc1LrGBF3pvz0emFmnjNg/CXUn+0n4beZefuA8gtZvu7U5jj3rMteOmD8n01yueMZbzs+MyIuAraPiPmdY86wz1CTd1/Kz/c/HXJcX0I5xg7z6wFl/ceyv4uIuwD/TkledqBcne5314Z4of281Nuu7zFkn75Xfd8JOIaVr4tWkz3W9Gut70fU92OZ2GTr6XRKM4f9IuIewNGU48Kvs73XnKH7fl/5HpRj4gmN8xyk153v2pSb83rNDK+r3dDdl3IV99mUX9jmkqnkPcd3R87M2yPiZ5Tj+oMpXyAma2U+r9ZttneeHHOOydJU9ESW7xsTmtbEubaP26z+e8UEox9AaRP5Ukob1jcDt0XEMcDrM/PsSS5+qv3sDmsH3Zvf/CnOd2XMB64edKCoyeuVlKuSXcuGzO+2+t7S32NvfS8ZMrxXvknDvCYtIh5P6drw6Zl5bUS8ntKQf//MvDEijqPcyfwmStc/45ntdbk861fYjrvU93+eYPp5nXhaun6b7Dr2xp9ou5+MZUPKb2PFvuI3nmDZLfcktGqpl7tT6qU/kRn2GU633jKHHV965cs602xeh101zjTXdt4nu4zJTjPMSm1rmTloGWOOZVEeUvErSlOjkylfgq+u425Cuaq5fkO80H5e6u3TE3UD1tunZ2K/G2Syx5p+ywaUDTp3bFLfW45Pk6qnmoj1umZ7DuVmUIDrI+KzwFsyc6KLP7N13F9W3y/tS5qB8hNyRBxNSZwfzvLEedT7ZM9U8p6ZypVW5vNaNqBs0DY70f43qXPPdD85cA9KMn7ZkCtKf5eZt2fmkZn5QMqdxs+mNDp/BvD9egPTZEz1ZLfVkPKt63v/SbXX0H/QF45Nprj8Qa4FNouIdbsDImIdyslz0JXl6Vo2LF//rm06402biJhH+Yn9i7n8ZradKD8L3wi1TUu5qWyXhlnO9roM2wZ7839gZsY4r943/GX1veUq2WTXsfc+0XY/E3rb7LBlDyufiql+9rORNEP5aRAGXOWo+/j2lBPAuY3TbEO5ynpRZv4VoO4zFwPz6vCu3tW+/n73z6H8orVDjaNlmmFma1t7GaW+FmXmbpn5ysx8W2YeQt/NfC0mcV7qrdszJ9inF3XGn2xd3MHwC1ybDCib7LFmKpbV98kcn1rricy8JjMPyMy7Uba3l1Ga6r2acs9H6zJn+rjf2x+XDRl+TX3vf6jJ0H24GrR/TeVYMZHJ5D0rM02L2fi8pvXcM22Jc0SsRWmjCKVxdrPMvDwzv5mZz6Nclr8n5UaGnttpu1o6FbvWn6+7Ftb33/SV9XaEuw0Y/6FD5t/7CXsy8f+G8tnsOWDYnnVeg+6Kng699V3YHVB30l7PBzOx/MMoB5nXdsq7X6I2aJxfb132GJIE7FXfZ6oue35Z31t7jTiZcsLcMyK6Pzl3TWod689zZwN3jYh7Dhh/YWOMk1abEZ1bl71gwCh7TOPixtuOd6S0DT9vyFXN2dD7WfLJA4btSflZ/8Rc8S758abZuzPOlKbJzJspPT9syODtddhyBjmDcsf/gyJi0NWohQ3zaLFjff/GgGGPmepMJzgvTXafnmpdXANsNegiCoPPOZONayp6y9h73LFWHHdK8WTm2Zn5KcrneAOlzfREhu771XQd939Jabe+YMhxuretnNdXNnR/jIgdKMnx+ayYBE/lWDGRyeQ9PWP2pfrUwt5xe9A0LWbj8/r7ebI7oOauk3siaktD6PrLZTLk5kBKs4HeA1DOZ2yvGEv6p6UkQo8aMJ916womK95NfjKl7eGdxoltyTixL2b4jU1jblxg+d2ly1jx7tLd6vgn0HeDDiWRvoDBN1W9t5aP6RN1UN3UshfUaU6m7+5uyg5yMoPvPh1aB4PWf5y66vWqcRvwiM6wN9T5/LBTvpAp3BzYmcejKcniczrlX6ix7FD/n085mZzQON/ezRhv6JTvVud7NZ2bJCbanoYsZynDb+K5S435cvpumu0bvhadmxJYfpNIS68ak1pHlveqcRTT06vG4iHrPWjbfkedZqZ71ejd5HYesEVfee/x8MnYXkiGfoaT3BZ6+8NED0C5gsk9AGV75s4DUMbchDxkPafcq8aQ+e1fp9m/r+zNtew1nXEfTLlK1bRNMYnzUi07m5IMP2XI/B7JisfvqdTFRxlwkysr9nSwsK98KseaxQw5PzD4Zq7+XjX2GzBNf68ak6qnuo3vMGCcban9UTd8jkH5opKMPZ88h+U37a1Urxp1ug/0PlNWPJ7dn9KV6t/ouwmb8R+A0utxY9ADUCZ1rBgn3t56Ju15T/+2NqxXjTE33Q1Z/qDtaSqf1yHdbX+8z7LWb+8YOfu9atSAD6F0nv9BSn+wt9RhJzHgzk/GJs6b1PHPonRW/Z91A+xtUEd3pv+PWn48pX/It1G7LeuLbck4sS9m+En/eGoiVpezmLLB387gDu97PVScChxOuXt2GeUnwTE7Hsu75Tq7rufbWPEO5RXqpq+8N7/zKDvlEZRvoUnnwSET1cGg9Z/gs34m5SB1CyVxfQ/Luzy7hM6BjZVMnClXmc8Ejhow7EEs703jCErvG8mQg/CA6XeoMSclwXxP/cxuruv4zMnU5TjLWcr4Pb88jvJz0R2UXlOOrJ/rUZSf02/ujL8ppU/RpHTVcyTlS9jX63wWTnUdKQlC7wvYb+t2+XHKfnD0kO14zDbE1BLnO7E8CTmV8ivDRylJ3Ldq+aeno97reiWl/dpHav31tp+f0nmQxkSf4QRx7FPraDGlv9WkNHvolR0+ZJrbKFfSPlnj651Avk7fibhvmtfU4VfWdXo/5WaYHLSMOs1/1eEX1vE/UqdP+o5FfeMHy0/if6pxfarGeVt3e5qgXjav9dCr8/5jbG9bO6QzzZjtpm/Y/oxNnLdleRea36yf+zcp2/5XxttGO/PehMmdlx7A8v3u57Ve31en7a3z1itZFztT9uPbKeeEwyndJd7I8m4VF3ammeyxZjGTSJxr+UNZ/iV7CWU/PpJynrhtqvVE2Sd6/XEvphzLPkH5IpDAgY3b3W61DnrbxHsov0jcXst364y/oHU76Uy3McuPZ7+k7GtfoHxRSOC1Q2K7kbJ9fqnWXa/LuZ8x+OFM+zDJY8WQeHvr2Zz3sHyf+w4lsf5arc9javlVwH0blz9se5rs53UIk0ic+5Z9c12Hr1D6bP/fWtZblz2b1mMSG0h2XrdQDr6n1A37yQx5QhNjE+d1KR3ZH0u5Unsz5RvVLynZf/eEthHl5HpR3XhWqBRWLnFeTGlHe3TdkP5K2bmfNGRem7B8R76F0ofivwz7sOo0B1JOQL0vGEuH1U1f+VqUzr1/XWP6a63rVw2q5/HqYND6N3zeD6MkMldQdvAL6mewbevOMIllHU7Z+bYaMnyfWs+3Ug60Y7pNmmD+d62xn1/ncSXlquOwp3uNuz0NmWYpEyRddRv5MOXEfDPlgHAGJcndZ8D4G1GaP/2+fv7XU07iR9LpenEK69h7Ot7FLH963+uZ4pMDhyxj2La9CeVL91/qPtFb9sPr/I6crnoHnk85GV1f1/O0WqcbTHZeE8RxCGOPkf2vgfOldMV0DOXY03uy4QEM6N6vb5qnU05817P8iaIvmSC+/et4N9bpjmecLt0o7WoPqPHcVOM7Bth9CnWzNeUBPlew/Gl5+zPBkwPHWY8VEudavjPlxH45y58a+LKJttHOPCZ1XqrTbElJfP5I2UdvoOzfR1F6s1lnZeqiTrMHyx90cR3lpugHMHHy0HSsYQqJcx12d8oDuc6jHHOuoiS8B021nihNqN5DOQdfSjk+XFQ/k727853g87xPXd9LKMnSJZSkdtDTE5u3kwHTzqMkYWfWeJdRLmCM6U62s71+nXKcvqVOu4ghv6pP9Vgx3nrSmPew4pMDn0Z5cMqNdT2/Adx7Essfb3uazOc10bY/LBfbjfJl8vr66j058MN1mge1rEfUmUnSSEXEP1N+zn5FZn581PFIsyUiFlKeYLYoy02N0pwQEftTnhL60sxcPNpoZkZE/JySVM/PibuAnfZeNSRpXBGx7YCyu1OeKngb5eczSZKmRURsWLuu7JbvT7kX5ActSTNMcz/OktTgG7WXgFMoP/ctoPwEuCGln9a/jC40SdJq6O7AbyLih5R7ztah3EC8B+U89PrWGZk4S5ptnwdeROkjdz6lveNJwIcz85ujDEyStFq6jNJb1WMoXdytT2lH/xng3Tn4SboD2cZZkiRJamAbZ0mSJKmBTTVWYeuvv35uscUWow5DkiRNwsUXX3xrZnafiqtVgInzKmyLLbbgoosuGnUYkiRpEiLiilHHoKmxqYYkSZLUwMRZkiRJamDiLEmSJDUwcZYkSZIamDhLkiRJDUycJUmSpAYmzpIkSVIDE2dJkiSpgYmzJEmS1MDEWZIkSWpg4jwDImLHiPhYRJwaEX+LiKVDxntyHefmiDgvIg6Y5VAlSZLUyMR5ZuwCPA1YCvxx0AgRsRvwHeD3wN7Ax4H3RcRrZilGSZIkTUJk5qhjWO1ExFqZeUf9+2PAkzNzQWecY4AtgYdl/RAi4oPAC4GtM/NvEy1nu+22y4suumi6w5ckSTMoIi7OzO1GHYcmzyvOM6CXNA8TEesDjwO+mit+c/kisBmw+wyGJ0mSpCkwcR6NewLrAX/qlJ9e33ea3XAkSZI0ERPn0di0vi/rL8zM64HbKVedJUmSNIesM+oA1C4iDgQO7P0/f/78EUYjaSYdHjHqECbtDd4zI2k15xXn0bimvm/SXxgRdwbWBq4eNFFmHpGZ2/Ve8+bNm9koJUmS9HcmzqNxDnArY9sy71zfu22fJUmSNGImziOQmbcAxwHPi1jh99j9KFejfzGSwCRJkjSUbZxnQERsCDyl/rsDsGFEPKf+/6vMPB94J3AC8KmI+BzwcOBVwL9n5q2zHbMkSZLGZ+I8M7YEvt4p6/3/UmBxZv4iIp4BvIfy0JNLgDdn5pGzFqUkSZKamTjPgMxcCkx4S3xmHgscO+MBSZIkaaXZxlmSJElqYOIsSZIkNTBxliRJkhqYOEuSJEkNTJwlSZKkBibOkiRJUgMTZ0mSJKmBibMkSZLUwMRZkiRJamDiLEmSJDUwcZYkSZIamDhLkiRJDUycJUmSpAYmzpIkSVIDE2dJkiSpgYmzJEmS1MDEWZIkSWpg4ixJkiQ1MHGWJEmSGpg4S5IkSQ1MnCVJkqQGJs6SJElSAxNnSZIkqYGJsyRJktTAxFmSJElqYOIsSZIkNTBxliRJkhqYOEuSJEkNTJwlSZKkBibOkiRJUgMTZ0mSJKmBibMkSZLUwMRZkiRJamDiLEmSJDUwcZYkSZIamDhLkiRJDUycJUmSpAYmzpIkSVIDE2dJkiSpgYmzJEmS1MDEWZIkSWpg4ixJkiQ1MHGWJEmSGpg4S5IkSQ1MnCVJkqQGJs6SJElSAxNnSZIkqYGJsyRJktTAxFmSJElqYOIsSZIkNTBxliRJkhqYOEuSJEkNTJwlSZKkBibOkiRJUgMTZ0mSJKmBibMkSZLUwMR5xCLimRHxy4i4LiIuj4jvRcSDRh2XJEmSVmTiPEIR8QTgW8BZwHOAlwNbAj+OiG1GGZskSZJWtM6oA1jDvQA4H3hxZiZARPwOOAd4ErB4dKFJkiSpn1ecR2td4IZe0lxdW9/9bCRJkuYQk7PR+jRw34g4ICI2jYi7AR8CLgS+OdrQJEmS1M/EeYQy8zjgWcAi4GrgAuARwOMzc9kIQ5MkSVKHifMIRcQjgM8BXwIeD+wDXAIcGxFbDRj/wIi4qPe64YYbZjVeSZKkNZmJ82h9CDgxM1+RmT/OzKOBvYHNgAO6I2fmEZm5Xe81b9682Y5XkiRpjWXiPFq7AKf0F2TmdcDZwL1GEpEkSZIGMnEeraXAQ/sLImJjYEfgvFEEJEmSpMFMnEfrw8DeEfHxiHhiRDwLOBZYH/jEaEOTJElSPx+AMlofBW4GXk15GMpNwKnAXpn551EGJkmSpBWZOI9QffDJp+tLkiRJc5hNNSRJkqQGJs6SJElSAxNnSZIkqYGJsyRJktTAxFmSJElqYOIsSZIkNTBxliRJkhqYOEuSJEkNTJwlSZKkBibOkiRJUgMTZ0mSJKmBibMkSZLUwMRZkiRJamDiLEmSJDUwcZYkSZIamDhLkiRJDUycJUmSpAYmzpIkSVIDE2dJkiSpgYmzJEmS1MDEWZIkSWpg4ixJkiQ1MHGWJEmSGpg4S5IkSQ1MnCVJkqQGJs6SJElSAxNnSZIkqYGJsyRJktTAxFmSJElqYOIsSZIkNTBxliRJkhqYOEuSJEkNTJwlSZKkBibOkiRJUgMTZ0mSJKmBibMkSZLUwMRZkiRJamDiLEmSJDUwcZYkSZIamDhLkiRJDUycJUmSpAYmzpIkSVIDE2dJkiSpgYmzJEmS1MDEWZIkSWpg4ixJkiQ1MHGWJEmSGpg4S5IkSQ1MnCVJkqQGJs6SJElSAxPnjoh4eUTcedRxSJIkaW4xcR7rA8BfIuLTEbH7qIORJEnS3GDiPNZdgUOA3YCfRsRpEXFARNxltGFJkiRplEycOzLzqsz8r8zcBdgT+BVwKHBxRHw1Ip4w2gglSZI0CibO48jMn2fm/sCOwMnAc4HvR8S5EfFvEbH2SAOUJEnSrDFxHkdE7BoR/w38CdgV+AzwTOAHwHuBT40wPEmSJM0iE+eOiJgfEa+KiFMpzTQeDRwM3DUz/19mfjczXwG8HHj2NC1z34g4OSL+GhFXR8SPI2Kb6Zi3JEmSpsc6ow5gDrqkvn8deHVmnjhkvN8BV67swiLi9cB/AIcDbwI2orSt3mBl5y1JkqTpY+I81luAz2XmNeONlJm/BbZfmQVFxI6UpPm1mfnRvkHfXZn5SpIkafrZVGOs+cCdBg2IiG0i4u3TuKx/Am7FttKSJElznonzWO8AthsybNs6fLrsDpwBvCQizo+I2yLidxGx9zQuQ5IkSdPAxHmsAHLIsG2BZdO4rK2B+wCLgLcBTwEuAL4TEbtM43IkSZK0kmzjDETEC4EX1n8TODIiru2MtgGlS7qfTuOi1wLmAftm5jE1liXAOZQbBV/cifNA4MDe//Pnz5/GUCRJkjQerzgX61DaNd+JcsV5g77/e6+/AYsp3dBNl94NiD/pFWTmrcDPgTFXnDPziMzcrveaN2/eNIYiSZKk8XjFGcjMzwKfBYiInwCvzMw/zcKiTwMePqC8l7xLkiRpjvCKc0dm7jVLSTPAd+r743oFEbE+sAfw61mKQZIkSQ284gxExBspfTdfWv8eT2bm+6Zp0d8BfgF8MiLeAlwKvAbYlPJIb0mSJM0RJs7FYcASSuJ62ATjJjAtiXNm3hERT6M8NfC/KG2pfwU8NjNPm45lSJIkaXqYOAOZudagv2dp2VdTHoTyT7O5XEmSJE2ObZwlSZKkBl5xBiJiy8mMn5mXz1QskiRJmptMnItLGf60wEHWnqlAJEmSNDeZOBf/xOQSZ0mSJK1hTJyBzFw86hgkSZI0t3lzoCRJktTAK85ARBwDvC4zz6x/jycz86mzEZckSZLmDhPn4k4sv/q+IbZ3liRJUoeJM5CZe/X9vXCEoUiSJGmOso2zJEmS1MArzgNExCbAvwEPB7YBLgF+CXw4M5eNLjJJkiSNilecOyLiIcA5wEGUts9/ru9vBc6OiF1HGJ4kSZJGxCvOY30YuADYOzMv7RVGxDbAsXX47iOKTZIkSSPiFeexHgQs6k+aATLzEmAR8OBRBCVJkqTRMnEe60Jg3SHD1gUumsVYJEmSNEeYOI+1CHhXRNy3v7D+fyjwjpFEJUmSpJGyjTN/f3Jgv3nAHyPiT8DlwJbATsClwD8CX5rdCCVJkjRqJs5F92mBZ9YXlKvyVwI/rf/faRbjkiRJ0hxh4oxPC5QkSdLEbOMsSZIkNfCK8wARsTbwBODewAadwZmZ75v9qCRJkjRKJs4d9UEnxwM7Uto9Rx3U3wbaxFmSJGkNY1ONsQ4HrgLuTkmadwN2AN4JnA3ca3ShSZIkaVS84jzWnsABwF/q/5GZS4FFEbEW8CHgqSOKTZIkSSPiFeexNgMuy8w7gOuBzfuGnQA8eiRRSZIkaaRMnMc6H9ii/n0m8Oy+YY+lJNOSJElaw9hUY6wfUnrU+CZwBPDFiHgo8DfgwcB/jjA2SZIkjYiJ81hvoj4dMDO/HBF/BZ5LebrgJ+pLkiRJaxgT547MvBm4ue//o4GjRxeRJEmS5gIT5yEiYgHwUGAbSg8bp9TeNSRJkrQGMnHuiIgNgf8GXgis3Tfo9oj4AvDKzLxpJMFJkiRpZOxVY6wjgecBbwTuQXnk9j0obZ/3rcMlSZK0hvGK81jPAw7KzCP7yi4EjogIgIOBl48gLkmSJI2QV5zHSuD0IcP+WIdLkiRpDWPiPNbXgRcNGfZi4KhZjEWSJElzhE01gIh4Xt+/JwKHRsSJwDeAy4EtgecAdwXePvsRSpIkadRMnIuvDCi7K/CIAeWfAhbPaDSSJEmac0yci+1HHYAkSZLmNhNnIDPPH3UMkiRJmttMnAeI0u/cU4BHA5sC1wAnAMdmpr1qSJIkrYFMnDsiYhPgGEr75ptYfnPgvwO/jIinZOa1o4tQkiRJo2B3dGO9F9gJeFZmbpSZ22fmRsCzavl7RxqdJEmSRsLEeax9gDdn5rf7CzPzaOAtdbgkSZLWMCbOY90ZGHaz4PnAxrMYiyRJkuYIE+exTgf+aciwlwKnzWIskiRJmiO8OXCsQ4FvRMTJlMdvXwpsBTwXeAilrbMkSZLWMCbOHZn57Yj4B+AQ4DAggAR+C/xDZv7v6KKTJEnSqJg494mI9YD9gZ9m5q4RsRGwCbAsM28cZWySJEkaLds498nMW4EPUPptJjNvzMyLTZolSZJk4jzWacAOow5CkiRJc4uJ81hvAt4aEY8adSCSJEmaO2zjPNaHgc2AEyLiWuAyys2BPZmZu4wkMkmSJI2MifNYvxx1AJIkSZp7TJyriNgZeAWwOfAX4KjM/OFoo5IkSdJcYeIMRMQewI+AdYErgLsAL4uIV2Xmx0YanCRJkuYEbw4sFgF/AhZk5taUxPlbwLtHGpUkSZLmDBPn4v7AOzPzQoDMvA54PbBpRNxtNgKIiHUi4vcRkRHx/NlYpiRJktqZOBebAxd3yi7qGzYbXgtsMUvLkiRJ0iSZOC+XE48yMyJiO+AdlD6kJUmSNAd5c+ByJ0TEoOT5xE55ZuZG07zsI4HvACdM83wlSZI0TUyci0WjWnBEPBl4InAfYP1RxSFJkqTxmTgDmTmSxDkiNqA8qXBRZl4SEQtGEYckSZImZhvn0ToIuBX4YMvIEXFgRFzUe91www0zG50kSZL+zsR5RCLiHsAbgYOBjSJiE2DjOnjDiJjfnSYzj8jM7XqvefPmzV7AkiRJazgT59HZntKm+Sjgmvr6XR32KcZ2jydJkqQRso3z6PwW2KtTtjXwZeBQ4IezHZAkSZKGM3EekcxcBizpL+u7OfD0zPzpLIckSZKkcdhUQ5IkSWrgFec5JDOXAjHqOCRJkjSWV5wlSZKkBibOkiRJUgMTZ0mSJKmBibMkSZLUwMRZkiRJamDiLEmSJDUwcZYkSZIamDhLkiRJDUycJUmSpAYmzpIkSVIDE2dJkiSpgYmzJEmS1MDEWZIkSWpg4ixJkiQ1MHGWJEmSGpg4S5IkSQ1MnCVJkqQGJs6SJElSAxNnSZIkqYGJsyRJktTAxFmSJElqYOIsSZIkNTBxliRJkhqYOEuSJEkNTJwlSZKkBibOkiRJUgMTZ0mSJKmBibMkSZLUwMRZkiRJamDiLEmSJDUwcZYkSZIamDhLkiRJDUycJUmSpAYmzpIkSVIDE2dJkiSpgYmzJEmS1MDEWZIkSWpg4ixJkiQ1MHGWJEmSGpg4S5IkSQ1MnCVJkqQGJs6SJElSAxNnSZIkqYGJsyRJktTAxFmSJElqYOIsSZIkNTBxliRJkhqYOEuSJEkNTJwlSZKkBibOkiRJUgMTZ0mSJKmBibMkSZLUwMRZkiRJamDiLEmSJDUwcZYkSZIamDhLkiRJDUycRyginhMR34qICyLirxFxWkS8PiLWHXVskiRJWtE6ow5gDfcGYCnwRuAyYHfgXcADgJeMLixJkiR1mTiP1tMz84q+/38SEQEcGhFvzMzLRhWYJEmSVmRTjRHqJM09p9T3bWczFkmSJI3PxHnu2RO4FThn1IFIkiRpORPnOSQidgZeC/xPZl436ngkSZK0nInzHBERmwPfBs4G3jxknAMj4qLe64YbbpjNECVJktZoJs5zQETcGTgWWA94cmbeOGi8zDwiM7frvebNmzercUqSJK3J7FVjxCJifeBoYAGwR2b+ZbQRSZIkaRAT5xGKiLWBrwAPAx6bmX8ecUiSJEkawsR5tD4C7AMcDKwdEY/oG3a6NwhKkiTNHSbOo/Xk+n5offXbC1gyq9FIkiRpKBPnEcrMBaOOQZIkSW3sVUOSJElqYOIsSZIkNTBxliRJkhqYOEuSJEkNTJwlSZKkBibOkiRJUgMTZ0mSJKmBibMkSZLUwMRZkiRJamDiLEmSJDUwcZYkSZIamDhLkiRJDUycJUmSpAYmzpIkSVIDE2dJkiSpgYmzJEmS1MDEWZIkSWpg4ixJkiQ1MHGWJEmSGpg4S5IkSQ1MnCVJkqQGJs6SJElSAxNnSZIkqYGJsyRJktTAxFmSJElqYOIsSZIkNTBxliRJkhqYOEuSJEkNTJwlSZKkBibOkiRJUgMTZ0mSJKmBibMkSZLUwMRZkiRJamDiLEmSJDUwcZYkSZIamDhLkiRJDUycJUmSpAYmzpIkSVIDE2dJkiSpgYmzJEmS1MDEWZIkSWpg4ixJkiQ1MHGWJEmSGpg4S5IkSQ1MnCVJkqQGJs6SJElSAxNnSZIkqYGJsyRJktTAxFmSJElqYOIsSZIkNTBxliRJkhqYOEuSJEkNTJwlSZKkBibOkiRJUgMTZ0mSJKmBibMkSZLUwMR5xCJix4g4JiJuiIgrI+K/I2KjUcclSZKkFa0z6gDWZBExHzgO+AvwXGAz4AhgK+DZIwxNkiRJHSbOo/VyYAvgoZl5OUBE3AR8IyIekpmnjDQ6SZIk/Z1NNUbrKcBxvaS5+g5wA/C00YQkSZKkQUycR2tn4E/9BZl5G3AmsNNIIpIkSdJAJs6jtSmwbED5NZT2zpIkSZojbOO8ComIA4ED+4puj4hLRxXPHDSP0sxFM8t6nh2rXD3/e8SoQ5iKVa6eV1HW84q2GHUAmhoT59G6BthkQPmmwFndwsw8gtLrhgaIiIsyc7tRx7G6s55nh/U8O6zn2WE9a3VhU43R+hOdtswRsTZwbzptnyVJkjRaJs6jdQywV0T0/2TzdMpPWt8bTUiSJEkaxMR5tD4OXA0cHRF7R8QLgI8BR2fmr0cb2irJZiyzw3qeHdbz7LCeZ4f1rNVCZOaoY1ijRcS9gQ8CjwZuBr4OvCEzvYlCkiRpDjFxliRJkhrYVEOSJElqYOKsOSMinhMR34qICyLirxFxWkS8PiLWHTL+goi4MSIyIrZuXMZWEfGJiLg0Im6OiLMi4nXTuiJz3EzXc0RsHhEfjYjz6/zPiIi3RMQa1f1lSz1HxCG1XruvDzcu4x8j4vS6LZ8eES+cuTWam2ayniNi44h4R0T8IiKujoirIuLHEbHHzK/Z3DIb23PffB4REXdExM3TvybSylmjTmSa894ALAXeCFwG7A68C3gA8JIB438IuA7YsGXmEbElcCKl/+zX1mXsCNx5JeNe1cxoPQPfotTr24BzgT2AQ4GNgbesRNyrmtZ6vhV4TGfaCR9sFBHPAj4PvA84FngK8IWIuD4zv7Oywa9CZrKe7w68Avg0cAiwdv1/SUQ8OTN/tJKxr0pmdHvuqV2yfrQuY9OphyvNkMz05WtOvIAtBpS9DUhgq075PsAVwAF1+NYN8/8McA6w0ajXdXWtZ+Aedbz9O+WfA5aOet3nWj1TkrGbpzj/04Fvdsq+A/xh1Ou+utQzsBGwYadsHeAM4JhRr/vqUs+deb6O8hyDd6/svHz5momXTTU0Z2TmFQOKT6nv2/YKImIj4AOUq5fXtMw7Iu4M7Ad8OjNvXMlQV2kzWc9A72fb6zvl17KGNQ1rreepiIgFlIcnfbkz6IvA/SLiHisz/1XJTNZzZt6YmX/tlN0G/A6468rMe1Uzk/XcExHbAouA1wB/m455StNtjTqRaZW0J+Wnv3P6yt5O+RnvU5OYz0OA9YErI+I7EXFLRFxT2zvPm75wV1nTUs+ZeTbwE+BtEfGAiLhzROwNvBg4cvrCXWUNquf1IuKyiLgtIs6OiDfXn6vHs3N97z5h9PT6vhNrtumq5zFqm95H4tNdYfrr+f3AD3PNagKjVYxtnDVnRcTOlLbI/5OZ13XKHp2ZGRGts+vd1HY4cBSlPeh9gMMoT2rcbxpDX6VMcz1DefrlVylX5Xrek5lr9AMQBtUzcDbwZuA3lAsZT6f8RL0D8C/jzK7X9nNZp7z3y8Bm0xDyKmma63mQt1KuNrs9T2M9R8QTgKfhlz7NcSbOmpMiYnPg2yw/EPd8FPh8Zv5qkrPs/bpyema+tP7945oQfiQi3pqZ565EyKukGahngE8C96PcMHQ+8Cjg4IhYlpnvW7mIV03D6jkzv9AZ9f8i4lrgLRFx2Jq4Ta6Mma7niPgH4GBgUWaePD1Rr3qmu54jYn3gI5Qv2BfMTNTS9LCphuac2h75WGA94Mm9NskRsS/wYOA/I2KTiNiE5T09bFzb5A7TuxL3k075cfV9l+mIfVUyE/UcEU8Fng88NzM/l5nHZ+Z7gP8ADo2I+TO3RnPTsHoex1eBoDQvGqa3PW/SKe9dib56kmGu8maonvvnv5DSpvxTmfnOqUe6apuhen4dpSndJ/uOORvU5W0SEXda2bil6eIVZ80p9crD0cACYI/M/Evf4J0oXcedNWDSPwPfo/zUN8hpwxZZ3zeYdLCrsBms510od9n/plN+CuXEeHfgD1MOfBUzQT1PZLzHuvba1+4E/LGvfFjb59XaDNZzb/4PofRYcizwr1OJcXUwg/W8E+XYMKjbumsoV6NfPYllSTPGxFlzRr2B5CvAw4DHZuafO6MsBpZ0yp4MvAl4NoMTPQAy84KI+B3wuM6gx1EO6KeMnWr1NJP1TOnnNYBdgf6fsh/SN3yN0FDPw+xH2SZ/PWyEzDwvIs4A9gW+3pn2tMw8f2pRr3pmsp7r/O8LfB84FdgvM29fiXBXWTNcz4dRjjv99gdeADwRuHgysUozycRZc8lHKP0GHwysHRGP6Bt2emYupZN41W65AE7MzEv7yn8M3CMzd+wb/SDguxHxecqDI+4DvAf4whrWlnQm6/kY4DzgaxGxiNLGeXdKl3afycxuN3Wrs4nq+bqIOJXSx/UZlIdrPB34Z+Bj9XMAICI+BbwkM/uP2W+n1PNhlMRub+CZwLNmbI3mphmr5/rQpB9QEr/3AA/qu1H2lszs/rKyOpuxes7MM+o09I2zELgjM5fMzOpIUzTqjqR9+eq9KMlaDnktHDLN/gx4MAfliunSAeM/i9Lbwy2UqxiHAeuNet1Xp3qm/Iz7BeAC4K+UE+I7gDuNet3nWj1T2n+eW+vpJuC3lD5s1+rMa3E5XI9Zxotr/d5S31806vVeneoZWDjOvJeOet1Xl3oesrxD8AEovubgKzInbN4lSZIkrfHsVUOSJElqYOIsSZIkNTBxliRJkhqYOEuSJEkNTJwlSZKkBibOkiRJUgMTZ0mrlYi4Z0QcHxHXR8RPImL7zvDNIuLy+hjlqS7jSRHx+4i4OSJWyT49I+L7EbGk7//9IyIjYusRhiVJc5qJs6TVzWeBmymPB7+t/t/vUODozFyZx6x/AbiI8jjgR67EfOaS71HW5apRByJJc5WP3Ja02oiIjYBHAQ/PzF9FxDLgpIjYMDP/GhEPAPYFdlqJZWwObA4clZknTHLa9TLz1qkueyZl5hXAFaOOQ5LmMq84S1qdrF/fb+q898o/CBxak8QxImKtiDgoIs6NiFvr+0ERsVYdvj/Lk8tP1aYNi4cFU4e/NSIOjYiLgZsjYoMoXhcRf4qIWyLiwohYFBFrd6a/a0QsjohLa7OQsyLira3x9o33iIg4uW8eLxwQ6wpNNSJiQf3/xRFxRERcWZu4fCIiNuxMe9+IOC4iboqICyLitRHxsYhYOqxuJGlV5BVnSauNzLw6Is4BXlkTzFcBZ2XmNRGxL+VK8UfGmcX7gNcB/wksAfYC3gVsBryB0pzhqfX9XfV9oqu0rwJ+A7wcWBf4G/BfwCuBw4CfAQ8E3kk5Jr8V/n5l+xfA2sAhwJnADnXc1niJiK2AHwJ/BvYDNqjL2qjOcyKHAscCLwDuD/xHXeeD6vw3AH5AaR7z0vp+ELANcHvD/CVp1ZGZvnz58rXavIDHAdcACVwNLAQ2BC4AHjfOdJsDtwJHdMo/ANwC3KX+v3Wd9/4NsSQlOV27r2wHSkL5ms64rwNuBDar/7+7xnOvlYz3sDrfzfvG2QW4A1jSV7Z/jXfr+v+C+v+3O/P/CuXLSO//V9R53buvbLO6zKWj3h58+fLlazpfNtWQtFrJzB9TktudgG0ycwnwFuBXmfnjiNg7Ik6rTQ8W13bRALtRrgh/tTPLLwPr1eFTcUxm9l95fRylmdzXImKd3gv4ESXBv1/feMdn5llD5tsa726UBPnK3giZeRpwRmP8/9f5/3Tgbp04Ts/Mv1+9zsyrKVfAJWm1YuIsabWTmbdk5hmZeUvtju5VwOsjYgvga8AiYEdKcv22Otmm9f3Szuwu7QyfrMs6/2/ZN9+/9b3+UMvvXt/vAlw8znxb490GuLwhrmGu6fx/K8vbjPfmP6i5yqBlStIqzTbOklZ3RwAfysylEfEM4JbM/BpARHye0kThLSxPELcCzu+bvtevcTeBbNXt5/nq+v5oSnvgrvPq+5XAXceZb2u8l7A8We+3FdOT3F4CPGxA+aBlStIqzSvOklZbEfF4YFdKO9+e9fp6r9gIiPr3LylXfp/Xmc2+lKusJ01TWD+iJNNbZeavB7yu6hvvMRFxzyHzaY33JGBhRNylN0JE7ALcd5rW5yRg54i4d9/8N6O0LZek1YpXnCWtlmq74Q8Cb8jMXrd0J1F6qXhvRBxHacLxFYDMvCoiPggcEBE3ASdQkr9/o9yANy0PBsnMsyLiCOAzEXF/4ERK8n5P4JnAMzLzFuD9wIuA4yPiUMpNhvcAds3Mf5tEvO+v6/l/EfEuSq8ahzK2icdULab0ovHdiDiYcmPiQZQr63dM0zIkaU4wcZa0unoNcGlmfr1XkJmXRcR+wOHAyyjdyR3aN80bKQnfPwNvojwd8GBWvGK90jLzDRFxJvCvwJspTTbOBr5LuYpMlq71HkXp/u3dwDxKzyCfnky8dZ2fAHyIciPhBcA7gBdTkuiVXZebI+KJwEcpT2m8AjgSeAgr8aAZSZqLIrPb/E6SpKmLiPUpV8iPy8yXjjoeSZouXnGWJK2UiHgj5WbGcyn9S78S2JZylVuSVhsmzpKklfU3SpOT7er/pwJ7Z+apowtJkqafTTUkSZKkBnZHJ0mSJDUwcZYkSZIamDhLkiRJDUycJUmSpAYmzpIkSVIDE2dJkiSpgYmzJEmS1OD/A+ma8GvlgTv2AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "####################################################################################################################\n",
    "#SECTION CONFIRMATIONS:\n",
    "####################################################################################################################\n",
    "def check_if_same_prot(seq1, seq2):\n",
    "    #check if two cds parts are in frame.\n",
    "    #print(seq1)\n",
    "    #print(Seq.Seq(seq1).translate(table=11))\n",
    "    #print(seq2)\n",
    "    #print(Seq.Seq(seq2).translate(table=11))\n",
    "    return  Seq.Seq(seq1).translate(table=11) == Seq.Seq(seq2).translate(table=11)\n",
    "\n",
    "#see if both part of cds encode for the same protein (recoded vs wt):\n",
    "same_prot = []\n",
    "for dic in recoded_in_frame_cdsparts:\n",
    "    pos = dic[\"pos\"] \n",
    "    recoded_part = dic[\"dna_recoded\"]\n",
    "    start = recoded_dna_seq.find(recoded_part)\n",
    "    end = start+len(recoded_part)\n",
    "    print(\"Est-ce que les positions entre la version wt et wtz sont pareilles : \",(start,end) == pos)\n",
    "    print(\"Est-ce que les sequences en adn sont pareilles?: \", recoded_part==dna_seq[start:end])\n",
    "    same_prot.append(check_if_same_prot(recoded_part,dna_seq[start:end]))\n",
    "\n",
    "print(\"Est-ce que toutes les parties de cds encodent pour la meme prot: \"+str(all(same_prot)))\n",
    "\n",
    "#print some graphs, just to be sure of our recoding approach:\n",
    "\n",
    "def test_recoding_function(nb,len_of_cds,perc):\n",
    "    cds_test = \"AAA\" * len_of_cds\n",
    "    x = [((recode_perc_change(perc,cds_test,rev_ecoli_codon_usage,i)[1])/len_of_cds)*100 for i in range(nb)]\n",
    "    figure(figsize=(8, 6), dpi=80)    \n",
    "    plt.title(f\"Distribution of % of recoding for {nb} dna sequences of {len_of_cds*3} bp long\", fontsize=18)\n",
    "    plt.hist(x, density=True, color=\"darkred\")  # density=False would make counts\n",
    "    plt.ylabel('Probability', fontsize=14)\n",
    "    plt.xlabel('%of recoding', fontsize=14)\n",
    "    plt.xticks(size = 14)\n",
    "    plt.yticks(size = 14)\n",
    "    plt.show()\n",
    "    \n",
    "test_recoding_function(10000,200,25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_features_colors(colors_set, pastel_factor,nb_of_features):\n",
    "#colors_set: the set of colors selected (any integers)\n",
    "#pastel_factor: the smoothness of colors (between 0-1, 0: no smoothness, 1: a lot of smoothness (near shades of greys))\n",
    "#nb_of_features: different features from a .tsv file\n",
    "#https://gist.github.com/adewes/5884820. This portion of the code have been taken from this website.\n",
    "    def get_random_color(pastel_factor):\n",
    "        return [(x+pastel_factor)/(1.0+pastel_factor) for x in [random.uniform(0,1.0) for i in [1,2,3]]]\n",
    "\n",
    "    def color_distance(c1,c2):\n",
    "        return sum([abs(x[0]-x[1]) for x in zip(c1,c2)])\n",
    "\n",
    "    def generate_new_color(existing_colors,pastel_factor):\n",
    "        max_distance = None\n",
    "        best_color = None\n",
    "        for i in range(0,100):\n",
    "            color = get_random_color(pastel_factor = pastel_factor)\n",
    "            if not existing_colors:\n",
    "                return color\n",
    "            best_distance = min([color_distance(color,c) for c in existing_colors])\n",
    "            if not max_distance or best_distance > max_distance:\n",
    "                max_distance = best_distance\n",
    "                best_color = color\n",
    "        return best_color\n",
    "\n",
    "    #To make your color choice reproducible, uncomment the following line:\n",
    "    random.seed(colors_set)\n",
    "\n",
    "    colors = []\n",
    "    for i in range(0,nb_of_features):\n",
    "        colors.append(generate_new_color(colors,pastel_factor = pastel_factor))\n",
    "    colors_hexa = []\n",
    "    for color in colors:\n",
    "        list_of_rgb255 = tuple([round(rgb*255) for rgb in color])\n",
    "        colors_hexa.append('#%02x%02x%02x' % list_of_rgb255) #convert rgb 255 color format to hexadecimal (needed in the gff3 format). Add the color in the list of colors\n",
    "    return colors_hexa\n",
    "\n",
    "#get infos with the initial fasta file and generate annotations:\n",
    "\n",
    "ajusted1_pos  = [{\"start\":dics[\"pos\"][0]+1, \"end\":dics[\"pos\"][1], \"annotations\":dics[\"annotations\"]} for dics in recoded_in_frame_cdsparts]\n",
    "corrected_pos = []\n",
    "rec_perc_annotation = []\n",
    "overlap_coor = []\n",
    "for dics in ajusted1_pos:\n",
    "    start = dics[\"start\"]\n",
    "    end = dics[\"end\"]\n",
    "    annotations = dics[\"annotations\"]\n",
    "    if start == 1:\n",
    "        new_start = [(dic[\"start\"],dic[\"end\"],dic[\"annotations\"]) for dic in ajusted1_pos if dic[\"end\"] == len_dna_seq] #should only be one value! (should work everytime)\n",
    "        if new_start:\n",
    "            start = new_start[0][0]\n",
    "            corrected_pos.append((start, end))\n",
    "            rec_perc_annotation.append((dic[\"annotations\"]+annotations)/2)     \n",
    "    else:\n",
    "        if start == new_start[0] and end == new_start[1] and annotations == new_start[2]:\n",
    "            continue\n",
    "        if end == len_dna_seq:\n",
    "            continue\n",
    "        corrected_pos.append((start, end))\n",
    "        rec_perc_annotation.append(annotations)\n",
    "\n",
    "if len(corrected_pos) != len(rec_perc_annotation):\n",
    "    print(\"THERE IS A PROBLEM WITH ANNOTATIONS, PLEASE REVISE YOUR CODE\")\n",
    "\n",
    "nw_ann = [[f\"recoded@{rec_perc_annotation[i]}\",str(pos[0]),str(pos[1]),str(pos[1]-pos[0]),\"+\",\"misc_feature\"] for i,pos in enumerate(corrected_pos)]\n",
    "df_nw_ann = pd.DataFrame(nw_ann, columns=[\"Features_name\",'start','end','length', 'strand','feat_type'])\n",
    "#print(df_nw_ann)\n",
    "#next step, get features from the .tsv file\n",
    "df = pd.read_csv('../ref_files_m13/Features_from_m13mp18-Kan_annotated_nonatg.txt', delimiter = \"\\t\", header=None)\n",
    "df.columns = [\"Features_name\",\"start..end\",\"length\",\"strand\",\"feat_type\"]\n",
    "df[['start', 'end']] = df['start..end'].str.split('.',1,expand=True)\n",
    "df['end'] = df['end'].str.replace('.',\"\")\n",
    "df[\"start\"] = pd.to_numeric(df[\"start\"], errors='coerce')\n",
    "df[\"end\"] = pd.to_numeric(df[\"end\"], errors='coerce')\n",
    "df['strand'] = df['strand'].str.replace('=>',\"+\")\n",
    "df['strand'] = df['strand'].str.replace('<=',\"-\")\n",
    "df['strand'] = df['strand'].str.replace('==',\".\")\n",
    "del df['start..end']\n",
    "df = df.append(df_nw_ann, ignore_index=True)\n",
    "\n",
    "#fix the number of disctinct features in dataframe:\n",
    "distinct_features_type = []\n",
    "distinct_misc_features_type = []\n",
    "for index, row in df.iterrows():\n",
    "    if row[\"feat_type\"] not in distinct_features_type and row[\"feat_type\"] != \"misc_feature\":\n",
    "        distinct_features_type.append(row[\"feat_type\"])\n",
    "    if row[\"feat_type\"] == \"misc_feature\":\n",
    "        for genomic_element in [\"prom\", \"rbs\",\"term\",\"rec\",\"non_atg\"]: #cette liste doit etre demandee a l'utilisateur\n",
    "            if genomic_element in row[\"Features_name\"]:\n",
    "                if genomic_element not in distinct_misc_features_type:\n",
    "                    distinct_misc_features_type.append(genomic_element)\n",
    "total_features = distinct_features_type+distinct_misc_features_type\n",
    "colors = generate_features_colors(10,0,len(total_features))\n",
    "feat_color = {feature:color for feature,color in zip(total_features,colors)}\n",
    "colors_indf = []\n",
    "for index, row in df.iterrows():\n",
    "    if row[\"feat_type\"] != \"misc_feature\":\n",
    "        colors_indf.append(feat_color[row[\"feat_type\"]])\n",
    "    else:\n",
    "        for genomic_element in [\"prom\", \"rbs\",\"term\",\"rec\",\"non_atg\"]: #cette liste doit etre demandee a l'utilisateur\n",
    "            if genomic_element in row[\"Features_name\"]:\n",
    "                colors_indf.append(feat_color[genomic_element])\n",
    "#get mean_perc\n",
    "mean_perc = round(statistics.mean([i for i in rec_perc_annotation]),3)\n",
    "df['colors'] = colors_indf\n",
    "gff3 = df.copy()\n",
    "gff3[\"ID\"] = f\"m13_rec{mean_perc}\"\n",
    "gff3[\"source\"] = \"LaboRodrigue_JFL\"\n",
    "gff3[\"start\"] = gff3[\"start\"]\n",
    "gff3[\"annotation\"] = 'ID='+gff3['feat_type']+';Name='+gff3['Features_name']+';InRegion;color='+gff3['colors']\n",
    "gff3[\"chromosome\"] = \"L1\"\n",
    "gff3[\"gene\"] = \"gene\"\n",
    "gff3[\"point\"] = \".\"\n",
    "\n",
    "gff3 = gff3[['ID', 'source', 'feat_type', 'start', 'end', 'point', 'strand', 'point', 'annotation']]\n",
    "gff3 = gff3.drop_duplicates(subset=['start',\"end\"], keep=\"last\")\n",
    "\n",
    "gff3.to_csv('../ref_files_m13/trial_Features_from_m13mp18-Kan_annotated.gff3', sep= \"\\t\")\n",
    "                    \n",
    "with open(str('../ref_files_m13/trial_Features_from_m13mp18-Kan_annotated.gff3'), 'w') as f_in:\n",
    "    f_in.write('##gff-version 3'+'\\n')\n",
    "    f_in.write(gff3.to_csv(header = False, index = False, sep = '\\t'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "length of recoded m13 genome : 7282 and wt m13 genome : 7282\n",
      "Number of punctual synonymous mutations in all m13 genome: 505. 429 codons are changed in all cds of the recoded m13 genome. 7.882004058061495% of the m13's genome is modified\n"
     ]
    }
   ],
   "source": [
    "#Stats on recoded genome:\n",
    "#is recoded and wt m13 have the same length?\n",
    "print(f\"length of recoded m13 genome : {len(recoded_dna_seq)} and wt m13 genome : {len(dna_seq)}\")\n",
    "\n",
    "#How many punctual mutations in all the recoded genome compared to wt:\n",
    "count = sum(1 for a, b in zip(dna_seq, recoded_dna_seq) if a != b)\n",
    "print(f\"Number of punctual synonymous mutations in all m13 genome: {count}. {total_changed_codons} codons are changed in all cds of the recoded m13 genome. {(count/6407)*100}% of the m13's genome is modified\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "#draft of may things!\n",
    "        \n",
    "        #append something\n",
    "        \n",
    "                \n",
    "            # else:\n",
    "            #     new_coors = {}\n",
    "            #     wrong_frame = True\n",
    "            #     wrong_frame_lastpos = next_coors[1]\n",
    "            #     if next_coors[1] < positions[1]: #in the case the next cds (next_coors) is not in frame and shorter than the cds it is compared with:\n",
    "            #         #appending the part before next_coors[0]:\n",
    "            #         if not cds_smaller:\n",
    "            #             parts_to_recode.append({str(dna_seq[positions[0]:next_coors[0]]):(positions[0],next_coors[0])})\n",
    "            #             last_start = next_coors[0]\n",
    "            #             last_end = next_coors[1]\n",
    "            #             cds_smaller = True\n",
    "            #             try:\n",
    "            #                 next_coors = all_locations[all_locations.index(next_coors[0],next_coors[1])+1]\n",
    "            #             except: #if we are at the end of the list, need to append the recoded parts not overlaping with the little cds:                                                                        \n",
    "            #                 #appending the last part of the recodable cds, excluding the cds not in frame:\n",
    "            #                 new_coors[dna_seq[next_coors[1]:positions[1]]] = (next_coors[1],positions[1])\n",
    "            #                 parts_to_recode.append({str(dna_seq[next_coors[1]:positions[1]]):(next_coors[1],positions[1])})\n",
    "            #         else: #if multiple smaller CDSs\n",
    "            #             parts_to_recode.append({str(dna_seq[last_start:next_coors[0]]):(positions[0],next_coors[0])})\n",
    "            #             last_start = next_coors[0]\n",
    "            #             try:\n",
    "            #                 next_coors = all_locations[all_locations.index(next_coors[0],next_coors[1])+1]\n",
    "            #             except: #if we are at the end of the list, need to append the recoded parts not overlaping with the little cds:                                                                        \n",
    "            #                 #appending the last part of the recodable cds, excluding the cds not in frame:\n",
    "            #                 new_coors[dna_seq[next_coors[1]:positions[1]]] = (next_coors[1],positions[1])\n",
    "            #             break\n",
    "            overlap_coor = #     else:\n",
    "            #         in_frame_startpos = last_end+((3-last_end-positions[0])%3)\n",
    "            #         in_frame_end_pos = ###ACHANGER POUR LE IN FRAME ET REMPLACER LE NEXT_COORD[0] PLUS BAS!\n",
    "            #         if cds_smaller and in_frame_startpos =< next_coors[0]+3: #these two might overlap! Get at least 1 codon !\n",
    "            #             in_frame_pos = (3-last_end-positions[0])%3\n",
    "            #             parts_to_recode.append({str(dna_seq[in_frame_pos:next_coors[0]]):(positions[0],next_coors[0])})\n",
    "            #         break\n",
    "            # try:\n",
    "            #     next_coors = all_locations[all_locations.index(next_coors[0],next_coors[1])+1]\n",
    "            # except: #if we are at the end of the list\n",
    "            #     break\n",
    "                \n",
    "    # elif wrong_frame: #do not append in dict if overlaping region not in frame\n",
    "    #     if cds_smaller:\n",
    "    #         parts_to_recode.append({str(dna_seq[next_coors[1]:positions[1]]):(next_coors[1],positions[1])})\n",
    "    #     continue\n",
    "    # elif new_coors:\n",
    "    #     parts_to_recode.append(new_coors)\n",
    "    # elif not new_coors:\n",
    "    #     parts_to_recode.append({str(dna_seq[positions[0]:positions[1]]):(positions[0],positions[1])})\n",
    "    # print(parts_to_recode)\n",
    "    \n",
    "# CDSs_no_overlaps = {}\n",
    "# whose_overlaps = {}   #gene key overlaps with gene value.\n",
    "# for i, (gene_name, coors) in enumerate(CDSs.items()):\n",
    "#     CDSs_no_overlaps[gene_name] = {}\n",
    "#     if i+1 < len(CDSs):\n",
    "#         next_coors = list(CDSs.values())[i+1]\n",
    "#         new_coors = {}\n",
    "#         while next_coors[0] <= coors[1] and coors[0] <= next_coors[1] : # if overlap between 2 cdss\n",
    "#             whose_overlaps[list(CDSs.keys())[i+1]] = gene_name\n",
    "#             if gene_name in whose_overlaps: #meaning gene key starts not at the begining of itself, rather after the known end of previously fixed overlap\n",
    "#                 max_start_of_gene = max([CDSs[gene][1] for gene in whose_overlaps if gene == gene_name])\n",
    "#             else:\n",
    "#                 max_start_of_gene = coors[0]\n",
    "#             if same_ORF(coors[1],next_coors[0]): #if overlaping cds are in frame, it is possible to recode\n",
    "#                 new_coors[str(dna_seq[max_start_of_gene:next_coors[0]])] = (max_start_of_gene,coors[1])\n",
    "#             else: #if there is just one cds not in frame, even if there is other overlaping cds, we cannot recode this part\n",
    "#                 new_coors[str(dna_seq[max_start_of_gene:next_coors[0]])] = (max_start_of_gene,next_coors[0])\n",
    "#                 break\n",
    "#             next_coors = list(CDSs.values())[list(CDSs.values()).index((next_coors[0],next_coors[1]))+1]\n",
    "#             print(next_coors,gene_name)\n",
    "#         if not new_coors:\n",
    "#             CDSs_no_overlaps[gene_name] = {str(dna_seq[coors[0]:coors[1]]):(coors[0],coors[1])}\n",
    "            \n",
    "# print(CDSs_no_overlaps)\n",
    "\n",
    "            \n",
    "            \n",
    "\n",
    "\n",
    "\n",
    "# cds_part_no_prob_overlap = overlap_situation(cds_parts,CDSs)\n",
    "# print(cds_part_no_prob_overlap)\n",
    "# #print(\"j suis quoi???\",cds_part_no_prob_overlap)\n",
    "\n",
    "# recoded_genome_parts = {}\n",
    "# for gene_name, parts in cds_part_no_prob_overlap.items():\n",
    "#     recoded_genome_parts[gene_name] = {}\n",
    "#     for part,coor in parts.items():\n",
    "#         #Search for orf coordinates and recoded the cds:\n",
    "#         start_part = coor[0]\n",
    "#         end_part = coor[1]\n",
    "#         start_codon = start_part+(start_part%3)\n",
    "#         end_codon = end_part-(end_part%3)\n",
    "#         if start_part < end_part:\n",
    "#             in_frame_part = dna_seq[start_codon:end_codon]\n",
    "#             recoded_part = str(recode_perc_change(50,in_frame_part,rev_ecoli_codon_usage))\n",
    "#             if recoded_part != \"\":\n",
    "#                 recoded_genome_parts[gene_name][recoded_part] = (start_codon,end_codon)\n",
    "#         else:\n",
    "#             print(\"minou\")\n",
    "#             in_frame_part_end = dna_seq[start_codon:max_len_dna]\n",
    "#             in_frame_part_start = dna_seq[0:end_codon]\n",
    "#             recoded_part_end = str(recode_perc_change(50,in_frame_part_end,rev_ecoli_codon_usage))\n",
    "#             recoded_part_start = str(recode_perc_change(50,in_frame_part_start,rev_ecoli_codon_usage))\n",
    "#             if recoded_part != \"\":\n",
    "#                 recoded_genome_parts[gene_name][in_frame_part_end] = (start_codon,max_len_dna)\n",
    "#                 recoded_genome_parts[gene_name][recoded_part_start] = (0,end_codon)       \n",
    "# print(recoded_genome_parts)\n",
    "\n",
    "\n",
    "#construct the final m13 vector:\n",
    "# all_recoded_start_coor = [str(list(recoded_parts.values())).split(\",\")[0][2:] for recoded_parts in recoded_genome_parts.values()]\n",
    "# all_recoded_start_coor = [(recoded_parts.values()) for recoded_parts in recoded_genome_parts.values()]\n",
    "# print(all_recoded_start_coor)\n",
    "# all_recoded_end_coor = [coor[1] for seq,coor in parts.items() for recoded_parts in recoded_genome_parts.values()]\n",
    "# flat_recoded_cds = {coor:seq for seq,coor in partsATGA.items() for parts in recoded_genome_parts.values()}\n",
    "# print(flat_recoded_cds)\n",
    "\n",
    "# max_len_dna = len(str(dna_seq))[1,2,3,4,5,6,7,8].index(2)+1\n",
    "# m13_WTZ = \"\"\n",
    "# last_pos = 0\n",
    "# for i in range(max_len_dna):\n",
    "#     if i in all_recoded_start_coor:\n",
    "#         m13_WTZ += dna_seq[last_pos:i]\n",
    "#         last_pos = i\n",
    "#     elif i in all_recoded_end_coor:\n",
    "#         m13_WTZ += dna_seq[last_pos:i]\n",
    "#         last_pos = i\n",
    "#     else:\n",
    "#         continue\n",
    "# print(len(m13_WTZ))\n",
    "        \n",
    "        \n",
    "#     print(\"old\",cds_parts, \"\\n\")    \n",
    "#     #le gene en chevauche-t-il un autre? Si oui, couper la partie correspondante dans la liste de gene\n",
    "#     cds_parts = overlap_situation(coor,cds_parts) \n",
    "#     print(\"new\",cds_parts,\"\\n\")\n",
    "#     print(old_cds_parts==cds_parts)    \n",
    "#     cds_recoded = {}\n",
    "#     traceback = []\n",
    "        \n",
    "                 \n",
    "        \n",
    "        \n",
    "        #when forbiden_regions[0] < coor[0]\n",
    "    #find the reading frame in each part of changeable CDSs\n",
    "    \n",
    "    #change the reading frame in cds\n",
    "    \n",
    "    #merge all substring into a new file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "86\n",
      "nb codon total 7\n",
      "{2: 'GCA', 5: 'ACG', 1: 'CGT', 6: 'GTT', 4: 'TCA', 3: 'TCT'}\n",
      "6\n",
      "6\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "# new_cds = \"\"\n",
    "# for pos in range(int(len(cds)/3)):    #25% de chance de le changer\n",
    "#     codon_in_cds = cds[pos*3:pos*3+3].upper()\n",
    "#     #set seed at each position in the CDS\n",
    "#     #random.seed(pos+pos_modifier)\n",
    "#     if codon_in_cds != \"ATG\" and codon_in_cds != \"TGG\":   #if we exclude those codon, because they have no synonymous.\n",
    "#         if pick_this_pos(perc): #and codon_in_cds != \"ATG\" and codon_in_cds != \"TGG\" and pos*3+3 != len(cds): ##########################!!!!!!!!!!!!!!!!!!!\n",
    "#             #if no TGG (cannot be changed), no start codon (cannot be changed)\n",
    "#             number_of_changed_codon += 1\n",
    "#             other_codons = {}\n",
    "#             for aa, codon_freq in codon_table.items():\n",
    "#                 if codon_in_cds in codon_freq:\n",
    "#                     #select the dict for the codon to recode, and evaluate the freq of synonymous codon without codon_in_cds.\n",
    "#                     for codon,freq in codon_freq.items():\n",
    "#                         if codon_in_cds != codon:\n",
    "#                             other_codons[codon] = freq\n",
    "#             #no need to ponderate frequency now that the codon_in_cds is removed because relative weights are ponderate by default with cumulative wieght (sum of weights)\n",
    "#             #https://docs.python.org/dev/library/random.html#random.choices\n",
    "#             #https://stackoverflow.com/questions/55440986/cumulative-weights-in-random-choices\n",
    "#             replacement_codon = random.choices(population=list(other_codons.keys()),weights=list(other_codons.values()),k=1)\n",
    "#             new_cds += ''.join(replacement_codon).upper()\n",
    "#         else:\n",
    "#             new_cds += codon_in_cds\n",
    "#     else:\n",
    "#         new_cds += codon_in_cds\n",
    "# #new_cds is recoded\n",
    "# return new_cds,number_of_changed_codon\n",
    "\n",
    "#recode_perc_change(perc,cds_test,rev_ecoli_codon_usage,1)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#test the recoding function:\n",
    "#BE SURE TO TEST THE FUNCTION WITHOUT THE VALUE OF SEED FIXED\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
