{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqUtils import GC\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import statistics\n",
    "import operator\n",
    "from tqdm.notebook import tqdm\n",
    "from functions import *\n",
    "import random\n",
    "from Bio import Restriction\n",
    "import copy\n",
    "from itertools import combinations\n",
    "from itertools import permutations\n",
    "from itertools import product\n",
    "import urllib.request as urllib2\n",
    "from dna_features_viewer import BiopythonTranslator\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5.846007e+48\n",
      "1.500000e+53\n"
     ]
    }
   ],
   "source": [
    "import decimal\n",
    "#nombre de génomes de cette longueur possibles:\n",
    "x = 4**81\n",
    "d = decimal.Decimal(x)\n",
    "print(format(d, '.6e'))\n",
    "#masse de l'univers observable:\n",
    "#1.5 × 10**53 kg\n",
    "#nombre de parties d'echecs possibles: 10**120\n",
    "#Le nombre d'atomes dans l'univers a un ordre de grandeur estimé à 10**80 atomes\n",
    "print(format(decimal.Decimal(1.5*10**53),'.6e'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#input section: everything we have to declare before doing this pipeline:\n",
    "\n",
    "path_to_all_cdss_m13 = \"../ref_files_m13/genes_m13_rutgers.txt\"\n",
    "path_to_whole_genome_m13 = \"../ref_files_m13/complete_sequence_m13_rutgers.fasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'K': {'AAA': 0.767, 'AAG': 0.233}, 'N': {'AAT': 0.448, 'AAC': 0.552}, 'I': {'ATA': 0.07, 'ATT': 0.509, 'ATC': 0.421}, 'M': {'ATG': 1.0}, 'T': {'ACA': 0.129, 'ACT': 0.165, 'ACC': 0.438, 'ACG': 0.268}, 'R': {'AGA': 0.036, 'AGG': 0.02, 'CGA': 0.063, 'CGT': 0.383, 'CGC': 0.402, 'CGG': 0.097}, 'S': {'AGT': 0.15, 'AGC': 0.278, 'TCA': 0.122, 'TCT': 0.146, 'TCC': 0.149, 'TCG': 0.154}, '*': {'TAA': 0.64, 'TAG': 0.07, 'TGA': 0.289}, 'Y': {'TAT': 0.568, 'TAC': 0.432}, 'L': {'TTA': 0.13, 'TTG': 0.128, 'CTA': 0.036, 'CTT': 0.103, 'CTC': 0.105, 'CTG': 0.498}, 'F': {'TTT': 0.573, 'TTC': 0.427}, 'C': {'TGT': 0.442, 'TGC': 0.558}, 'W': {'TGG': 1.0}, 'Q': {'CAA': 0.347, 'CAG': 0.653}, 'H': {'CAT': 0.569, 'CAC': 0.431}, 'P': {'CCA': 0.19, 'CCT': 0.157, 'CCC': 0.123, 'CCG': 0.529}, 'E': {'GAA': 0.69, 'GAG': 0.31}, 'D': {'GAT': 0.626, 'GAC': 0.374}, 'V': {'GTA': 0.154, 'GTT': 0.258, 'GTC': 0.216, 'GTG': 0.372}, 'A': {'GCA': 0.212, 'GCT': 0.161, 'GCC': 0.27, 'GCG': 0.357}, 'G': {'GGA': 0.107, 'GGT': 0.338, 'GGC': 0.406, 'GGG': 0.149}}\n"
     ]
    }
   ],
   "source": [
    "#Create a codon table for m13:\n",
    "# all_cds_m13 = {str(record.description).split(\"gene=\")[1].split(\"]\",1)[0]:str(record.seq)\n",
    "#                       for record in SeqIO.parse(path_to_all_cdss_m13, \"fasta\")}\n",
    "\n",
    "# codon_table_m13 = create_codon_table(list(all_cds_m13.values()), 11)\n",
    "\n",
    "#Create a codon table for e.coli:\n",
    "all_name_cds_ecoli = {str(record.description).split(\"gene=\")[1].split(\"]\",1)[0]:str(record.seq)\n",
    "                      for record in SeqIO.parse(\"../ref_files_m13/GCF_000005845.2_ASM584v2_cds_from_genomic.fna\", \"fasta\")}\n",
    "codon_table_ecoli = create_codon_table(list(all_name_cds_ecoli.values()), 11)\n",
    "\n",
    "# Make a reversed dict codon usage of whole genome\n",
    "rev_ecoli_codon_usage = reverse_codon_table(codon_table_ecoli)\n",
    "\n",
    "print(rev_ecoli_codon_usage)\n",
    "\n",
    "\n",
    "#IF NEEDED %GC for each gene:\n",
    "# for key, value in all_name_cds_m13.items():\n",
    "#     print(key,GC(value))\n",
    "# codon_table_m13 = create_codon_table(list(all_name_cds_m13.values()), 11)\n",
    "\n",
    "# Make a reversed dict codon usage of m13 whole genome:\n",
    "#rev_codon_table_m13 = reverse_codon_table(codon_table_m13)\n",
    "#print(rev_codon_table_m13)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Le % GC moyen dans tout le génome de m13 WT est de: 40.76791009832995\n"
     ]
    }
   ],
   "source": [
    "#find %GC in the complete fasta:\n",
    "for record in SeqIO.parse(path_to_whole_genome_m13, \"fasta\"):\n",
    "    complete_genome_m13 = record.seq\n",
    "print(\"Le % GC moyen dans tout le génome de m13 WT est de: \" +str(GC(complete_genome_m13)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#just for fun: draft for recoding stuff and remove motif:\n",
    "def get_key_by_value(nested_dict, codon):\n",
    "    for i in nested_dict:\n",
    "        if codon in nested_dict[i].keys():\n",
    "            return i\n",
    "        \n",
    "def normalize_list(list_to_normalize):\n",
    "    return [i/sum(list_to_normalize) for i in list_to_normalize]\n",
    "\n",
    "def recode_stuff(original_codon_table, futur_codon_table, seq, seed):\n",
    "    #create a codon table that will change during the modification of the recoded_sequence\n",
    "    modifiable_codon_table = copy.deepcopy(futur_codon_table)\n",
    "    recoded_cds = \"\"\n",
    "    for pos in range(int(len(seq)/3)):\n",
    "        # Set the seed\n",
    "        random.seed(pos+seed)\n",
    "        codon = seq[(pos*3):(pos*3)+3]\n",
    "        amino_acid = get_key_by_value(original_codon_table,codon)\n",
    "        recoded_codon = random.choices(population = list(futur_codon_table[amino_acid].keys()),\n",
    "                       weights = list(futur_codon_table[amino_acid].values()),\n",
    "                       k=1)\n",
    "        recoded_cds += recoded_codon[0]\n",
    "    return recoded_cds\n",
    "\n",
    "def remove_any_motif(all_forbidden_seqs, recoded_cds, obj_codon_table):\n",
    "    \n",
    "    #add reverse complement to list_forbidden_seqs :\n",
    "    list_forbidden_seqs = []\n",
    "    for motif in all_forbidden_seqs:\n",
    "        list_forbidden_seqs.append(motif)\n",
    "        if Seq.Seq(motif).reverse_complement() not in list_forbidden_seqs:\n",
    "            list_forbidden_seqs.append(str(Seq.Seq(motif).reverse_complement()))\n",
    "            \n",
    "    #need a section to certify if the job is possible:\n",
    "    if any(len(forbidden_seq) < 3 for forbidden_seq in list_forbidden_seqs):\n",
    "        print(\"********WARNING********\\nneed to remove motif(s) that are less than 3 bp in the list of forbidden sequences\")\n",
    "        print(\"no output provided\")\n",
    "        print(\"************************\")\n",
    "    else:\n",
    "        #if list_forbidden_seqs contains removable sequences only\n",
    "        n = 0\n",
    "        while any(forbidden_seq in recoded_cds for forbidden_seq in list_forbidden_seqs) and n < 10000:\n",
    "            n+=1\n",
    "            for forbidden_seq in list_forbidden_seqs:\n",
    "                pos_of_forbidden_seq = [motif.start() for motif in re.finditer(forbidden_seq,recoded_cds)]\n",
    "                #find ORF\n",
    "                for it,pos in enumerate(pos_of_forbidden_seq):\n",
    "                    motif_end = pos+len(forbidden_seq)\n",
    "                    min_in_frame = pos-(pos%3)\n",
    "                    max_in_frame = motif_end+((3-motif_end%3)%3)\n",
    "                    substring_no_motif = recode_stuff(rev_bacillus_codon_usage, rev_bacillus_codon_usage,\n",
    "                                 recoded_cds[min_in_frame:max_in_frame], it)\n",
    "                    recoded_cds = recoded_cds[:min_in_frame]+substring_no_motif+recoded_cds[max_in_frame:]\n",
    "        if n < 10000:\n",
    "            return recoded_cds\n",
    "        else:\n",
    "            print(f\"no output because the following sequence:\\n{recoded_cds}\\ncontains at least one forbidden sequence.\\n********NO OUTPUT*********\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTCTCACCCTTTTGAATCTTTACCTACACATTACTCAGGCATTGCATTTAAAATATATGAGGGTTCTAAAAATTTTTATCC\n"
     ]
    }
   ],
   "source": [
    "#try to remove restriction sites from all m13 promoters:\n",
    "#trouver toutes les fenêtres de 81 pb pour chacun des gènes dans m13:\n",
    "#II, X, V, VII, IX and VIII are transcribed very frequently aka 1st domain:\n",
    "#it is separated from the second domain (genes III, VI, I and IV) by a Rho-independent transcriptional terminator:\n",
    "#autrement dit, il faut enlever toutes les fenêtres sauf II et III:\n",
    "\n",
    "#Règle: 20 apres la premiere base de la séquence consensus du -35:\n",
    "#Toutes les lettres peuvent être modifiée pour \n",
    "\n",
    "#gene X: aagtctttcgggcttcctcttAATcTttttGaTGcaAtccgctttgcttctGacTATAATagtcagggtaaagacctgatt\n",
    "\n",
    "# le gene II comprend l'entierete de cette sequence:\n",
    "#ATTENTION, ON NE VEUT PAS ELIMINER LE RBS!\n",
    "#le rbs du gene X n'est pas present dans cette sequence. \n",
    "prom = \"ttctcacccttttgAATcTttacCTacACAttactcaggcattgcattTAaAATatatgagggttctaaaaatttttatcc\"\n",
    "print(prom.upper())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTTTAAAATTAATAACGTTCGGGCAAAGGATTTAATACGAGTTGTCGAATTGTTTGTAAAGTCTAATACTTCTAAATCCTC\n"
     ]
    }
   ],
   "source": [
    "def any_upper_case(substring):\n",
    "    for nt in substring:\n",
    "        if nt.isupper():\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "def position_of_uppercase(substring):\n",
    "    return [pos for pos,nt in enumerate(substring) if nt.isupper()]\n",
    "\n",
    "def find_all_replacement_codons(window, codon_table, start_codon):\n",
    "    if len(window)%3 != 0:\n",
    "        print(\"revise your code, Mr Julien!\")\n",
    "        return\n",
    "    switch_codons = {}\n",
    "    for n in range(int(len(window)/3)):\n",
    "        codon = window[n*3:n*3+3]\n",
    "        for aa, codontype_freq in codon_table.items():\n",
    "            if codon in codontype_freq:\n",
    "                switch_codons[start_codon+n*3] = list(codontype_freq.keys())\n",
    "    return switch_codons\n",
    "\n",
    "def check_if_same_prot(seq1, seq2):\n",
    "    return  Seq.Seq(seq1).translate(table=11) == Seq.Seq(seq2).translate(table=11)\n",
    "\n",
    "print(\"ttttaaaattAATaacgttCggGcaAaggatttaatacgagTtgTcgAATtgtttgtaAagtctaatacttctaaatcctc\".upper())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "For the promoter of X (in gene II):\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{(0, 21): ['AAGTCTTTCGGGCTTCCTCTT'], (21, 27): ['AATTTA', 'AATTTG', 'AATCTA', 'AATCTT', 'AATCTC', 'AATCTG', 'AACTTA', 'AACTTG', 'AACCTA', 'AACCTT', 'AACCTC', 'AACCTG'], (27, 30): ['TTT'], (30, 39): ['GATGCAATA', 'GATGCAATT', 'GATGCAATC', 'GATGCTATA', 'GATGCTATT', 'GATGCTATC', 'GATGCCATA', 'GATGCCATT', 'GATGCCATC', 'GATGCGATA', 'GATGCGATT', 'GATGCGATC', 'GACGCAATA', 'GACGCAATT', 'GACGCAATC', 'GACGCTATA', 'GACGCTATT', 'GACGCTATC', 'GACGCCATA', 'GACGCCATT', 'GACGCCATC', 'GACGCGATA', 'GACGCGATT', 'GACGCGATC'], (39, 51): ['CGCTTTGCTTCT'], (51, 60): ['GATTATAAT', 'GATTATAAC', 'GATTACAAT', 'GATTACAAC', 'GACTATAAT', 'GACTATAAC', 'GACTACAAT', 'GACTACAAC'], (60, 81): ['AGTCAGGGTAAAGACCTGATT']}\n",
      "2304\n"
     ]
    }
   ],
   "source": [
    "#for prom X:\n",
    "#this code get all possible potentially altered promoter in a single cds.\n",
    "\n",
    "prom = \"aagtctttcgggcttcctcttAATcTttttGaTGcaAtccgctttgcttctGacTATAATagtcagggtaaagacctgatt\"       #gene X\n",
    "\n",
    "assert len(prom) == 81, \"normalement, la longueur de la séquence promotrice devrait être de 81 pb exactement\"\n",
    "assert not any_upper_case(prom[0:3]),\"pas de lettre majuscule dans les trois premiers nucleotides, sinon ca plante...\"\n",
    "ref_gene = all_cds_m13[\"II\"].lower()\n",
    "\n",
    "start_of_prom_in_ref_gene = ref_gene.find(prom.lower())\n",
    "substrings_to_consider = {}\n",
    "traceback = 0\n",
    "boolean = True\n",
    "promseq_in_refgene = ref_gene[start_of_prom_in_ref_gene:start_of_prom_in_ref_gene+len(prom)]\n",
    "\n",
    "for pos, nt in enumerate(prom):\n",
    "    #set the seed:\n",
    "    if pos >= traceback:\n",
    "        if nt.isupper():\n",
    "            #est-ce qu'il y a une série de codons contenant une lettre majuscule?:\n",
    "            pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos\n",
    "            #start codon: le codon in frame (in prom) duquel on commence à changer des codons\n",
    "            start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "            pos_of_prom_in_frame = pos-(pos_uppercase_in_ref_gene%3)\n",
    "            rest_of_81bp = prom[pos_of_prom_in_frame+3:]\n",
    "            window = prom[start_codon:start_codon+3]\n",
    "            n = 0\n",
    "            for next_codon in rest_of_81bp:\n",
    "                if any(t.isupper() for t in rest_of_81bp[n:n+3]):\n",
    "                    #extend the portion of prom to change:\n",
    "                    window = window+rest_of_81bp[n:n+3]\n",
    "                    n+=3\n",
    "                else:\n",
    "                    break\n",
    "            switch_codon = find_all_replacement_codons(window.upper(),rev_ecoli_codon_usage, start_codon)\n",
    "            list_switch_codon = [possibilities for position, possibilities in switch_codon.items()]\n",
    "            substrings_to_consider[start_codon,start_codon+n+3] = [''.join(i) for i in list(product(*list_switch_codon))]\n",
    "            traceback = start_codon+n+3\n",
    "        else:\n",
    "            if boolean:\n",
    "                #save the position to append correctly in substrings_to_consider\n",
    "                pos_1st_lower_case = pos\n",
    "                boolean = False\n",
    "            if pos < len(prom)-1:\n",
    "                if prom[pos+1].isupper():\n",
    "                    #il faut alors savoir ou est le prochain cadre de lecture:\n",
    "                    pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos+1\n",
    "                    start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "                    substrings_to_consider[pos_1st_lower_case,start_codon] = [prom[pos_1st_lower_case:start_codon].upper()]\n",
    "                    boolean = True\n",
    "            else:\n",
    "                substrings_to_consider[pos_1st_lower_case,pos+1] = [prom[pos_1st_lower_case:pos+1].upper()]\n",
    "    \n",
    "    \n",
    "print(substrings_to_consider)\n",
    "list_of_all_possibilities = [all_substring for position, all_substring in substrings_to_consider.items()]\n",
    "all_possible_promoter_X = [''.join(i) for i in list(product(*list_of_all_possibilities))]\n",
    "\n",
    "print(len(all_possible_promoter_X))\n",
    "#if no NOOOOOO, meaning all sequences encode for the same amino acid. \n",
    "for i in all_possible_promoter_X:\n",
    "    if not check_if_same_prot(ref_gene[:start_of_prom_in_ref_gene]+i+ref_gene[start_of_prom_in_ref_gene+len(i):],ref_gene):\n",
    "        print(\"NOOOOOOOOOOOOOO\")\n",
    "\n",
    "#write output file:\n",
    "with open(\"./altered_proms/promX/altered_promX.fa\",\"w\") as fout:\n",
    "    string = \"\"\n",
    "    for i,seq_prom in enumerate(all_possible_promoter_X):\n",
    "        string+=f\">promX_{i}\\n{seq_prom}\\n\"\n",
    "    fout.write(string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "longueur du dico pour l'algo de RF: 0\n",
      "je suis le plus petit score de ipro70 et et non-promoteur pour CNNprom: ('promX_1967', ['AAGTCTTTCGGGCTTCCTCTTAACCTCTTTGATGCTATCCGCTTTGCTTCTGACTACAACAGTCAGGGTAAAGACCTGATT', 0.3451])\n"
     ]
    }
   ],
   "source": [
    "#make calculations and get URL on wed server:\n",
    "#1st prom2L:\n",
    "n=0\n",
    "iProm2L_results = {}\n",
    "for i,line in enumerate(urllib2.urlopen(\"http://bioinformatics.hitsz.edu.cn/iPromoter-2L/static/result/162.217.60.184_1629309064.37.txt\")):\n",
    "    if \"The length of the query sequence\" in str(line):\n",
    "        prom_name = str(line).strip(\" (\")[1].split(\" )\")[0]\n",
    "    elif \"non-promoter sequence\" in str(line):\n",
    "        n=1\n",
    "    elif n == 1:\n",
    "        n=0\n",
    "        iProm2L_results[prom_name] = str(line).split(\":\")[1]      \n",
    "print(\"longueur du dico pour l'algo de RF:\", len(iProm2L_results))\n",
    "\n",
    "# with open('./altered_proms/promX/iProm2L_X.txt', 'a') as the_file:\n",
    "#     for prom_alt in iProm2L_results.items():\n",
    "#         the_file.write(str(prom_alt+\"\\t\"+iProm2L_results)+\"\\n\")\n",
    "        \n",
    "#ipro70 analysis:\n",
    "with open('./altered_proms/promX/ipro70_results', 'r') as fin:\n",
    "    ipro70_results = {}\n",
    "    for i, line in enumerate(fin):\n",
    "        if i%3 == 0:\n",
    "            prom_name = line.replace(\" \",\"_\").strip(\"\\n\")\n",
    "        if i%3 == 1:\n",
    "            prom_seq = line\n",
    "        if i%3 == 2 and \"Non-promoter\" in line:\n",
    "            ipro70_results[prom_name] = [prom_seq.strip(\"\\n\"),float(line.split(\"score: \")[1].strip(\"\\n\"))]\n",
    "#CNN prom analysis:            \n",
    "with open('./altered_proms/promX/CNN_prom_results', 'r') as fin:\n",
    "    CNN_prom = []\n",
    "    for i, line in enumerate(fin):\n",
    "        if i%2 == 0:\n",
    "            prom_name = line[1:].replace(\" \",\"_\").strip(\"\\n\")\n",
    "        if i%2 == 1 and \"Not a Promoter\" in line:\n",
    "            CNN_prom.append(prom_name)\n",
    "\n",
    "common_proms_CNN_ipro70 = {prom_name:prom_seq for prom_name, prom_seq in ipro70_results.items() if prom_name in CNN_prom}\n",
    "prom_seq_lowscore = min(common_proms_CNN_ipro70.items(), key=lambda x: x[1][1])\n",
    "print(\"je suis le plus petit score de ipro70 et un non-promoteur pour CNNprom:\", prom_seq_lowscore)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "print(0%3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "1000\n",
      "ATGGAGCAGGTCGCGGATTTCGACACAATTTATCAGGCGATGATACAAATCTCCGTTGTACTTTGTTTCGCGCTTGGTATAATCGCTGGGGGTCAAAGA\n"
     ]
    }
   ],
   "source": [
    "#PROM II, WE'D LIKE NOT TO MODIFY THIS \n",
    "#for prom II:\n",
    "#THIS CASE IS VERY DIFFERENT, BECAUSE THIS PROMOTER IS IN AN INTERGENIC REGION, SO EVERY SUBSTITUTION IS POSSIBLE.\n",
    "#get every substitution for every uppercase.\n",
    "#There are too many possibilities too be all generated in a reasonable amount of time:\n",
    "#The algorithm is a while loop where nmax promoter are generated. \n",
    "\n",
    "prom = \"tttaacgcgaattttaacaaAAtaTtaacGTTtACAatttaaatatttgcTtaTAcAATcttcctGTttttggggcttttc\"       #gene II\n",
    " \n",
    "assert len(prom) == 81, \"normalement, la longueur de la séquence promotrice devrait être de 81 pb exactement\"\n",
    "assert not any_upper_case(prom[0:3]), \"pas de lettre majuscule dans les trois premiers nucleotides, sinon ca plante...\"\n",
    "switch_codon = {}\n",
    "substrings_to_consider = {}\n",
    "nts_to_consider = [\"A\",\"T\",\"C\",\"G\"]\n",
    "prev_number = 0\n",
    "nmax = 1000\n",
    "number = 0\n",
    "all_possibilities = []\n",
    "while number < nmax:\n",
    "    number +=1 \n",
    "    for i, nt in enumerate(prom):\n",
    "        if nt.isupper():\n",
    "            np.random.seed(i+number)\n",
    "            choice = [n for n in nts_to_consider if n != nt]\n",
    "            switch_codon[i] = np.random.choice(choice)\n",
    "            prev_number = i\n",
    "        else:\n",
    "            if i < len(prom)-1:\n",
    "                if prom[i+1].isupper():\n",
    "                    switch_codon[(prev_number,i)] = prom[prev_number:i].upper()\n",
    "            else:\n",
    "                switch_codon[prev_number,i] = prom[prev_number:i].upper()\n",
    "\n",
    "    an_altered_promoter = [''.join(all_substring) for position, all_substring in switch_codon.items()]\n",
    "    all_possibilities.append(''.join(an_altered_promoter))\n",
    "    \n",
    "print(len(set(all_possibilities)) == len(all_possibilities))\n",
    "print(len(all_possibilities))\n",
    "\n",
    "\n",
    "# with open(\"./tests_all_prom_seqX/X_test1.fa\",\"a\") as fout:\n",
    "#     string = \"\"\n",
    "#     for i,seq_prom in enumerate(all_possible_promoter_X):\n",
    "#         string+=f\">prom{i}\\n{seq_prom}\\n\"\n",
    "#     fout.write(string)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I DONT KNOW WHAT TO DO WITH GENE viii"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{(0, 27): ['TTCGACACAATTTATCAGGCGATGATA'], (27, 33): ['CAAATA', 'CAAATT', 'CAAATC', 'CAGATA', 'CAGATT', 'CAGATC'], (33, 36): ['TCC'], (36, 42): ['GTAGTA', 'GTAGTT', 'GTAGTC', 'GTAGTG', 'GTTGTA', 'GTTGTT', 'GTTGTC', 'GTTGTG', 'GTCGTA', 'GTCGTT', 'GTCGTC', 'GTCGTG', 'GTGGTA', 'GTGGTT', 'GTGGTC', 'GTGGTG'], (42, 54): ['CTTTGTTTCGCG'], (54, 66): ['TTAGGAATAATA', 'TTAGGAATAATT', 'TTAGGAATAATC', 'TTAGGAATTATA', 'TTAGGAATTATT', 'TTAGGAATTATC', 'TTAGGAATCATA', 'TTAGGAATCATT', 'TTAGGAATCATC', 'TTAGGTATAATA', 'TTAGGTATAATT', 'TTAGGTATAATC', 'TTAGGTATTATA', 'TTAGGTATTATT', 'TTAGGTATTATC', 'TTAGGTATCATA', 'TTAGGTATCATT', 'TTAGGTATCATC', 'TTAGGCATAATA', 'TTAGGCATAATT', 'TTAGGCATAATC', 'TTAGGCATTATA', 'TTAGGCATTATT', 'TTAGGCATTATC', 'TTAGGCATCATA', 'TTAGGCATCATT', 'TTAGGCATCATC', 'TTAGGGATAATA', 'TTAGGGATAATT', 'TTAGGGATAATC', 'TTAGGGATTATA', 'TTAGGGATTATT', 'TTAGGGATTATC', 'TTAGGGATCATA', 'TTAGGGATCATT', 'TTAGGGATCATC', 'TTGGGAATAATA', 'TTGGGAATAATT', 'TTGGGAATAATC', 'TTGGGAATTATA', 'TTGGGAATTATT', 'TTGGGAATTATC', 'TTGGGAATCATA', 'TTGGGAATCATT', 'TTGGGAATCATC', 'TTGGGTATAATA', 'TTGGGTATAATT', 'TTGGGTATAATC', 'TTGGGTATTATA', 'TTGGGTATTATT', 'TTGGGTATTATC', 'TTGGGTATCATA', 'TTGGGTATCATT', 'TTGGGTATCATC', 'TTGGGCATAATA', 'TTGGGCATAATT', 'TTGGGCATAATC', 'TTGGGCATTATA', 'TTGGGCATTATT', 'TTGGGCATTATC', 'TTGGGCATCATA', 'TTGGGCATCATT', 'TTGGGCATCATC', 'TTGGGGATAATA', 'TTGGGGATAATT', 'TTGGGGATAATC', 'TTGGGGATTATA', 'TTGGGGATTATT', 'TTGGGGATTATC', 'TTGGGGATCATA', 'TTGGGGATCATT', 'TTGGGGATCATC', 'CTAGGAATAATA', 'CTAGGAATAATT', 'CTAGGAATAATC', 'CTAGGAATTATA', 'CTAGGAATTATT', 'CTAGGAATTATC', 'CTAGGAATCATA', 'CTAGGAATCATT', 'CTAGGAATCATC', 'CTAGGTATAATA', 'CTAGGTATAATT', 'CTAGGTATAATC', 'CTAGGTATTATA', 'CTAGGTATTATT', 'CTAGGTATTATC', 'CTAGGTATCATA', 'CTAGGTATCATT', 'CTAGGTATCATC', 'CTAGGCATAATA', 'CTAGGCATAATT', 'CTAGGCATAATC', 'CTAGGCATTATA', 'CTAGGCATTATT', 'CTAGGCATTATC', 'CTAGGCATCATA', 'CTAGGCATCATT', 'CTAGGCATCATC', 'CTAGGGATAATA', 'CTAGGGATAATT', 'CTAGGGATAATC', 'CTAGGGATTATA', 'CTAGGGATTATT', 'CTAGGGATTATC', 'CTAGGGATCATA', 'CTAGGGATCATT', 'CTAGGGATCATC', 'CTTGGAATAATA', 'CTTGGAATAATT', 'CTTGGAATAATC', 'CTTGGAATTATA', 'CTTGGAATTATT', 'CTTGGAATTATC', 'CTTGGAATCATA', 'CTTGGAATCATT', 'CTTGGAATCATC', 'CTTGGTATAATA', 'CTTGGTATAATT', 'CTTGGTATAATC', 'CTTGGTATTATA', 'CTTGGTATTATT', 'CTTGGTATTATC', 'CTTGGTATCATA', 'CTTGGTATCATT', 'CTTGGTATCATC', 'CTTGGCATAATA', 'CTTGGCATAATT', 'CTTGGCATAATC', 'CTTGGCATTATA', 'CTTGGCATTATT', 'CTTGGCATTATC', 'CTTGGCATCATA', 'CTTGGCATCATT', 'CTTGGCATCATC', 'CTTGGGATAATA', 'CTTGGGATAATT', 'CTTGGGATAATC', 'CTTGGGATTATA', 'CTTGGGATTATT', 'CTTGGGATTATC', 'CTTGGGATCATA', 'CTTGGGATCATT', 'CTTGGGATCATC', 'CTCGGAATAATA', 'CTCGGAATAATT', 'CTCGGAATAATC', 'CTCGGAATTATA', 'CTCGGAATTATT', 'CTCGGAATTATC', 'CTCGGAATCATA', 'CTCGGAATCATT', 'CTCGGAATCATC', 'CTCGGTATAATA', 'CTCGGTATAATT', 'CTCGGTATAATC', 'CTCGGTATTATA', 'CTCGGTATTATT', 'CTCGGTATTATC', 'CTCGGTATCATA', 'CTCGGTATCATT', 'CTCGGTATCATC', 'CTCGGCATAATA', 'CTCGGCATAATT', 'CTCGGCATAATC', 'CTCGGCATTATA', 'CTCGGCATTATT', 'CTCGGCATTATC', 'CTCGGCATCATA', 'CTCGGCATCATT', 'CTCGGCATCATC', 'CTCGGGATAATA', 'CTCGGGATAATT', 'CTCGGGATAATC', 'CTCGGGATTATA', 'CTCGGGATTATT', 'CTCGGGATTATC', 'CTCGGGATCATA', 'CTCGGGATCATT', 'CTCGGGATCATC', 'CTGGGAATAATA', 'CTGGGAATAATT', 'CTGGGAATAATC', 'CTGGGAATTATA', 'CTGGGAATTATT', 'CTGGGAATTATC', 'CTGGGAATCATA', 'CTGGGAATCATT', 'CTGGGAATCATC', 'CTGGGTATAATA', 'CTGGGTATAATT', 'CTGGGTATAATC', 'CTGGGTATTATA', 'CTGGGTATTATT', 'CTGGGTATTATC', 'CTGGGTATCATA', 'CTGGGTATCATT', 'CTGGGTATCATC', 'CTGGGCATAATA', 'CTGGGCATAATT', 'CTGGGCATAATC', 'CTGGGCATTATA', 'CTGGGCATTATT', 'CTGGGCATTATC', 'CTGGGCATCATA', 'CTGGGCATCATT', 'CTGGGCATCATC', 'CTGGGGATAATA', 'CTGGGGATAATT', 'CTGGGGATAATC', 'CTGGGGATTATA', 'CTGGGGATTATT', 'CTGGGGATTATC', 'CTGGGGATCATA', 'CTGGGGATCATT', 'CTGGGGATCATC'], (66, 81): ['GCTGGGGGTCAAAGA']}\n",
      "10000\n"
     ]
    }
   ],
   "source": [
    "#for promoter of gene VIII:\n",
    "#Note: the sequence has a 3' end shrunk, because of its overlap with CDS VII and VIII. However, has it is now, the +1 is included.\n",
    "#Note: the rbs of gene IX overlaps with the promoter of gene VIII. \n",
    "#This section will be in lower case (protected). So, the last 'T' in the -10 box will not be changed.the +1 box is therefore, not included.\n",
    "#Note: the rbs sequence is gggggtcaaag. How things are working, no bases inn rbs should be affected.\n",
    "\n",
    "prom = \"ttcgacacaatttatcaggcgatgatacAAaTctccGTTGtactttgtttcgcgctTggTATAATcgctgggggtcaaaga\"\n",
    "assert len(prom) == 81, \"normalement, la longueur de la séquence promotrice devrait être de 81 pb exactement\"\n",
    "assert not any_upper_case(prom[0:3]),\"pas de lettre majuscule dans les trois premiers nucleotides, sinon ca plante...\"\n",
    "ref_gene = all_cds_m13[\"VII\"].lower()\n",
    "\n",
    "start_of_prom_in_ref_gene = ref_gene.find(prom.lower())\n",
    "substrings_to_consider = {}\n",
    "traceback = 0\n",
    "boolean = True\n",
    "promseq_in_refgene = ref_gene[start_of_prom_in_ref_gene:start_of_prom_in_ref_gene+len(prom)]\n",
    "\n",
    "for pos, nt in enumerate(prom):\n",
    "    #set the seed:\n",
    "    if pos >= traceback:\n",
    "        if nt.isupper():\n",
    "            #est-ce qu'il y a une série de codons contenant une lettre majuscule?:\n",
    "            pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos\n",
    "            #start codon: le codon in frame (in prom) duquel on commence à changer des codons\n",
    "            start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "            pos_of_prom_in_frame = pos-(pos_uppercase_in_ref_gene%3)\n",
    "            rest_of_81bp = prom[pos_of_prom_in_frame+3:]\n",
    "            window = prom[start_codon:start_codon+3]\n",
    "            n = 0\n",
    "            for next_codon in rest_of_81bp:\n",
    "                if any(t.isupper() for t in rest_of_81bp[n:n+3]):\n",
    "                    #extend the portion of prom to change:\n",
    "                    window = window+rest_of_81bp[n:n+3]\n",
    "                    n+=3\n",
    "                else:\n",
    "                    break\n",
    "            switch_codon = find_all_replacement_codons(window.upper(),rev_ecoli_codon_usage, start_codon)\n",
    "            list_switch_codon = [possibilities for position, possibilities in switch_codon.items()]\n",
    "            substrings_to_consider[start_codon,start_codon+n+3] = [''.join(i) for i in list(product(*list_switch_codon))]\n",
    "            traceback = start_codon+n+3\n",
    "        else:\n",
    "            if boolean:\n",
    "                #save the position to append correctly in substrings_to_consider\n",
    "                pos_1st_lower_case = pos\n",
    "                boolean = False\n",
    "            if pos < len(prom)-1:\n",
    "                if prom[pos+1].isupper():\n",
    "                    #il faut alors savoir ou est le prochain cadre de lecture:\n",
    "                    pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos+1\n",
    "                    start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "                    substrings_to_consider[pos_1st_lower_case,start_codon] = [prom[pos_1st_lower_case:start_codon].upper()]\n",
    "                    boolean = True\n",
    "            else:\n",
    "                substrings_to_consider[pos_1st_lower_case,pos+1] = [prom[pos_1st_lower_case:pos+1].upper()]\n",
    "    \n",
    "    \n",
    "print(substrings_to_consider)\n",
    "list_of_all_possibilities = [all_substring for position, all_substring in substrings_to_consider.items()]\n",
    "all_possible_promoter_VIII = [''.join(i) for i in list(product(*list_of_all_possibilities))]\n",
    "\n",
    "#if no NOOOOOO, meaning all sequences encode for the same amino acid. \n",
    "for i in all_possible_promoter_VIII:\n",
    "    if not check_if_same_prot(ref_gene[:start_of_prom_in_ref_gene]+i+ref_gene[start_of_prom_in_ref_gene+len(i):],ref_gene):\n",
    "        print(\"NOOOOOOOOOOOOOO\")\n",
    "        \n",
    "#Select random element in a list to get a much smaller input:      \n",
    "random.seed(i)\n",
    "reduced_all_possible_promoter_VIII = np.random.choice(all_possible_promoter_VIII, 10000, replace=False)\n",
    "print(len(reduced_all_possible_promoter_VIII))\n",
    "\n",
    "with open(\"./altered_proms/promVIII/altered_promVIII.fa\",\"w\") as fout:\n",
    "    string = \"\"\n",
    "    for i,seq_prom in enumerate(reduced_all_possible_promoter_VIII):\n",
    "        string+=f\">prom_VIII_{i}\\n{seq_prom}\\n\"\n",
    "    fout.write(string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10000\n"
     ]
    }
   ],
   "source": [
    "#make calculations for VIII and get URL on web server:\n",
    "#1st prom2L:\n",
    "n=0\n",
    "iProm2L_results = {}\n",
    "for i,line in enumerate(urllib2.urlopen(\"http://bioinformatics.hitsz.edu.cn/iPromoter-2L/static/result/162.217.60.184_1629311469.29.txt\")):\n",
    "    if \"The length of the query sequence\" in str(line):\n",
    "        prom_name = \"prom\"+str(line.decode(\"utf-8\")).split(\"prom\")[1].split(\" )\")[0]\n",
    "    elif \"non-promoter sequence\" in str(line):\n",
    "        n=1\n",
    "    elif n == 1:\n",
    "        n=0\n",
    "        iProm2L_results[prom_name] = str(line).split(\":\")[1]\n",
    "        \n",
    "print(len(iProm2L_results))\n",
    "\n",
    "# with open('./altered_proms/promVIII/iProm2L_VIII.txt', 'a') as the_file:\n",
    "#     for prom_alt in iProm2L_results.items():\n",
    "#         the_file.write(str(prom_alt+\"\\t\"+iProm2L_results)+\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{(0, 13): ['TTCTCACCCTTTT'], (13, 19): ['GAAAGT', 'GAAAGC', 'GAATCA', 'GAATCT', 'GAATCC', 'GAATCG', 'GAGAGT', 'GAGAGC', 'GAGTCA', 'GAGTCT', 'GAGTCC', 'GAGTCG'], (19, 22): ['TTA'], (22, 31): ['CCAACACAT', 'CCAACACAC', 'CCAACTCAT', 'CCAACTCAC', 'CCAACCCAT', 'CCAACCCAC', 'CCAACGCAT', 'CCAACGCAC', 'CCTACACAT', 'CCTACACAC', 'CCTACTCAT', 'CCTACTCAC', 'CCTACCCAT', 'CCTACCCAC', 'CCTACGCAT', 'CCTACGCAC', 'CCCACACAT', 'CCCACACAC', 'CCCACTCAT', 'CCCACTCAC', 'CCCACCCAT', 'CCCACCCAC', 'CCCACGCAT', 'CCCACGCAC', 'CCGACACAT', 'CCGACACAC', 'CCGACTCAT', 'CCGACTCAC', 'CCGACCCAT', 'CCGACCCAC', 'CCGACGCAT', 'CCGACGCAC'], (31, 46): ['TACTCAGGCATTGCA'], (46, 55): ['TTTAAAATA', 'TTTAAAATT', 'TTTAAAATC', 'TTTAAGATA', 'TTTAAGATT', 'TTTAAGATC', 'TTCAAAATA', 'TTCAAAATT', 'TTCAAAATC', 'TTCAAGATA', 'TTCAAGATT', 'TTCAAGATC'], (55, 81): ['TATGAGGGTTCTAAAAATTTTTATCC']}\n",
      "4608\n"
     ]
    }
   ],
   "source": [
    "# prom II': this prom has substitution that makes it closer than the consensus sequensus (compared to fd phage sequence)\n",
    "\n",
    "prom = \"ttctcacccttttgAATcTttacCTacACAttactcaggcattgcattTAaAATatatgagggttctaaaaatttttatcc\"\n",
    "\n",
    "assert len(prom) == 81, \"normalement, la longueur de la séquence promotrice devrait être de 81 pb exactement\"\n",
    "assert not any_upper_case(prom[0:3]),\"pas de lettre majuscule dans les trois premiers nucleotides, sinon ca plante...\"\n",
    "ref_gene = all_cds_m13[\"II\"].lower()\n",
    "\n",
    "start_of_prom_in_ref_gene = ref_gene.find(prom.lower())\n",
    "substrings_to_consider = {}\n",
    "traceback = 0\n",
    "boolean = True\n",
    "promseq_in_refgene = ref_gene[start_of_prom_in_ref_gene:start_of_prom_in_ref_gene+len(prom)]\n",
    "\n",
    "for pos, nt in enumerate(prom):\n",
    "    #set the seed:\n",
    "    if pos >= traceback:\n",
    "        if nt.isupper():\n",
    "            #est-ce qu'il y a une série de codons contenant une lettre majuscule?:\n",
    "            pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos\n",
    "            #start codon: le codon in frame (in prom) duquel on commence à changer des codons\n",
    "            start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "            pos_of_prom_in_frame = pos-(pos_uppercase_in_ref_gene%3)\n",
    "            rest_of_81bp = prom[pos_of_prom_in_frame+3:]\n",
    "            window = prom[start_codon:start_codon+3]\n",
    "            n = 0\n",
    "            for next_codon in rest_of_81bp:\n",
    "                if any(t.isupper() for t in rest_of_81bp[n:n+3]):\n",
    "                    #extend the portion of prom to change:\n",
    "                    window = window+rest_of_81bp[n:n+3]\n",
    "                    n+=3\n",
    "                else:\n",
    "                    break\n",
    "            switch_codon = find_all_replacement_codons(window.upper(),rev_ecoli_codon_usage, start_codon)\n",
    "            list_switch_codon = [possibilities for position, possibilities in switch_codon.items()]\n",
    "            substrings_to_consider[start_codon,start_codon+n+3] = [''.join(i) for i in list(product(*list_switch_codon))]\n",
    "            traceback = start_codon+n+3\n",
    "        else:\n",
    "            if boolean:\n",
    "                #save the position to append correctly in substrings_to_consider\n",
    "                pos_1st_lower_case = pos\n",
    "                boolean = False\n",
    "            if pos < len(prom)-1:\n",
    "                if prom[pos+1].isupper():\n",
    "                    #il faut alors savoir ou est le prochain cadre de lecture:\n",
    "                    pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos+1\n",
    "                    start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "                    substrings_to_consider[pos_1st_lower_case,start_codon] = [prom[pos_1st_lower_case:start_codon].upper()]\n",
    "                    boolean = True\n",
    "            else:\n",
    "                substrings_to_consider[pos_1st_lower_case,pos+1] = [prom[pos_1st_lower_case:pos+1].upper()]\n",
    "    \n",
    "    \n",
    "print(substrings_to_consider)\n",
    "list_of_all_possibilities = [all_substring for position, all_substring in substrings_to_consider.items()]\n",
    "all_possible_promoter_IIprime = [''.join(i) for i in list(product(*list_of_all_possibilities))]\n",
    "\n",
    "#if no NOOOOOO, meaning all sequences encode for the same amino acid. \n",
    "for i in all_possible_promoter_IIprime:\n",
    "    if not check_if_same_prot(ref_gene[:start_of_prom_in_ref_gene]+i+ref_gene[start_of_prom_in_ref_gene+len(i):],ref_gene):\n",
    "        print(\"NOOOOOOOOOOOOOO\")\n",
    "        \n",
    "print(len(all_possible_promoter_IIprime))\n",
    "#Select random element in a list to get a much smaller input: \n",
    "if len(all_possible_promoter_IIprime) > 10000:\n",
    "    random.seed(i)\n",
    "    reduced_all_possible_promoter_IIprime = np.random.choice(all_possible_promoter_IIprime, 10000, replace=False)\n",
    "\n",
    "with open(\"./altered_proms/promII_prime/promII_prime.fa\",\"w\") as fout:\n",
    "    string = \"\"\n",
    "    for i,seq_prom in enumerate(all_possible_promoter_IIprime):\n",
    "        string+=f\">prom_II_prime_{i}\\n{seq_prom}\\n\"\n",
    "    fout.write(string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{}\n",
      "je suis le plus petit score de ipro70 et un non-promoteur pour CNNprom: ('prom_II_prime_4607', ['TTCTCACCCTTTTGAGTCGTTACCGACGCACTACTCAGGCATTGCATTCAAGATCTATGAGGGTTCTAAAAATTTTTATCC', 0.1415])\n"
     ]
    }
   ],
   "source": [
    "#make calculations for II' and get URL on web server:\n",
    "#1st prom2L:\n",
    "n=0\n",
    "iProm2L_results = {}\n",
    "for i,line in enumerate(urllib2.urlopen(\"http://bioinformatics.hitsz.edu.cn/iPromoter-2L/static/result/162.217.60.184_1629318730.04.txt\")):\n",
    "    if \"The length of the query sequence\" in str(line):\n",
    "        prom_name = \"prom\"+str(line.decode(\"utf-8\")).split(\"prom\")[1].split(\" )\")[0]\n",
    "    elif \"non-promoter sequence\" in str(line):\n",
    "        n=1\n",
    "    elif n == 1:\n",
    "        n=0\n",
    "        iProm2L_results[prom_name] = str(line).split(\":\")[1]\n",
    "print(iProm2L_results)\n",
    "\n",
    "#ipro70 analysis:\n",
    "with open('./altered_proms/promII_prime/ipro70_results', 'r') as fin:\n",
    "    ipro70_results = {}\n",
    "    for i, line in enumerate(fin):\n",
    "        if i%3 == 0:\n",
    "            prom_name = line.replace(\" \",\"_\").strip(\"\\n\")\n",
    "        if i%3 == 1:\n",
    "            prom_seq = line\n",
    "        if i%3 == 2 and \"Non-promoter\" in line:\n",
    "            ipro70_results[prom_name] = [prom_seq.strip(\"\\n\"),float(line.split(\"score: \")[1].strip(\"\\n\"))]\n",
    "            \n",
    "#CNN prom analysis            \n",
    "with open('./altered_proms/promII_prime/CNN_prom_results', 'r') as fin:\n",
    "    CNN_prom = []\n",
    "    for i, line in enumerate(fin):\n",
    "        if i%2 == 0:\n",
    "            prom_name = line[1:].replace(\" \",\"_\").strip(\"\\n\")\n",
    "        if i%2 == 1 and \"Not a Promoter\" in line:\n",
    "            CNN_prom.append(prom_name)\n",
    "\n",
    "common_proms_CNN_ipro70 = {prom_name:prom_seq for prom_name, prom_seq in ipro70_results.items() if prom_name in CNN_prom}\n",
    "prom_seq_lowscore = min(common_proms_CNN_ipro70.items(), key=lambda x: x[1][1])\n",
    "print(\"je suis le plus petit score de ipro70 et un non-promoteur pour CNNprom:\", prom_seq_lowscore)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{(0, 15): ['ACCTATGATTTTGAT'], (15, 21): ['AAATTT', 'AAATTC', 'AAGTTT', 'AAGTTC'], (21, 24): ['ACT'], (24, 30): ['ATAGAT', 'ATAGAC', 'ATTGAT', 'ATTGAC', 'ATCGAT', 'ATCGAC'], (30, 48): ['TCTTCTCAGCGTCTTAAT'], (48, 51): ['TTA', 'TTG', 'CTA', 'CTT', 'CTC', 'CTG'], (51, 54): ['AGC'], (54, 57): ['TAT', 'TAC'], (57, 60): ['CGC'], (60, 63): ['TAT', 'TAC'], (63, 81): ['GTTTTCAAGGATTCTAAG']}\n",
      "576\n"
     ]
    }
   ],
   "source": [
    "#prom IV:\n",
    "\n",
    "prom = \"acctatgattttgatAAAtTcactaTTGACtcttctcagcgtcttaatcTAagcTatcgctATgttttcaaggattctaag\"\n",
    "\n",
    "assert len(prom) == 81, \"normalement, la longueur de la séquence promotrice devrait être de 81 pb exactement\"\n",
    "assert not any_upper_case(prom[0:3]),\"pas de lettre majuscule dans les trois premiers nucleotides, sinon ca plante...\"\n",
    "ref_gene = all_cds_m13[\"I\"].lower()\n",
    "\n",
    "start_of_prom_in_ref_gene = ref_gene.find(prom.lower())\n",
    "substrings_to_consider = {}\n",
    "traceback = 0\n",
    "boolean = True\n",
    "promseq_in_refgene = ref_gene[start_of_prom_in_ref_gene:start_of_prom_in_ref_gene+len(prom)]\n",
    "\n",
    "for pos, nt in enumerate(prom):\n",
    "    #set the seed:\n",
    "    if pos >= traceback:\n",
    "        if nt.isupper():\n",
    "            #est-ce qu'il y a une série de codons contenant une lettre majuscule?:\n",
    "            pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos\n",
    "            #start codon: le codon in frame (in prom) duquel on commence à changer des codons\n",
    "            start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "            pos_of_prom_in_frame = pos-(pos_uppercase_in_ref_gene%3)\n",
    "            rest_of_81bp = prom[pos_of_prom_in_frame+3:]\n",
    "            window = prom[start_codon:start_codon+3]\n",
    "            n = 0\n",
    "            for next_codon in rest_of_81bp:\n",
    "                if any(t.isupper() for t in rest_of_81bp[n:n+3]):\n",
    "                    #extend the portion of prom to change:\n",
    "                    window = window+rest_of_81bp[n:n+3]\n",
    "                    n+=3\n",
    "                else:\n",
    "                    break\n",
    "            switch_codon = find_all_replacement_codons(window.upper(),rev_ecoli_codon_usage, start_codon)\n",
    "            list_switch_codon = [possibilities for position, possibilities in switch_codon.items()]\n",
    "            substrings_to_consider[start_codon,start_codon+n+3] = [''.join(i) for i in list(product(*list_switch_codon))]\n",
    "            traceback = start_codon+n+3\n",
    "        else:\n",
    "            if boolean:\n",
    "                #save the position to append correctly in substrings_to_consider\n",
    "                pos_1st_lower_case = pos\n",
    "                boolean = False\n",
    "            if pos < len(prom)-1:\n",
    "                if prom[pos+1].isupper():\n",
    "                    #il faut alors savoir ou est le prochain cadre de lecture:\n",
    "                    pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos+1\n",
    "                    start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "                    substrings_to_consider[pos_1st_lower_case,start_codon] = [prom[pos_1st_lower_case:start_codon].upper()]\n",
    "                    boolean = True\n",
    "            else:\n",
    "                substrings_to_consider[pos_1st_lower_case,pos+1] = [prom[pos_1st_lower_case:pos+1].upper()]\n",
    "    \n",
    "    \n",
    "print(substrings_to_consider)\n",
    "list_of_all_possibilities = [all_substring for position, all_substring in substrings_to_consider.items()]\n",
    "all_possible_promoter_IV = [''.join(i) for i in list(product(*list_of_all_possibilities))]\n",
    "\n",
    "#if no NOOOOOO, meaning all sequences encode for the same amino acid. \n",
    "for i in all_possible_promoter_IV:\n",
    "    if not check_if_same_prot(ref_gene[:start_of_prom_in_ref_gene]+i+ref_gene[start_of_prom_in_ref_gene+len(i):],ref_gene):\n",
    "        print(\"NOOOOOOOOOOOOOO\")\n",
    "        \n",
    "print(len(all_possible_promoter_IV))\n",
    "#Select random element in a list to get a much smaller input: \n",
    "if len(all_possible_promoter_IV) > 10000:\n",
    "    random.seed(i)\n",
    "    all_possible_promoter_IV = np.random.choice(all_possible_promoter_IV, 10000, replace=False)  \n",
    "    \n",
    "with open(\"./altered_proms/promIV/promIV_prime.fa\",\"w\") as fout:\n",
    "    string = \"\"\n",
    "    for i,seq_prom in enumerate(all_possible_promoter_IV):\n",
    "        string+=f\">prom_IV_{i}\\n{seq_prom}\\n\"\n",
    "    fout.write(string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'prom_IV_259': 'ACCTATGATTTTGATAAATTCACTATCGATTCTTCTCAGCGTCTTAATCTCAGCTACCGCTACGTTTTCAAGGATTCTAAG', 'prom_IV_262': 'ACCTATGATTTTGATAAATTCACTATCGATTCTTCTCAGCGTCTTAATCTGAGCTACCGCTATGTTTTCAAGGATTCTAAG', 'prom_IV_263': 'ACCTATGATTTTGATAAATTCACTATCGATTCTTCTCAGCGTCTTAATCTGAGCTACCGCTACGTTTTCAAGGATTCTAAG', 'prom_IV_407': 'ACCTATGATTTTGATAAGTTTACTATCGATTCTTCTCAGCGTCTTAATCTGAGCTACCGCTACGTTTTCAAGGATTCTAAG', 'prom_IV_454': 'ACCTATGATTTTGATAAGTTCACTATAGATTCTTCTCAGCGTCTTAATCTGAGCTACCGCTATGTTTTCAAGGATTCTAAG', 'prom_IV_455': 'ACCTATGATTTTGATAAGTTCACTATAGATTCTTCTCAGCGTCTTAATCTGAGCTACCGCTACGTTTTCAAGGATTCTAAG', 'prom_IV_534': 'ACCTATGATTTTGATAAGTTCACTATCGATTCTTCTCAGCGTCTTAATTTGAGCTACCGCTATGTTTTCAAGGATTCTAAG', 'prom_IV_535': 'ACCTATGATTTTGATAAGTTCACTATCGATTCTTCTCAGCGTCTTAATTTGAGCTACCGCTACGTTTTCAAGGATTCTAAG', 'prom_IV_546': 'ACCTATGATTTTGATAAGTTCACTATCGATTCTTCTCAGCGTCTTAATCTCAGCTACCGCTATGTTTTCAAGGATTCTAAG', 'prom_IV_547': 'ACCTATGATTTTGATAAGTTCACTATCGATTCTTCTCAGCGTCTTAATCTCAGCTACCGCTACGTTTTCAAGGATTCTAAG', 'prom_IV_548': 'ACCTATGATTTTGATAAGTTCACTATCGATTCTTCTCAGCGTCTTAATCTGAGCTATCGCTATGTTTTCAAGGATTCTAAG', 'prom_IV_549': 'ACCTATGATTTTGATAAGTTCACTATCGATTCTTCTCAGCGTCTTAATCTGAGCTATCGCTACGTTTTCAAGGATTCTAAG', 'prom_IV_550': 'ACCTATGATTTTGATAAGTTCACTATCGATTCTTCTCAGCGTCTTAATCTGAGCTACCGCTATGTTTTCAAGGATTCTAAG', 'prom_IV_551': 'ACCTATGATTTTGATAAGTTCACTATCGATTCTTCTCAGCGTCTTAATCTGAGCTACCGCTACGTTTTCAAGGATTCTAAG', 'prom_IV_571': 'ACCTATGATTTTGATAAGTTCACTATCGACTCTTCTCAGCGTCTTAATCTCAGCTACCGCTACGTTTTCAAGGATTCTAAG', 'prom_IV_574': 'ACCTATGATTTTGATAAGTTCACTATCGACTCTTCTCAGCGTCTTAATCTGAGCTACCGCTATGTTTTCAAGGATTCTAAG', 'prom_IV_575': 'ACCTATGATTTTGATAAGTTCACTATCGACTCTTCTCAGCGTCTTAATCTGAGCTACCGCTACGTTTTCAAGGATTCTAAG'}\n",
      "93\n",
      "55\n",
      "('prom_IV_259', 'ACCTATGATTTTGATAAATTCACTATCGATTCTTCTCAGCGTCTTAATCTCAGCTACCGCTACGTTTTCAAGGATTCTAAG')\n"
     ]
    }
   ],
   "source": [
    "#make calculations for IV and get URL on web server:\n",
    "#1st prom2L:\n",
    "n=0\n",
    "iProm2L_results = {}\n",
    "for i,line in enumerate(urllib2.urlopen(\"http://bioinformatics.hitsz.edu.cn/iPromoter-2L/static/result/162.217.60.184_1629388529.22.txt\")):\n",
    "    if \"The length of the query sequence\" in str(line):\n",
    "        prom_name = \"prom\"+str(line.decode(\"utf-8\")).split(\"prom\")[1].split(\" )\")[0]\n",
    "    elif \"non-promoter sequence\" in str(line):\n",
    "        n=1\n",
    "    elif n == 1:\n",
    "        n=0\n",
    "        iProm2L_results[prom_name] = str(line).split(\":\")[1][:-3]\n",
    "print(iProm2L_results)\n",
    "\n",
    "#ipro70 analysis:\n",
    "with open('./altered_proms/promIV/ipro70_results', 'r') as fin:\n",
    "    ipro70_results = {}\n",
    "    for i, line in enumerate(fin):\n",
    "        if i%3 == 0:\n",
    "            prom_name = line.replace(\" \",\"_\").strip(\"\\n\")\n",
    "        if i%3 == 1:\n",
    "            prom_seq = line\n",
    "        if i%3 == 2 and \"Non-promoter\" in line:\n",
    "            ipro70_results[prom_name] = [prom_seq.strip(\"\\n\"),float(line.split(\"score: \")[1].strip(\"\\n\"))]\n",
    "print(len(ipro70_results))\n",
    "\n",
    "#CNN prom analysis            \n",
    "with open('./altered_proms/promIV/CNN_prom_results', 'r') as fin:\n",
    "    CNN_prom = []\n",
    "    for i, line in enumerate(fin):\n",
    "        if i%2 == 0:\n",
    "            prom_name = line[1:].replace(\" \",\"_\").strip(\"\\n\")\n",
    "        if i%2 == 1 and \"Not a Promoter\" in line:\n",
    "            CNN_prom.append(prom_name)\n",
    "print(len(CNN_prom))\n",
    "\n",
    "common_proms_CNN_ipro70 = {prom_name:prom_seq for prom_name, prom_seq in iProm2L_results.items() if prom_name in CNN_prom}\n",
    "prom_seq_lowscore = min(common_proms_CNN_ipro70.items(), key=lambda x: x[1][1])\n",
    "\n",
    "print(prom_seq_lowscore)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{(0, 21): ['GTTCCGTTAGTTCGTTTTATT'], (21, 24): ['AAT', 'AAC'], (24, 30): ['GTAGAT'], (30, 36): ['TTTAGT', 'TTTAGC', 'TTTTCA', 'TTTTCT', 'TTTTCC', 'TTTTCG', 'TTCAGT', 'TTCAGC', 'TTCTCA', 'TTCTCT', 'TTCTCC', 'TTCTCG'], (36, 51): ['TCCCAACGTCCTGAC'], (51, 60): ['TGGTATAAT', 'TGGTATAAC', 'TGGTACAAT', 'TGGTACAAC'], (60, 63): ['GAG'], (63, 66): ['CCA', 'CCT', 'CCC', 'CCG'], (66, 81): ['GTTCTTAAAATCGCA']}\n",
      "384\n"
     ]
    }
   ],
   "source": [
    "#gene V promoter:\n",
    "#RBS: its rbs is located imediatly after, but should be completely fine:\n",
    "#two CDSs overlap with this element. Keep that in mind while writing this code.\n",
    "#after checking, the gene are both in the same ORF.\n",
    "\n",
    "#for prom V:\n",
    "\n",
    "prom = \"gttccgttagttcgttttattAAcgtagatTTttCttcccaacgtcctgacTggTATAATgagcCAgttcttaaaatcgca\"       #gene V\n",
    "\n",
    "assert len(prom) == 81, \"normalement, la longueur de la séquence promotrice devrait être de 81 pb exactement\"\n",
    "assert not any_upper_case(prom[0:3]),\"pas de lettre majuscule dans les trois premiers nucleotides, sinon ca plante...\"\n",
    "ref_gene = all_cds_m13[\"II\"].lower()\n",
    "\n",
    "start_of_prom_in_ref_gene = ref_gene.find(prom.lower())\n",
    "substrings_to_consider = {}\n",
    "traceback = 0\n",
    "boolean = True\n",
    "promseq_in_refgene = ref_gene[start_of_prom_in_ref_gene:start_of_prom_in_ref_gene+len(prom)]\n",
    "\n",
    "for pos, nt in enumerate(prom):\n",
    "    #set the seed:\n",
    "    if pos >= traceback:\n",
    "        if nt.isupper():\n",
    "            #est-ce qu'il y a une série de codons contenant une lettre majuscule?:\n",
    "            pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos\n",
    "            #start codon: le codon in frame (in prom) duquel on commence à changer des codons\n",
    "            start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "            pos_of_prom_in_frame = pos-(pos_uppercase_in_ref_gene%3)\n",
    "            rest_of_81bp = prom[pos_of_prom_in_frame+3:]\n",
    "            window = prom[start_codon:start_codon+3]\n",
    "            n = 0\n",
    "            for next_codon in rest_of_81bp:\n",
    "                if any(t.isupper() for t in rest_of_81bp[n:n+3]):\n",
    "                    #extend the portion of prom to change:\n",
    "                    window = window+rest_of_81bp[n:n+3]\n",
    "                    n+=3\n",
    "                else:\n",
    "                    break\n",
    "            switch_codon = find_all_replacement_codons(window.upper(),rev_ecoli_codon_usage, start_codon)\n",
    "            list_switch_codon = [possibilities for position, possibilities in switch_codon.items()]\n",
    "            substrings_to_consider[start_codon,start_codon+n+3] = [''.join(i) for i in list(product(*list_switch_codon))]\n",
    "            traceback = start_codon+n+3\n",
    "        else:\n",
    "            if boolean:\n",
    "                #save the position to append correctly in substrings_to_consider\n",
    "                pos_1st_lower_case = pos\n",
    "                boolean = False\n",
    "            if pos < len(prom)-1:\n",
    "                if prom[pos+1].isupper():\n",
    "                    #il faut alors savoir ou est le prochain cadre de lecture:\n",
    "                    pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos+1\n",
    "                    start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "                    substrings_to_consider[pos_1st_lower_case,start_codon] = [prom[pos_1st_lower_case:start_codon].upper()]\n",
    "                    boolean = True\n",
    "            else:\n",
    "                substrings_to_consider[pos_1st_lower_case,pos+1] = [prom[pos_1st_lower_case:pos+1].upper()]\n",
    "    \n",
    "    \n",
    "print(substrings_to_consider)\n",
    "list_of_all_possibilities = [all_substring for position, all_substring in substrings_to_consider.items()]\n",
    "all_possible_promoter_V = [''.join(i) for i in list(product(*list_of_all_possibilities))]\n",
    "\n",
    "#if no NOOOOOO, meaning all sequences encode for the same amino acid. \n",
    "for i in all_possible_promoter_V:\n",
    "    if not check_if_same_prot(ref_gene[:start_of_prom_in_ref_gene]+i+ref_gene[start_of_prom_in_ref_gene+len(i):],ref_gene):\n",
    "        print(\"NOOOOOOOOOOOOOO\")\n",
    "        \n",
    "print(len(all_possible_promoter_V))\n",
    "#Select random element in a list to get a much smaller input: \n",
    "if len(all_possible_promoter_V) > 10000:\n",
    "    random.seed(i)\n",
    "    all_possible_promoter_V = np.random.choice(all_possible_promoter_V, 10000, replace=False)  \n",
    "    \n",
    "with open(\"./altered_proms/promV/promV_prime.fa\",\"w\") as fout:\n",
    "    string = \"\"\n",
    "    for i,seq_prom in enumerate(all_possible_promoter_V):\n",
    "        string+=f\">prom_V_{i}\\n{seq_prom}\\n\"\n",
    "    fout.write(string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'prom_V_46': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTTTCATCCCAACGTCCTGACTGGTACAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_90': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTTTCGTCCCAACGTCCTGACTGGTACAATGAGCCCGTTCTTAAAATCGCA', 'prom_V_91': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTTTCGTCCCAACGTCCTGACTGGTACAATGAGCCGGTTCTTAAAATCGCA', 'prom_V_93': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTTTCGTCCCAACGTCCTGACTGGTACAACGAGCCTGTTCTTAAAATCGCA', 'prom_V_94': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTTTCGTCCCAACGTCCTGACTGGTACAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_95': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTTTCGTCCCAACGTCCTGACTGGTACAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_104': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGTTCCCAACGTCCTGACTGGTACAATGAGCCAGTTCTTAAAATCGCA', 'prom_V_105': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGTTCCCAACGTCCTGACTGGTACAATGAGCCTGTTCTTAAAATCGCA', 'prom_V_106': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGTTCCCAACGTCCTGACTGGTACAATGAGCCCGTTCTTAAAATCGCA', 'prom_V_107': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGTTCCCAACGTCCTGACTGGTACAATGAGCCGGTTCTTAAAATCGCA', 'prom_V_108': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGTTCCCAACGTCCTGACTGGTACAACGAGCCAGTTCTTAAAATCGCA', 'prom_V_109': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGTTCCCAACGTCCTGACTGGTACAACGAGCCTGTTCTTAAAATCGCA', 'prom_V_110': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGTTCCCAACGTCCTGACTGGTACAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_111': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGTTCCCAACGTCCTGACTGGTACAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_114': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGCTCCCAACGTCCTGACTGGTATAATGAGCCCGTTCTTAAAATCGCA', 'prom_V_115': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGCTCCCAACGTCCTGACTGGTATAATGAGCCGGTTCTTAAAATCGCA', 'prom_V_118': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGCTCCCAACGTCCTGACTGGTATAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_119': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGCTCCCAACGTCCTGACTGGTATAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_120': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAATGAGCCAGTTCTTAAAATCGCA', 'prom_V_121': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAATGAGCCTGTTCTTAAAATCGCA', 'prom_V_122': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAATGAGCCCGTTCTTAAAATCGCA', 'prom_V_123': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAATGAGCCGGTTCTTAAAATCGCA', 'prom_V_124': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAACGAGCCAGTTCTTAAAATCGCA', 'prom_V_125': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAACGAGCCTGTTCTTAAAATCGCA', 'prom_V_126': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_127': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_136': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCATCCCAACGTCCTGACTGGTACAATGAGCCAGTTCTTAAAATCGCA', 'prom_V_137': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCATCCCAACGTCCTGACTGGTACAATGAGCCTGTTCTTAAAATCGCA', 'prom_V_138': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCATCCCAACGTCCTGACTGGTACAATGAGCCCGTTCTTAAAATCGCA', 'prom_V_139': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCATCCCAACGTCCTGACTGGTACAATGAGCCGGTTCTTAAAATCGCA', 'prom_V_140': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCATCCCAACGTCCTGACTGGTACAACGAGCCAGTTCTTAAAATCGCA', 'prom_V_141': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCATCCCAACGTCCTGACTGGTACAACGAGCCTGTTCTTAAAATCGCA', 'prom_V_142': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCATCCCAACGTCCTGACTGGTACAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_143': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCATCCCAACGTCCTGACTGGTACAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_158': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCTTCCCAACGTCCTGACTGGTACAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_159': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCTTCCCAACGTCCTGACTGGTACAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_170': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCCTCCCAACGTCCTGACTGGTACAATGAGCCCGTTCTTAAAATCGCA', 'prom_V_171': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCCTCCCAACGTCCTGACTGGTACAATGAGCCGGTTCTTAAAATCGCA', 'prom_V_174': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCCTCCCAACGTCCTGACTGGTACAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_175': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCCTCCCAACGTCCTGACTGGTACAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_178': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCGTCCCAACGTCCTGACTGGTATAATGAGCCCGTTCTTAAAATCGCA', 'prom_V_179': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCGTCCCAACGTCCTGACTGGTATAATGAGCCGGTTCTTAAAATCGCA', 'prom_V_182': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCGTCCCAACGTCCTGACTGGTATAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_183': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCGTCCCAACGTCCTGACTGGTATAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_184': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCGTCCCAACGTCCTGACTGGTACAATGAGCCAGTTCTTAAAATCGCA', 'prom_V_185': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCGTCCCAACGTCCTGACTGGTACAATGAGCCTGTTCTTAAAATCGCA', 'prom_V_186': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCGTCCCAACGTCCTGACTGGTACAATGAGCCCGTTCTTAAAATCGCA', 'prom_V_187': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCGTCCCAACGTCCTGACTGGTACAATGAGCCGGTTCTTAAAATCGCA', 'prom_V_188': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCGTCCCAACGTCCTGACTGGTACAACGAGCCAGTTCTTAAAATCGCA', 'prom_V_189': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCGTCCCAACGTCCTGACTGGTACAACGAGCCTGTTCTTAAAATCGCA', 'prom_V_190': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCGTCCCAACGTCCTGACTGGTACAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_191': 'GTTCCGTTAGTTCGTTTTATTAATGTAGATTTCTCGTCCCAACGTCCTGACTGGTACAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_238': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTTTCATCCCAACGTCCTGACTGGTACAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_239': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTTTCATCCCAACGTCCTGACTGGTACAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_282': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTTTCGTCCCAACGTCCTGACTGGTACAATGAGCCCGTTCTTAAAATCGCA', 'prom_V_283': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTTTCGTCCCAACGTCCTGACTGGTACAATGAGCCGGTTCTTAAAATCGCA', 'prom_V_284': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTTTCGTCCCAACGTCCTGACTGGTACAACGAGCCAGTTCTTAAAATCGCA', 'prom_V_285': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTTTCGTCCCAACGTCCTGACTGGTACAACGAGCCTGTTCTTAAAATCGCA', 'prom_V_286': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTTTCGTCCCAACGTCCTGACTGGTACAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_287': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTTTCGTCCCAACGTCCTGACTGGTACAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_294': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGTTCCCAACGTCCTGACTGGTATAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_295': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGTTCCCAACGTCCTGACTGGTATAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_296': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGTTCCCAACGTCCTGACTGGTACAATGAGCCAGTTCTTAAAATCGCA', 'prom_V_297': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGTTCCCAACGTCCTGACTGGTACAATGAGCCTGTTCTTAAAATCGCA', 'prom_V_298': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGTTCCCAACGTCCTGACTGGTACAATGAGCCCGTTCTTAAAATCGCA', 'prom_V_299': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGTTCCCAACGTCCTGACTGGTACAATGAGCCGGTTCTTAAAATCGCA', 'prom_V_300': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGTTCCCAACGTCCTGACTGGTACAACGAGCCAGTTCTTAAAATCGCA', 'prom_V_301': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGTTCCCAACGTCCTGACTGGTACAACGAGCCTGTTCTTAAAATCGCA', 'prom_V_302': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGTTCCCAACGTCCTGACTGGTACAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_303': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGTTCCCAACGTCCTGACTGGTACAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_306': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGCTCCCAACGTCCTGACTGGTATAATGAGCCCGTTCTTAAAATCGCA', 'prom_V_307': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGCTCCCAACGTCCTGACTGGTATAATGAGCCGGTTCTTAAAATCGCA', 'prom_V_310': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGCTCCCAACGTCCTGACTGGTATAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_311': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGCTCCCAACGTCCTGACTGGTATAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_312': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAATGAGCCAGTTCTTAAAATCGCA', 'prom_V_313': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAATGAGCCTGTTCTTAAAATCGCA', 'prom_V_314': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAATGAGCCCGTTCTTAAAATCGCA', 'prom_V_315': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAATGAGCCGGTTCTTAAAATCGCA', 'prom_V_316': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAACGAGCCAGTTCTTAAAATCGCA', 'prom_V_317': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAACGAGCCTGTTCTTAAAATCGCA', 'prom_V_318': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_319': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_328': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCATCCCAACGTCCTGACTGGTACAATGAGCCAGTTCTTAAAATCGCA', 'prom_V_329': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCATCCCAACGTCCTGACTGGTACAATGAGCCTGTTCTTAAAATCGCA', 'prom_V_330': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCATCCCAACGTCCTGACTGGTACAATGAGCCCGTTCTTAAAATCGCA', 'prom_V_331': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCATCCCAACGTCCTGACTGGTACAATGAGCCGGTTCTTAAAATCGCA', 'prom_V_332': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCATCCCAACGTCCTGACTGGTACAACGAGCCAGTTCTTAAAATCGCA', 'prom_V_333': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCATCCCAACGTCCTGACTGGTACAACGAGCCTGTTCTTAAAATCGCA', 'prom_V_334': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCATCCCAACGTCCTGACTGGTACAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_335': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCATCCCAACGTCCTGACTGGTACAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_346': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCTTCCCAACGTCCTGACTGGTACAATGAGCCCGTTCTTAAAATCGCA', 'prom_V_347': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCTTCCCAACGTCCTGACTGGTACAATGAGCCGGTTCTTAAAATCGCA', 'prom_V_350': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCTTCCCAACGTCCTGACTGGTACAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_351': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCTTCCCAACGTCCTGACTGGTACAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_362': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCCTCCCAACGTCCTGACTGGTACAATGAGCCCGTTCTTAAAATCGCA', 'prom_V_363': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCCTCCCAACGTCCTGACTGGTACAATGAGCCGGTTCTTAAAATCGCA', 'prom_V_366': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCCTCCCAACGTCCTGACTGGTACAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_367': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCCTCCCAACGTCCTGACTGGTACAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_370': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCGTCCCAACGTCCTGACTGGTATAATGAGCCCGTTCTTAAAATCGCA', 'prom_V_371': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCGTCCCAACGTCCTGACTGGTATAATGAGCCGGTTCTTAAAATCGCA', 'prom_V_374': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCGTCCCAACGTCCTGACTGGTATAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_375': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCGTCCCAACGTCCTGACTGGTATAACGAGCCGGTTCTTAAAATCGCA', 'prom_V_376': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCGTCCCAACGTCCTGACTGGTACAATGAGCCAGTTCTTAAAATCGCA', 'prom_V_377': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCGTCCCAACGTCCTGACTGGTACAATGAGCCTGTTCTTAAAATCGCA', 'prom_V_378': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCGTCCCAACGTCCTGACTGGTACAATGAGCCCGTTCTTAAAATCGCA', 'prom_V_379': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCGTCCCAACGTCCTGACTGGTACAATGAGCCGGTTCTTAAAATCGCA', 'prom_V_380': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCGTCCCAACGTCCTGACTGGTACAACGAGCCAGTTCTTAAAATCGCA', 'prom_V_381': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCGTCCCAACGTCCTGACTGGTACAACGAGCCTGTTCTTAAAATCGCA', 'prom_V_382': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCGTCCCAACGTCCTGACTGGTACAACGAGCCCGTTCTTAAAATCGCA', 'prom_V_383': 'GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCTCGTCCCAACGTCCTGACTGGTACAACGAGCCGGTTCTTAAAATCGCA'}\n",
      "384\n",
      "235\n",
      "('prom_V_314', ['GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAATGAGCCCGTTCTTAAAATCGCA', 0.0154])\n"
     ]
    }
   ],
   "source": [
    "#make calculations for V and get URL on web server:\n",
    "#1st prom2L:\n",
    "n=0\n",
    "iProm2L_results = {}\n",
    "for i,line in enumerate(urllib2.urlopen(\"http://bioinformatics.hitsz.edu.cn/iPromoter-2L/static/result/162.217.60.184_1629399021.47.txt\")):\n",
    "    if \"The length of the query sequence\" in str(line):\n",
    "        prom_name = \"prom\"+str(line.decode(\"utf-8\")).split(\"prom\")[1].split(\" )\")[0]\n",
    "    elif \"non-promoter sequence\" in str(line):\n",
    "        n=1\n",
    "    elif n == 1:\n",
    "        n=0\n",
    "        iProm2L_results[prom_name] = str(line).split(\":\")[1][:-3]\n",
    "print(iProm2L_results)\n",
    "\n",
    "#ipro70 analysis:\n",
    "with open('./altered_proms/promV/ipro70_results', 'r') as fin:\n",
    "    ipro70_results = {}\n",
    "    for i, line in enumerate(fin):\n",
    "        if i%3 == 0:\n",
    "            prom_name = line.replace(\" \",\"_\").strip(\"\\n\")\n",
    "        if i%3 == 1:\n",
    "            prom_seq = line\n",
    "        if i%3 == 2 and \"Non-promoter\" in line:\n",
    "            ipro70_results[prom_name] = [prom_seq.strip(\"\\n\"),float(line.split(\"score: \")[1].strip(\"\\n\"))]\n",
    "print(len(ipro70_results))\n",
    "\n",
    "#CNN prom analysis            \n",
    "with open('./altered_proms/promV/CNN_prom_results', 'r') as fin:\n",
    "    CNN_prom = []\n",
    "    for i, line in enumerate(fin):\n",
    "        if i%2 == 0:\n",
    "            prom_name = line[1:].replace(\" \",\"_\").strip(\"\\n\")\n",
    "        if i%2 == 1 and \"Not a Promoter\" in line:\n",
    "            CNN_prom.append(prom_name)\n",
    "print(len(CNN_prom))\n",
    "\n",
    "common_proms_CNN_ipro70 = {prom_name:prom_seq for prom_name, prom_seq in ipro70_results.items() if prom_name in CNN_prom and prom_name in iProm2L_results.keys()}\n",
    "prom_seq_lowscore = min(common_proms_CNN_ipro70.items(), key=lambda x: x[1][1])\n",
    "\n",
    "print(prom_seq_lowscore)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pos_1st_lower_case 34\n",
      "{(0, 7): ['TAAACCG'], 8: ['A', 'C', 'G'], 9: ['T', 'C', 'G'], (9, 10): ['A'], 11: ['T', 'C', 'G'], 12: ['T', 'C', 'G'], 13: ['A', 'C', 'G'], (13, 19): ['TTAAAG'], 20: ['A', 'T', 'G'], (20, 37): ['CTCCTTTTGGAGCCTTT']}\n",
      "Voici le nombre de promoteurs générés:  139968\n"
     ]
    }
   ],
   "source": [
    "#prom III:\n",
    "#attention, la portion 5' de cette sequence se retrouve dans une portion terminatrice \n",
    "#cas d'exception puisque cette région ne chevauche pas complètement un gène. \n",
    "#note: Le potentiel +1 et la base d'avant (CT) se trouvent dans un terminateur et n'ont donc pas été changé.\n",
    "#note: comme c'est en ce moment, le +1 n'est pas muté.\n",
    "\n",
    "prom = \"cggtatcaagctgtttaagAAaTtcacCTcGAaAgcaagctga\"\n",
    "rest_prom = \"taaaccgaTAcAATtaaaggCtccttttggagcctttt\"\n",
    "\n",
    "assert (len(prom)+len(rest_prom)) == 81, \"normalement, la longueur de la séquence promotrice devrait être de 81 pb exactement\"\n",
    "assert not any_upper_case(prom[0:3]),\"pas de lettre majuscule dans les trois premiers nucleotides, sinon ca plante...\"\n",
    "ref_gene = all_cds_m13[\"VIII\"].lower()\n",
    "\n",
    "start_of_prom_in_ref_gene = ref_gene.find(prom.lower())\n",
    "substrings_to_consider = {}\n",
    "traceback = 0\n",
    "boolean = True\n",
    "for pos, nt in enumerate(prom):\n",
    "    #set the seed:\n",
    "    random.seed(pos)\n",
    "    promseq_in_refgene = ref_gene[start_of_prom_in_ref_gene:start_of_prom_in_ref_gene+len(prom)]\n",
    "    if pos >= traceback:\n",
    "        if nt.isupper():\n",
    "            #est-ce qu'il y a une série de codons contenant une lettre majuscule?:\n",
    "            pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos\n",
    "            #start codon: le codon in frame (in prom) duquel on commence à changer des codons\n",
    "            start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "            pos_of_prom_in_frame = pos-(pos_uppercase_in_ref_gene%3)\n",
    "            rest_of_81bp = prom[pos_of_prom_in_frame+3:]\n",
    "            window = prom[start_codon:start_codon+3]\n",
    "            n = 0\n",
    "            for next_codon in rest_of_81bp:\n",
    "                if any(t.isupper() for t in rest_of_81bp[n:n+3]):\n",
    "                    #extend the portion of prom to change:\n",
    "                    window = window+rest_of_81bp[n:n+3]\n",
    "                    n+=3\n",
    "                else:\n",
    "                    break\n",
    "            switch_codon = find_all_replacement_codons(window.upper(),rev_ecoli_codon_usage, start_codon)\n",
    "            list_switch_codon = [possibilities for position, possibilities in switch_codon.items()]\n",
    "            substrings_to_consider[start_codon,start_codon+n+3] = [''.join(i) for i in list(product(*list_switch_codon))]\n",
    "            traceback = start_codon+n+3\n",
    "        else:\n",
    "            if boolean:\n",
    "                pos_1st_lower_case = pos\n",
    "                boolean = False\n",
    "            if pos < len(prom)-1:\n",
    "                if prom[pos+1].isupper():\n",
    "                    #il faut alors savoir ou est le prochain cadre de lecture:\n",
    "                    pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos+1\n",
    "                    start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "                    substrings_to_consider[pos_1st_lower_case,start_codon] = [prom[pos_1st_lower_case:start_codon].upper()]\n",
    "                    boolean = True\n",
    "            else:\n",
    "                print(\"pos_1st_lower_case\", pos_1st_lower_case)\n",
    "                substrings_to_consider[pos_1st_lower_case,pos+1] = [prom[pos_1st_lower_case:pos+1].upper()]\n",
    "                \n",
    "list_of_all_possibilities = [all_substring for position, all_substring in substrings_to_consider.items()]\n",
    "all_possible_promoter_preIII = [''.join(i) for i in list(product(*list_of_all_possibilities))]\n",
    "\n",
    "for i in all_possible_promoter_preIII:\n",
    "    if not check_if_same_prot(ref_gene[:start_of_prom_in_ref_gene]+i+ref_gene[start_of_prom_in_ref_gene+len(i):],ref_gene):\n",
    "        print(\"NOOOOOOOOOOOOOO\")\n",
    "\n",
    "#After that, get all possible promoter for rest_prom not in gene VIII:\n",
    "\n",
    "switch_codon = {}\n",
    "substrings_to_consider = {}\n",
    "nts_list = [\"A\",\"T\",\"C\",\"G\"]\n",
    "prev_number = 0\n",
    "all_possibilities = []\n",
    "for i, nt in enumerate(rest_prom):\n",
    "    if nt.isupper():\n",
    "        nts_to_consider = [n for n in nts_list if n != nt]\n",
    "        switch_codon[i] = nts_to_consider\n",
    "        prev_number = i\n",
    "    else:\n",
    "        if i < len(rest_prom)-1:\n",
    "            if rest_prom[i+1].isupper():\n",
    "                switch_codon[(prev_number,i)] = [rest_prom[prev_number:i].upper()]\n",
    "        else:\n",
    "            switch_codon[prev_number,i] = [rest_prom[prev_number:i].upper()]\n",
    "            \n",
    "print(switch_codon)\n",
    "list_of_all_possibilities = [all_substring for position, all_substring in switch_codon.items()]\n",
    "all_possible_promoter_rest = [''.join(i) for i in list(product(*list_of_all_possibilities))]\n",
    "all_possible_promoter_III = [''.join(i) for i in list(product(*[all_possible_promoter_preIII,all_possible_promoter_rest]))]\n",
    "\n",
    "print(\"Voici le nombre de promoteurs générés: \", len(all_possible_promoter_III))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{(0, 14): ['TTGTCTTTGGCGCT'], (14, 17): ['GGA', 'GGT', 'GGC', 'GGG'], (17, 20): ['AAA'], (20, 29): ['CCATATGAA', 'CCATATGAG', 'CCATACGAA', 'CCATACGAG', 'CCTTATGAA', 'CCTTATGAG', 'CCTTACGAA', 'CCTTACGAG', 'CCCTATGAA', 'CCCTATGAG', 'CCCTACGAA', 'CCCTACGAG', 'CCGTATGAA', 'CCGTATGAG', 'CCGTACGAA', 'CCGTACGAG'], (29, 41): ['TTTTCTATTGAT'], (41, 44): ['TGT', 'TGC'], (44, 47): ['GAC'], (47, 53): ['AAAATA', 'AAAATT', 'AAAATC', 'AAGATA', 'AAGATT', 'AAGATC'], (53, 56): ['AAC'], (56, 62): ['TTATTT', 'TTATTC', 'TTGTTT', 'TTGTTC', 'CTATTT', 'CTATTC', 'CTTTTT', 'CTTTTC', 'CTCTTT', 'CTCTTC', 'CTGTTT', 'CTGTTC'], (62, 81): ['CGTGGTGTCTTTGCGTTTC']}\n",
      "9216\n"
     ]
    }
   ],
   "source": [
    "# prom of gene VI\n",
    "#RBS relatively far away from its promoter.\n",
    "\n",
    "prom = \"ttgtctttggcgctggTaaacCaTatgAattttctattgattgTgacAaAATaaacttATtccgtggtgtctttgcgtttc\"\n",
    "\n",
    "assert len(prom) == 81, \"normalement, la longueur de la séquence promotrice devrait être de 81 pb exactement\"\n",
    "assert not any_upper_case(prom[0:3]),\"pas de lettre majuscule dans les trois premiers nucleotides, sinon ca plante...\"\n",
    "ref_gene = all_cds_m13[\"III\"].lower()\n",
    "\n",
    "start_of_prom_in_ref_gene = ref_gene.find(prom.lower())\n",
    "substrings_to_consider = {}\n",
    "traceback = 0\n",
    "boolean = True\n",
    "promseq_in_refgene = ref_gene[start_of_prom_in_ref_gene:start_of_prom_in_ref_gene+len(prom)]\n",
    "\n",
    "for pos, nt in enumerate(prom):\n",
    "    #set the seed:\n",
    "    if pos >= traceback:\n",
    "        if nt.isupper():\n",
    "            #est-ce qu'il y a une série de codons contenant une lettre majuscule?:\n",
    "            pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos\n",
    "            #start codon: le codon in frame (in prom) duquel on commence à changer des codons\n",
    "            start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "            pos_of_prom_in_frame = pos-(pos_uppercase_in_ref_gene%3)\n",
    "            rest_of_81bp = prom[pos_of_prom_in_frame+3:]\n",
    "            window = prom[start_codon:start_codon+3]\n",
    "            n = 0\n",
    "            for next_codon in rest_of_81bp:\n",
    "                if any(t.isupper() for t in rest_of_81bp[n:n+3]):\n",
    "                    #extend the portion of prom to change:\n",
    "                    window = window+rest_of_81bp[n:n+3]\n",
    "                    n+=3\n",
    "                else:\n",
    "                    break\n",
    "            switch_codon = find_all_replacement_codons(window.upper(),rev_ecoli_codon_usage, start_codon)\n",
    "            list_switch_codon = [possibilities for position, possibilities in switch_codon.items()]\n",
    "            substrings_to_consider[start_codon,start_codon+n+3] = [''.join(i) for i in list(product(*list_switch_codon))]\n",
    "            traceback = start_codon+n+3\n",
    "        else:\n",
    "            if boolean:\n",
    "                #save the position to append correctly in substrings_to_consider\n",
    "                pos_1st_lower_case = pos\n",
    "                boolean = False\n",
    "            if pos < len(prom)-1:\n",
    "                if prom[pos+1].isupper():\n",
    "                    #il faut alors savoir ou est le prochain cadre de lecture:\n",
    "                    pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos+1\n",
    "                    start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "                    substrings_to_consider[pos_1st_lower_case,start_codon] = [prom[pos_1st_lower_case:start_codon].upper()]\n",
    "                    boolean = True\n",
    "            else:\n",
    "                substrings_to_consider[pos_1st_lower_case,pos+1] = [prom[pos_1st_lower_case:pos+1].upper()]\n",
    "    \n",
    "    \n",
    "print(substrings_to_consider)\n",
    "list_of_all_possibilities = [all_substring for position, all_substring in substrings_to_consider.items()]\n",
    "all_possible_promoter_VI = [''.join(i) for i in list(product(*list_of_all_possibilities))]\n",
    "\n",
    "#if no NOOOOOO, meaning all sequences encode for the same amino acid. \n",
    "for i in all_possible_promoter_VI:\n",
    "    if not check_if_same_prot(ref_gene[:start_of_prom_in_ref_gene]+i+ref_gene[start_of_prom_in_ref_gene+len(i):],ref_gene):\n",
    "        print(\"NOOOOOOOOOOOOOO\")\n",
    "        \n",
    "print(len(all_possible_promoter_VI))\n",
    "#Select random element in a list to get a much smaller input: \n",
    "if len(all_possible_promoter_VI) > 10000:\n",
    "    random.seed(i)\n",
    "    all_possible_promoter_VI = np.random.choice(all_possible_promoter_VI, 10000, replace=False)  \n",
    "    \n",
    "with open(\"./altered_proms/promVI/promVI.fa\",\"w\") as fout:\n",
    "    string = \"\"\n",
    "    for i,seq_prom in enumerate(all_possible_promoter_VI):\n",
    "        string+=f\">prom_VI_{i}\\n{seq_prom}\\n\"\n",
    "    fout.write(string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4125\n",
      "3964\n",
      "2655\n",
      "('prom_VI_9140', ['TTGTCTTTGGCGCTGGGAAACCGTACGAGTTTTCTATTGATTGTGACAAGATCAACCTCTTTCGTGGTGTCTTTGCGTTTC', 0.0706])\n"
     ]
    }
   ],
   "source": [
    "#make calculations for VI and get URL on web server:\n",
    "#1st prom2L:\n",
    "n=0\n",
    "iProm2L_results = {}\n",
    "for i,line in enumerate(urllib2.urlopen(\"http://bioinformatics.hitsz.edu.cn/iPromoter-2L/static/result/162.217.60.184_1629400054.61.txt\")):\n",
    "    if \"The length of the query sequence\" in str(line):\n",
    "        prom_name = \"prom\"+str(line.decode(\"utf-8\")).split(\"prom\")[1].split(\" )\")[0]\n",
    "    elif \"non-promoter sequence\" in str(line):\n",
    "        n=1\n",
    "    elif n == 1:\n",
    "        n=0\n",
    "        iProm2L_results[prom_name] = str(line).split(\":\")[1][:-3]\n",
    "print(len(iProm2L_results))\n",
    "\n",
    "#ipro70 analysis:\n",
    "with open('./altered_proms/promVI/ipro70_results', 'r') as fin:\n",
    "    ipro70_results = {}\n",
    "    for i, line in enumerate(fin):\n",
    "        if i%3 == 0:\n",
    "            prom_name = line.replace(\" \",\"_\").strip(\"\\n\")\n",
    "        if i%3 == 1:\n",
    "            prom_seq = line\n",
    "        if i%3 == 2 and \"Non-promoter\" in line:\n",
    "            ipro70_results[prom_name] = [prom_seq.strip(\"\\n\"),float(line.split(\"score: \")[1].strip(\"\\n\"))]\n",
    "print(len(ipro70_results))\n",
    "\n",
    "#CNN prom analysis            \n",
    "with open('./altered_proms/promVI/CNN_results', 'r') as fin:\n",
    "    CNN_prom = []\n",
    "    for i, line in enumerate(fin):\n",
    "        if i%2 == 0:\n",
    "            prom_name = line[1:].replace(\" \",\"_\").strip(\"\\n\")\n",
    "        if i%2 == 1 and \"Not a Promoter\" in line:\n",
    "            CNN_prom.append(prom_name)\n",
    "print(len(CNN_prom))\n",
    "\n",
    "common_proms_CNN_ipro70 = {prom_name:prom_seq for prom_name, prom_seq in ipro70_results.items() if prom_name in CNN_prom and prom_name in iProm2L_results.keys()}\n",
    "prom_seq_lowscore = min(common_proms_CNN_ipro70.items(), key=lambda x: x[1][1])\n",
    "\n",
    "print(prom_seq_lowscore)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{(0, 23): ['TTGTTCAGGGTGTTCAGTTAATT'], (23, 26): ['TTA', 'TTG', 'CTA', 'CTT', 'CTC', 'CTG'], (26, 35): ['CCGTCTAAT'], (35, 38): ['GCA', 'GCT', 'GCC', 'GCG'], (38, 50): ['CTTCCCTGTTTT'], (50, 62): ['TATGTAATATTA', 'TATGTAATATTG', 'TATGTAATACTA', 'TATGTAATACTT', 'TATGTAATACTC', 'TATGTAATACTG', 'TATGTAATTTTA', 'TATGTAATTTTG', 'TATGTAATTCTA', 'TATGTAATTCTT', 'TATGTAATTCTC', 'TATGTAATTCTG', 'TATGTAATCTTA', 'TATGTAATCTTG', 'TATGTAATCCTA', 'TATGTAATCCTT', 'TATGTAATCCTC', 'TATGTAATCCTG', 'TATGTTATATTA', 'TATGTTATATTG', 'TATGTTATACTA', 'TATGTTATACTT', 'TATGTTATACTC', 'TATGTTATACTG', 'TATGTTATTTTA', 'TATGTTATTTTG', 'TATGTTATTCTA', 'TATGTTATTCTT', 'TATGTTATTCTC', 'TATGTTATTCTG', 'TATGTTATCTTA', 'TATGTTATCTTG', 'TATGTTATCCTA', 'TATGTTATCCTT', 'TATGTTATCCTC', 'TATGTTATCCTG', 'TATGTCATATTA', 'TATGTCATATTG', 'TATGTCATACTA', 'TATGTCATACTT', 'TATGTCATACTC', 'TATGTCATACTG', 'TATGTCATTTTA', 'TATGTCATTTTG', 'TATGTCATTCTA', 'TATGTCATTCTT', 'TATGTCATTCTC', 'TATGTCATTCTG', 'TATGTCATCTTA', 'TATGTCATCTTG', 'TATGTCATCCTA', 'TATGTCATCCTT', 'TATGTCATCCTC', 'TATGTCATCCTG', 'TATGTGATATTA', 'TATGTGATATTG', 'TATGTGATACTA', 'TATGTGATACTT', 'TATGTGATACTC', 'TATGTGATACTG', 'TATGTGATTTTA', 'TATGTGATTTTG', 'TATGTGATTCTA', 'TATGTGATTCTT', 'TATGTGATTCTC', 'TATGTGATTCTG', 'TATGTGATCTTA', 'TATGTGATCTTG', 'TATGTGATCCTA', 'TATGTGATCCTT', 'TATGTGATCCTC', 'TATGTGATCCTG', 'TACGTAATATTA', 'TACGTAATATTG', 'TACGTAATACTA', 'TACGTAATACTT', 'TACGTAATACTC', 'TACGTAATACTG', 'TACGTAATTTTA', 'TACGTAATTTTG', 'TACGTAATTCTA', 'TACGTAATTCTT', 'TACGTAATTCTC', 'TACGTAATTCTG', 'TACGTAATCTTA', 'TACGTAATCTTG', 'TACGTAATCCTA', 'TACGTAATCCTT', 'TACGTAATCCTC', 'TACGTAATCCTG', 'TACGTTATATTA', 'TACGTTATATTG', 'TACGTTATACTA', 'TACGTTATACTT', 'TACGTTATACTC', 'TACGTTATACTG', 'TACGTTATTTTA', 'TACGTTATTTTG', 'TACGTTATTCTA', 'TACGTTATTCTT', 'TACGTTATTCTC', 'TACGTTATTCTG', 'TACGTTATCTTA', 'TACGTTATCTTG', 'TACGTTATCCTA', 'TACGTTATCCTT', 'TACGTTATCCTC', 'TACGTTATCCTG', 'TACGTCATATTA', 'TACGTCATATTG', 'TACGTCATACTA', 'TACGTCATACTT', 'TACGTCATACTC', 'TACGTCATACTG', 'TACGTCATTTTA', 'TACGTCATTTTG', 'TACGTCATTCTA', 'TACGTCATTCTT', 'TACGTCATTCTC', 'TACGTCATTCTG', 'TACGTCATCTTA', 'TACGTCATCTTG', 'TACGTCATCCTA', 'TACGTCATCCTT', 'TACGTCATCCTC', 'TACGTCATCCTG', 'TACGTGATATTA', 'TACGTGATATTG', 'TACGTGATACTA', 'TACGTGATACTT', 'TACGTGATACTC', 'TACGTGATACTG', 'TACGTGATTTTA', 'TACGTGATTTTG', 'TACGTGATTCTA', 'TACGTGATTCTT', 'TACGTGATTCTC', 'TACGTGATTCTG', 'TACGTGATCTTA', 'TACGTGATCTTG', 'TACGTGATCCTA', 'TACGTGATCCTT', 'TACGTGATCCTC', 'TACGTGATCCTG'], (62, 68): ['TCTGTA'], (68, 71): ['AAA', 'AAG'], (71, 81): ['GCTGCTATTT']}\n",
      "0\n",
      "6912\n"
     ]
    }
   ],
   "source": [
    "#for promoter gene I:\n",
    "#rbs far away:\n",
    "\n",
    "prom = \"ttgttcagggtgttcagttaattcTcccgtctaatgCgcttccctgtttttaTgtTATtcTctctgtaAaggctgctattt\"\n",
    "\n",
    "assert len(prom) == 81, \"normalement, la longueur de la séquence promotrice devrait être de 81 pb exactement\"\n",
    "assert not any_upper_case(prom[0:3]),\"pas de lettre majuscule dans les trois premiers nucleotides, sinon ca plante...\"\n",
    "ref_gene = all_cds_m13[\"VI\"].lower()\n",
    "\n",
    "start_of_prom_in_ref_gene = ref_gene.find(prom.lower())\n",
    "substrings_to_consider = {}\n",
    "traceback = 0\n",
    "boolean = True\n",
    "promseq_in_refgene = ref_gene[start_of_prom_in_ref_gene:start_of_prom_in_ref_gene+len(prom)]\n",
    "\n",
    "for pos, nt in enumerate(prom):\n",
    "    #set the seed:\n",
    "    if pos >= traceback:\n",
    "        if nt.isupper():\n",
    "            #est-ce qu'il y a une série de codons contenant une lettre majuscule?:\n",
    "            pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos\n",
    "            #start codon: le codon in frame (in prom) duquel on commence à changer des codons\n",
    "            start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "            pos_of_prom_in_frame = pos-(pos_uppercase_in_ref_gene%3)\n",
    "            rest_of_81bp = prom[pos_of_prom_in_frame+3:]\n",
    "            window = prom[start_codon:start_codon+3]\n",
    "            n = 0\n",
    "            for next_codon in rest_of_81bp:\n",
    "                if any(t.isupper() for t in rest_of_81bp[n:n+3]):\n",
    "                    #extend the portion of prom to change:\n",
    "                    window = window+rest_of_81bp[n:n+3]\n",
    "                    n+=3\n",
    "                else:\n",
    "                    break\n",
    "            switch_codon = find_all_replacement_codons(window.upper(),rev_ecoli_codon_usage, start_codon)\n",
    "            list_switch_codon = [possibilities for position, possibilities in switch_codon.items()]\n",
    "            substrings_to_consider[start_codon,start_codon+n+3] = [''.join(i) for i in list(product(*list_switch_codon))]\n",
    "            traceback = start_codon+n+3\n",
    "        else:\n",
    "            if boolean:\n",
    "                #save the position to append correctly in substrings_to_consider\n",
    "                pos_1st_lower_case = pos\n",
    "                boolean = False\n",
    "            if pos < len(prom)-1:\n",
    "                if prom[pos+1].isupper():\n",
    "                    #il faut alors savoir ou est le prochain cadre de lecture:\n",
    "                    pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos+1\n",
    "                    start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "                    substrings_to_consider[pos_1st_lower_case,start_codon] = [prom[pos_1st_lower_case:start_codon].upper()]\n",
    "                    boolean = True\n",
    "            else:\n",
    "                substrings_to_consider[pos_1st_lower_case,pos+1] = [prom[pos_1st_lower_case:pos+1].upper()]\n",
    "    \n",
    "    \n",
    "print(substrings_to_consider)\n",
    "list_of_all_possibilities = [all_substring for position, all_substring in substrings_to_consider.items()]\n",
    "all_possible_promoter_I = [''.join(i) for i in list(product(*list_of_all_possibilities))]\n",
    "\n",
    "#if no NOOOOOO, meaning all sequences encode for the same amino acid. \n",
    "for n,i in enumerate(all_possible_promoter_I):\n",
    "    if not check_if_same_prot(ref_gene[:start_of_prom_in_ref_gene]+i+ref_gene[start_of_prom_in_ref_gene+len(i):],ref_gene):\n",
    "        print(\"NOOOOOOOOOOOOOO\")\n",
    "    if n%10000 == 0:\n",
    "        print(n)\n",
    "        \n",
    "print(len(all_possible_promoter_I))\n",
    "#Select random element in a list to get a much smaller input: \n",
    "if len(all_possible_promoter_I) > 10000:\n",
    "    random.seed(i)\n",
    "    all_possible_promoter_I = np.random.choice(all_possible_promoter_I, 10000, replace=False)  \n",
    "    \n",
    "with open(\"./altered_proms/promI/promI.fa\",\"w\") as fout:\n",
    "    string = \"\"\n",
    "    for i,seq_prom in enumerate(all_possible_promoter_I):\n",
    "        string+=f\">prom_I_{i}\\n{seq_prom}\\n\"\n",
    "    fout.write(string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{}\n",
      "1920\n",
      "618\n",
      "('prom_I_5727', ['TTGTTCAGGGTGTTCAGTTAATTCTCCCGTCTAATGCGCTTCCCTGTTTTTACGTGATATTGTCTGTAAAGGCTGCTATTT', 0.1257])\n"
     ]
    }
   ],
   "source": [
    "#make calculations for I and get URL on web server:\n",
    "#1st prom2L:\n",
    "n=0\n",
    "iProm2L_results = {}\n",
    "for i,line in enumerate(urllib2.urlopen(\"http://bioinformatics.hitsz.edu.cn/iPromoter-2L/static/result/162.217.60.184_1629406192.71.txt\")):\n",
    "    if \"The length of the query sequence\" in str(line):\n",
    "        prom_name = \"prom\"+str(line.decode(\"utf-8\")).split(\"prom\")[1].split(\" )\")[0]\n",
    "    elif \"non-promoter sequence\" in str(line):\n",
    "        n=1\n",
    "    elif n == 1:\n",
    "        n=0\n",
    "        iProm2L_results[prom_name] = str(line).split(\":\")[1][:-3]\n",
    "print(iProm2L_results)\n",
    "\n",
    "#ipro70 analysis:\n",
    "with open('./altered_proms/promI/ipro70_results', 'r') as fin:\n",
    "    ipro70_results = {}\n",
    "    for i, line in enumerate(fin):\n",
    "        if i%3 == 0:\n",
    "            prom_name = line.replace(\" \",\"_\").strip(\"\\n\")\n",
    "        if i%3 == 1:\n",
    "            prom_seq = line\n",
    "        if i%3 == 2 and \"Non-promoter\" in line:\n",
    "            ipro70_results[prom_name] = [prom_seq.strip(\"\\n\"),float(line.split(\"score: \")[1].strip(\"\\n\"))]\n",
    "print(len(ipro70_results))\n",
    "\n",
    "#CNN prom analysis            \n",
    "with open('./altered_proms/promI/CNN_prom_results', 'r') as fin:\n",
    "    CNN_prom = []\n",
    "    for i, line in enumerate(fin):\n",
    "        if i%2 == 0:\n",
    "            prom_name = line[1:].replace(\" \",\"_\").strip(\"\\n\")\n",
    "        if i%2 == 1 and \"Not a Promoter\" in line:\n",
    "            CNN_prom.append(prom_name)\n",
    "print(len(CNN_prom))\n",
    "\n",
    "common_proms_CNN_ipro70 = {prom_name:prom_seq for prom_name, prom_seq in ipro70_results.items() if prom_name in CNN_prom}\n",
    "prom_seq_lowscore = min(common_proms_CNN_ipro70.items(), key=lambda x: x[1][1])\n",
    "\n",
    "print(prom_seq_lowscore)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{(0, 10): ['TGTAACTGGC'], (10, 16): ['AAATTA', 'AAATTG', 'AAACTA', 'AAACTT', 'AAACTC', 'AAACTG', 'AAGTTA', 'AAGTTG', 'AAGCTA', 'AAGCTT', 'AAGCTC', 'AAGCTG'], (16, 19): ['GGC'], (19, 28): ['AGTGGAAAA', 'AGTGGAAAG', 'AGTGGTAAA', 'AGTGGTAAG', 'AGTGGCAAA', 'AGTGGCAAG', 'AGTGGGAAA', 'AGTGGGAAG', 'AGCGGAAAA', 'AGCGGAAAG', 'AGCGGTAAA', 'AGCGGTAAG', 'AGCGGCAAA', 'AGCGGCAAG', 'AGCGGGAAA', 'AGCGGGAAG', 'TCAGGAAAA', 'TCAGGAAAG', 'TCAGGTAAA', 'TCAGGTAAG', 'TCAGGCAAA', 'TCAGGCAAG', 'TCAGGGAAA', 'TCAGGGAAG', 'TCTGGAAAA', 'TCTGGAAAG', 'TCTGGTAAA', 'TCTGGTAAG', 'TCTGGCAAA', 'TCTGGCAAG', 'TCTGGGAAA', 'TCTGGGAAG', 'TCCGGAAAA', 'TCCGGAAAG', 'TCCGGTAAA', 'TCCGGTAAG', 'TCCGGCAAA', 'TCCGGCAAG', 'TCCGGGAAA', 'TCCGGGAAG', 'TCGGGAAAA', 'TCGGGAAAG', 'TCGGGTAAA', 'TCGGGTAAG', 'TCGGGCAAA', 'TCGGGCAAG', 'TCGGGGAAA', 'TCGGGGAAG'], (28, 40): ['ACGCTCGTTAGC'], (40, 52): ['GTAGGAAAAATA', 'GTAGGAAAAATT', 'GTAGGAAAAATC', 'GTAGGAAAGATA', 'GTAGGAAAGATT', 'GTAGGAAAGATC', 'GTAGGTAAAATA', 'GTAGGTAAAATT', 'GTAGGTAAAATC', 'GTAGGTAAGATA', 'GTAGGTAAGATT', 'GTAGGTAAGATC', 'GTAGGCAAAATA', 'GTAGGCAAAATT', 'GTAGGCAAAATC', 'GTAGGCAAGATA', 'GTAGGCAAGATT', 'GTAGGCAAGATC', 'GTAGGGAAAATA', 'GTAGGGAAAATT', 'GTAGGGAAAATC', 'GTAGGGAAGATA', 'GTAGGGAAGATT', 'GTAGGGAAGATC', 'GTTGGAAAAATA', 'GTTGGAAAAATT', 'GTTGGAAAAATC', 'GTTGGAAAGATA', 'GTTGGAAAGATT', 'GTTGGAAAGATC', 'GTTGGTAAAATA', 'GTTGGTAAAATT', 'GTTGGTAAAATC', 'GTTGGTAAGATA', 'GTTGGTAAGATT', 'GTTGGTAAGATC', 'GTTGGCAAAATA', 'GTTGGCAAAATT', 'GTTGGCAAAATC', 'GTTGGCAAGATA', 'GTTGGCAAGATT', 'GTTGGCAAGATC', 'GTTGGGAAAATA', 'GTTGGGAAAATT', 'GTTGGGAAAATC', 'GTTGGGAAGATA', 'GTTGGGAAGATT', 'GTTGGGAAGATC', 'GTCGGAAAAATA', 'GTCGGAAAAATT', 'GTCGGAAAAATC', 'GTCGGAAAGATA', 'GTCGGAAAGATT', 'GTCGGAAAGATC', 'GTCGGTAAAATA', 'GTCGGTAAAATT', 'GTCGGTAAAATC', 'GTCGGTAAGATA', 'GTCGGTAAGATT', 'GTCGGTAAGATC', 'GTCGGCAAAATA', 'GTCGGCAAAATT', 'GTCGGCAAAATC', 'GTCGGCAAGATA', 'GTCGGCAAGATT', 'GTCGGCAAGATC', 'GTCGGGAAAATA', 'GTCGGGAAAATT', 'GTCGGGAAAATC', 'GTCGGGAAGATA', 'GTCGGGAAGATT', 'GTCGGGAAGATC', 'GTGGGAAAAATA', 'GTGGGAAAAATT', 'GTGGGAAAAATC', 'GTGGGAAAGATA', 'GTGGGAAAGATT', 'GTGGGAAAGATC', 'GTGGGTAAAATA', 'GTGGGTAAAATT', 'GTGGGTAAAATC', 'GTGGGTAAGATA', 'GTGGGTAAGATT', 'GTGGGTAAGATC', 'GTGGGCAAAATA', 'GTGGGCAAAATT', 'GTGGGCAAAATC', 'GTGGGCAAGATA', 'GTGGGCAAGATT', 'GTGGGCAAGATC', 'GTGGGGAAAATA', 'GTGGGGAAAATT', 'GTGGGGAAAATC', 'GTGGGGAAGATA', 'GTGGGGAAGATT', 'GTGGGGAAGATC'], (52, 58): ['CAGGAT'], (58, 64): ['AAAATA', 'AAAATT', 'AAAATC', 'AAGATA', 'AAGATT', 'AAGATC'], (64, 81): ['GTAGCTGGGTGCAAAAT']}\n",
      "331776\n",
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "40000\n",
      "50000\n",
      "60000\n",
      "70000\n",
      "80000\n",
      "90000\n",
      "100000\n",
      "110000\n",
      "120000\n",
      "130000\n",
      "140000\n",
      "150000\n",
      "160000\n",
      "170000\n",
      "180000\n",
      "190000\n",
      "200000\n",
      "210000\n",
      "220000\n",
      "230000\n",
      "240000\n",
      "250000\n",
      "260000\n",
      "270000\n",
      "280000\n",
      "290000\n",
      "300000\n",
      "310000\n",
      "320000\n",
      "330000\n",
      "331776\n"
     ]
    }
   ],
   "source": [
    "#for promoter gene I':\n",
    "#rbs far away:\n",
    "#Note: There is a difference in this sequence and the sequence of the fd phage.\n",
    "#To be sure this promoter is altered, every difference have been mark as uppercase.\n",
    "\n",
    "prom = \"tgtaactggcaAATtaggctCTgGAaAgacgctcgttagcgtTggTAagATtcaggataAAATtgtagctgggtgcaaaat\"\n",
    "\n",
    "assert len(prom) == 81, \"normalement, la longueur de la séquence promotrice devrait être de 81 pb exactement\"\n",
    "assert not any_upper_case(prom[0:3]),\"pas de lettre majuscule dans les trois premiers nucleotides, sinon ca plante...\"\n",
    "ref_gene = all_cds_m13[\"I\"].lower()\n",
    "\n",
    "start_of_prom_in_ref_gene = ref_gene.find(prom.lower())\n",
    "substrings_to_consider = {}\n",
    "traceback = 0\n",
    "boolean = True\n",
    "promseq_in_refgene = ref_gene[start_of_prom_in_ref_gene:start_of_prom_in_ref_gene+len(prom)]\n",
    "\n",
    "for pos, nt in enumerate(prom):\n",
    "    #set the seed:\n",
    "    if pos >= traceback:\n",
    "        if nt.isupper():\n",
    "            #est-ce qu'il y a une série de codons contenant une lettre majuscule?:\n",
    "            pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos\n",
    "            #start codon: le codon in frame (in prom) duquel on commence à changer des codons\n",
    "            start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "            pos_of_prom_in_frame = pos-(pos_uppercase_in_ref_gene%3)\n",
    "            rest_of_81bp = prom[pos_of_prom_in_frame+3:]\n",
    "            window = prom[start_codon:start_codon+3]\n",
    "            n = 0\n",
    "            for next_codon in rest_of_81bp:\n",
    "                if any(t.isupper() for t in rest_of_81bp[n:n+3]):\n",
    "                    #extend the portion of prom to change:\n",
    "                    window = window+rest_of_81bp[n:n+3]\n",
    "                    n+=3\n",
    "                else:\n",
    "                    break\n",
    "            switch_codon = find_all_replacement_codons(window.upper(),rev_ecoli_codon_usage, start_codon)\n",
    "            list_switch_codon = [possibilities for position, possibilities in switch_codon.items()]\n",
    "            substrings_to_consider[start_codon,start_codon+n+3] = [''.join(i) for i in list(product(*list_switch_codon))]\n",
    "            traceback = start_codon+n+3\n",
    "        else:\n",
    "            if boolean:\n",
    "                #save the position to append correctly in substrings_to_consider\n",
    "                pos_1st_lower_case = pos\n",
    "                boolean = False\n",
    "            if pos < len(prom)-1:\n",
    "                if prom[pos+1].isupper():\n",
    "                    #il faut alors savoir ou est le prochain cadre de lecture:\n",
    "                    pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos+1\n",
    "                    start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "                    substrings_to_consider[pos_1st_lower_case,start_codon] = [prom[pos_1st_lower_case:start_codon].upper()]\n",
    "                    boolean = True\n",
    "            else:\n",
    "                substrings_to_consider[pos_1st_lower_case,pos+1] = [prom[pos_1st_lower_case:pos+1].upper()]\n",
    "\n",
    "print(substrings_to_consider)\n",
    "list_of_all_possibilities = [all_substring for position, all_substring in substrings_to_consider.items()]\n",
    "all_possible_promoter_I_prime = [''.join(i) for i in list(product(*list_of_all_possibilities))]\n",
    "print(len(all_possible_promoter_I_prime))\n",
    "\n",
    "#if no NOOOOOO, meaning all sequences encode for the same amino acid. \n",
    "for n,i in enumerate(all_possible_promoter_I_prime):\n",
    "    if not check_if_same_prot(ref_gene[:start_of_prom_in_ref_gene]+i+ref_gene[start_of_prom_in_ref_gene+len(i):],ref_gene):\n",
    "        print(\"NOOOOOOOOOOOOOO\")\n",
    "    if n%10000 == 0:\n",
    "        print(n)\n",
    "        \n",
    "print(len(all_possible_promoter_I_prime))\n",
    "#Select random element in a list to get a much smaller input: \n",
    "if len(all_possible_promoter_I_prime) > 10000:\n",
    "    random.seed(i)\n",
    "    reduced_all_possible_promoter_I_prime = np.random.choice(all_possible_promoter_I_prime, 10000, replace=False)  \n",
    "    \n",
    "with open(\"./altered_proms/promI_prime/promI_prime.fa\",\"w\") as fout:\n",
    "    string = \"\"\n",
    "    for i,seq_prom in enumerate(reduced_all_possible_promoter_I_prime):\n",
    "        string+=f\">prom_I_prime_{i}\\n{seq_prom}\\n\"\n",
    "    fout.write(string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7038\n",
      "7171\n",
      "8001\n",
      "('prom_I_prime_1440', ['TGTAACTGGCAAGCTTGGCTCGGGCAAGACGCTCGTTAGCGTAGGCAAGATCCAGGATAAGATCGTAGCTGGGTGCAAAAT', 0.0109])\n"
     ]
    }
   ],
   "source": [
    "#make calculations for I' and get URL on web server:\n",
    "#1st prom2L:\n",
    "n=0\n",
    "iProm2L_results = {}\n",
    "for i,line in enumerate(urllib2.urlopen(\"http://bioinformatics.hitsz.edu.cn/iPromoter-2L/static/result/162.217.60.184_1629409518.75.txt\")):\n",
    "    if \"The length of the query sequence\" in str(line):\n",
    "        prom_name = \"prom\"+str(line.decode(\"utf-8\")).split(\"prom\")[1].split(\" )\")[0]\n",
    "    elif \"non-promoter sequence\" in str(line):\n",
    "        n=1\n",
    "    elif n == 1:\n",
    "        n=0\n",
    "        iProm2L_results[prom_name] = str(line).split(\":\")[1][:-3]\n",
    "print(len(iProm2L_results))\n",
    "\n",
    "#ipro70 analysis:\n",
    "with open('./altered_proms/promI_prime/ipro70_results', 'r') as fin:\n",
    "    ipro70_results = {}\n",
    "    for i, line in enumerate(fin):\n",
    "        if i%3 == 0:\n",
    "            prom_name = line.replace(\" \",\"_\").strip(\"\\n\")\n",
    "        if i%3 == 1:\n",
    "            prom_seq = line\n",
    "        if i%3 == 2 and \"Non-promoter\" in line:\n",
    "            ipro70_results[prom_name] = [prom_seq.strip(\"\\n\"),float(line.split(\"score: \")[1].strip(\"\\n\"))]\n",
    "print(len(ipro70_results))\n",
    "\n",
    "#CNN prom analysis            \n",
    "with open('./altered_proms/promI_prime/CNN_prom_results', 'r') as fin:\n",
    "    CNN_prom = []\n",
    "    for i, line in enumerate(fin):\n",
    "        if i%2 == 0:\n",
    "            prom_name = line[1:].replace(\" \",\"_\").strip(\"\\n\")\n",
    "        if i%2 == 1 and \"Not a Promoter\" in line:\n",
    "            CNN_prom.append(prom_name)\n",
    "print(len(CNN_prom))\n",
    "\n",
    "common_proms_CNN_ipro70 = {prom_name:prom_seq for prom_name, prom_seq in ipro70_results.items() if prom_name in CNN_prom and prom_name in iProm2L_results.keys()}\n",
    "prom_seq_lowscore = min(common_proms_CNN_ipro70.items(), key=lambda x: x[1][1])\n",
    "\n",
    "print(prom_seq_lowscore)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{(0, 10): ['TTTTAAAATT'], (10, 13): ['AAT', 'AAC'], (13, 19): ['AACGTT'], (19, 28): ['AGAGCAAAA', 'AGAGCAAAG', 'AGAGCTAAA', 'AGAGCTAAG', 'AGAGCCAAA', 'AGAGCCAAG', 'AGAGCGAAA', 'AGAGCGAAG', 'AGGGCAAAA', 'AGGGCAAAG', 'AGGGCTAAA', 'AGGGCTAAG', 'AGGGCCAAA', 'AGGGCCAAG', 'AGGGCGAAA', 'AGGGCGAAG', 'CGAGCAAAA', 'CGAGCAAAG', 'CGAGCTAAA', 'CGAGCTAAG', 'CGAGCCAAA', 'CGAGCCAAG', 'CGAGCGAAA', 'CGAGCGAAG', 'CGTGCAAAA', 'CGTGCAAAG', 'CGTGCTAAA', 'CGTGCTAAG', 'CGTGCCAAA', 'CGTGCCAAG', 'CGTGCGAAA', 'CGTGCGAAG', 'CGCGCAAAA', 'CGCGCAAAG', 'CGCGCTAAA', 'CGCGCTAAG', 'CGCGCCAAA', 'CGCGCCAAG', 'CGCGCGAAA', 'CGCGCGAAG', 'CGGGCAAAA', 'CGGGCAAAG', 'CGGGCTAAA', 'CGGGCTAAG', 'CGGGCCAAA', 'CGGGCCAAG', 'CGGGCGAAA', 'CGGGCGAAG'], (28, 40): ['GATTTAATACGA'], (40, 52): ['GTAGTAGAATTA', 'GTAGTAGAATTG', 'GTAGTAGAACTA', 'GTAGTAGAACTT', 'GTAGTAGAACTC', 'GTAGTAGAACTG', 'GTAGTAGAGTTA', 'GTAGTAGAGTTG', 'GTAGTAGAGCTA', 'GTAGTAGAGCTT', 'GTAGTAGAGCTC', 'GTAGTAGAGCTG', 'GTAGTTGAATTA', 'GTAGTTGAATTG', 'GTAGTTGAACTA', 'GTAGTTGAACTT', 'GTAGTTGAACTC', 'GTAGTTGAACTG', 'GTAGTTGAGTTA', 'GTAGTTGAGTTG', 'GTAGTTGAGCTA', 'GTAGTTGAGCTT', 'GTAGTTGAGCTC', 'GTAGTTGAGCTG', 'GTAGTCGAATTA', 'GTAGTCGAATTG', 'GTAGTCGAACTA', 'GTAGTCGAACTT', 'GTAGTCGAACTC', 'GTAGTCGAACTG', 'GTAGTCGAGTTA', 'GTAGTCGAGTTG', 'GTAGTCGAGCTA', 'GTAGTCGAGCTT', 'GTAGTCGAGCTC', 'GTAGTCGAGCTG', 'GTAGTGGAATTA', 'GTAGTGGAATTG', 'GTAGTGGAACTA', 'GTAGTGGAACTT', 'GTAGTGGAACTC', 'GTAGTGGAACTG', 'GTAGTGGAGTTA', 'GTAGTGGAGTTG', 'GTAGTGGAGCTA', 'GTAGTGGAGCTT', 'GTAGTGGAGCTC', 'GTAGTGGAGCTG', 'GTTGTAGAATTA', 'GTTGTAGAATTG', 'GTTGTAGAACTA', 'GTTGTAGAACTT', 'GTTGTAGAACTC', 'GTTGTAGAACTG', 'GTTGTAGAGTTA', 'GTTGTAGAGTTG', 'GTTGTAGAGCTA', 'GTTGTAGAGCTT', 'GTTGTAGAGCTC', 'GTTGTAGAGCTG', 'GTTGTTGAATTA', 'GTTGTTGAATTG', 'GTTGTTGAACTA', 'GTTGTTGAACTT', 'GTTGTTGAACTC', 'GTTGTTGAACTG', 'GTTGTTGAGTTA', 'GTTGTTGAGTTG', 'GTTGTTGAGCTA', 'GTTGTTGAGCTT', 'GTTGTTGAGCTC', 'GTTGTTGAGCTG', 'GTTGTCGAATTA', 'GTTGTCGAATTG', 'GTTGTCGAACTA', 'GTTGTCGAACTT', 'GTTGTCGAACTC', 'GTTGTCGAACTG', 'GTTGTCGAGTTA', 'GTTGTCGAGTTG', 'GTTGTCGAGCTA', 'GTTGTCGAGCTT', 'GTTGTCGAGCTC', 'GTTGTCGAGCTG', 'GTTGTGGAATTA', 'GTTGTGGAATTG', 'GTTGTGGAACTA', 'GTTGTGGAACTT', 'GTTGTGGAACTC', 'GTTGTGGAACTG', 'GTTGTGGAGTTA', 'GTTGTGGAGTTG', 'GTTGTGGAGCTA', 'GTTGTGGAGCTT', 'GTTGTGGAGCTC', 'GTTGTGGAGCTG', 'GTCGTAGAATTA', 'GTCGTAGAATTG', 'GTCGTAGAACTA', 'GTCGTAGAACTT', 'GTCGTAGAACTC', 'GTCGTAGAACTG', 'GTCGTAGAGTTA', 'GTCGTAGAGTTG', 'GTCGTAGAGCTA', 'GTCGTAGAGCTT', 'GTCGTAGAGCTC', 'GTCGTAGAGCTG', 'GTCGTTGAATTA', 'GTCGTTGAATTG', 'GTCGTTGAACTA', 'GTCGTTGAACTT', 'GTCGTTGAACTC', 'GTCGTTGAACTG', 'GTCGTTGAGTTA', 'GTCGTTGAGTTG', 'GTCGTTGAGCTA', 'GTCGTTGAGCTT', 'GTCGTTGAGCTC', 'GTCGTTGAGCTG', 'GTCGTCGAATTA', 'GTCGTCGAATTG', 'GTCGTCGAACTA', 'GTCGTCGAACTT', 'GTCGTCGAACTC', 'GTCGTCGAACTG', 'GTCGTCGAGTTA', 'GTCGTCGAGTTG', 'GTCGTCGAGCTA', 'GTCGTCGAGCTT', 'GTCGTCGAGCTC', 'GTCGTCGAGCTG', 'GTCGTGGAATTA', 'GTCGTGGAATTG', 'GTCGTGGAACTA', 'GTCGTGGAACTT', 'GTCGTGGAACTC', 'GTCGTGGAACTG', 'GTCGTGGAGTTA', 'GTCGTGGAGTTG', 'GTCGTGGAGCTA', 'GTCGTGGAGCTT', 'GTCGTGGAGCTC', 'GTCGTGGAGCTG', 'GTGGTAGAATTA', 'GTGGTAGAATTG', 'GTGGTAGAACTA', 'GTGGTAGAACTT', 'GTGGTAGAACTC', 'GTGGTAGAACTG', 'GTGGTAGAGTTA', 'GTGGTAGAGTTG', 'GTGGTAGAGCTA', 'GTGGTAGAGCTT', 'GTGGTAGAGCTC', 'GTGGTAGAGCTG', 'GTGGTTGAATTA', 'GTGGTTGAATTG', 'GTGGTTGAACTA', 'GTGGTTGAACTT', 'GTGGTTGAACTC', 'GTGGTTGAACTG', 'GTGGTTGAGTTA', 'GTGGTTGAGTTG', 'GTGGTTGAGCTA', 'GTGGTTGAGCTT', 'GTGGTTGAGCTC', 'GTGGTTGAGCTG', 'GTGGTCGAATTA', 'GTGGTCGAATTG', 'GTGGTCGAACTA', 'GTGGTCGAACTT', 'GTGGTCGAACTC', 'GTGGTCGAACTG', 'GTGGTCGAGTTA', 'GTGGTCGAGTTG', 'GTGGTCGAGCTA', 'GTGGTCGAGCTT', 'GTGGTCGAGCTC', 'GTGGTCGAGCTG', 'GTGGTGGAATTA', 'GTGGTGGAATTG', 'GTGGTGGAACTA', 'GTGGTGGAACTT', 'GTGGTGGAACTC', 'GTGGTGGAACTG', 'GTGGTGGAGTTA', 'GTGGTGGAGTTG', 'GTGGTGGAGCTA', 'GTGGTGGAGCTT', 'GTGGTGGAGCTC', 'GTGGTGGAGCTG'], (52, 58): ['TTTGTA'], (58, 61): ['AAA', 'AAG'], (61, 81): ['TCTAATACTTCTAAATCCTC']}\n",
      "36864\n",
      "36864\n"
     ]
    }
   ],
   "source": [
    "#prom IV'\n",
    "#No problem with anything!\n",
    "\n",
    "prom = \"ttttaaaattAATaacgttCggGcaAaggatttaatacgagTtgTcgAATtgtttgtaAagtctaatacttctaaatcctc\"\n",
    "\n",
    "\n",
    "assert len(prom) == 81, \"normalement, la longueur de la séquence promotrice devrait être de 81 pb exactement\"\n",
    "assert not any_upper_case(prom[0:3]),\"pas de lettre majuscule dans les trois premiers nucleotides, sinon ca plante...\"\n",
    "ref_gene = all_cds_m13[\"IV\"].lower()\n",
    "\n",
    "start_of_prom_in_ref_gene = ref_gene.find(prom.lower())\n",
    "substrings_to_consider = {}\n",
    "traceback = 0\n",
    "boolean = True\n",
    "promseq_in_refgene = ref_gene[start_of_prom_in_ref_gene:start_of_prom_in_ref_gene+len(prom)]\n",
    "\n",
    "for pos, nt in enumerate(prom):\n",
    "    #set the seed:\n",
    "    if pos >= traceback:\n",
    "        if nt.isupper():\n",
    "            #est-ce qu'il y a une série de codons contenant une lettre majuscule?:\n",
    "            pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos\n",
    "            #start codon: le codon in frame (in prom) duquel on commence à changer des codons\n",
    "            start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "            pos_of_prom_in_frame = pos-(pos_uppercase_in_ref_gene%3)\n",
    "            rest_of_81bp = prom[pos_of_prom_in_frame+3:]\n",
    "            window = prom[start_codon:start_codon+3]\n",
    "            n = 0\n",
    "            for next_codon in rest_of_81bp:\n",
    "                if any(t.isupper() for t in rest_of_81bp[n:n+3]):\n",
    "                    #extend the portion of prom to change:\n",
    "                    window = window+rest_of_81bp[n:n+3]\n",
    "                    n+=3\n",
    "                else:\n",
    "                    break\n",
    "            switch_codon = find_all_replacement_codons(window.upper(),rev_ecoli_codon_usage, start_codon)\n",
    "            list_switch_codon = [possibilities for position, possibilities in switch_codon.items()]\n",
    "            substrings_to_consider[start_codon,start_codon+n+3] = [''.join(i) for i in list(product(*list_switch_codon))]\n",
    "            traceback = start_codon+n+3\n",
    "        else:\n",
    "            if boolean:\n",
    "                #save the position to append correctly in substrings_to_consider\n",
    "                pos_1st_lower_case = pos\n",
    "                boolean = False\n",
    "            if pos < len(prom)-1:\n",
    "                if prom[pos+1].isupper():\n",
    "                    #il faut alors savoir ou est le prochain cadre de lecture:\n",
    "                    pos_uppercase_in_ref_gene = start_of_prom_in_ref_gene + pos+1\n",
    "                    start_codon = (pos_uppercase_in_ref_gene-(pos_uppercase_in_ref_gene%3))-start_of_prom_in_ref_gene\n",
    "                    substrings_to_consider[pos_1st_lower_case,start_codon] = [prom[pos_1st_lower_case:start_codon].upper()]\n",
    "                    boolean = True\n",
    "            else:\n",
    "                substrings_to_consider[pos_1st_lower_case,pos+1] = [prom[pos_1st_lower_case:pos+1].upper()]\n",
    "\n",
    "print(substrings_to_consider)\n",
    "list_of_all_possibilities = [all_substring for position, all_substring in substrings_to_consider.items()]\n",
    "all_possible_promoter_IV_prime = [''.join(i) for i in list(product(*list_of_all_possibilities))]\n",
    "print(len(all_possible_promoter_IV_prime))\n",
    "\n",
    "#if no NOOOOOO, meaning all sequences encode for the same amino acid. \n",
    "for i in all_possible_promoter_IV_prime:\n",
    "    if not check_if_same_prot(ref_gene[:start_of_prom_in_ref_gene]+i+ref_gene[start_of_prom_in_ref_gene+len(i):],ref_gene):\n",
    "        print(\"NOOOOOOOOOOOOOO\")\n",
    "        \n",
    "print(len(all_possible_promoter_IV_prime))\n",
    "#Select random element in a list to get a much smaller input: \n",
    "if len(all_possible_promoter_IV_prime) > 10000:\n",
    "    random.seed(i)\n",
    "    reduced_all_possible_promoter_IV_prime = np.random.choice(all_possible_promoter_IV_prime, 10000, replace=False)  \n",
    "    \n",
    "with open(\"./altered_proms/promIV_prime/promIV_prime.fa\",\"w\") as fout:\n",
    "    string = \"\"\n",
    "    for i,seq_prom in enumerate(reduced_all_possible_promoter_IV_prime):\n",
    "        string+=f\">promIV_prime_{i}\\n{seq_prom}\\n\"\n",
    "    fout.write(string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "9528\n",
      "526\n",
      "('promIV_prime_6950', ['TTTTAAAATTAATAACGTTCGGGCCAAGGATTTAATACGAGTAGTCGAGCTCTTTGTAAAGTCTAATACTTCTAAATCCTC', 0.0095])\n"
     ]
    }
   ],
   "source": [
    "#make calculations for IV' and get URL on web server:\n",
    "#1st prom2L:\n",
    "n=0\n",
    "iProm2L_results = {}\n",
    "for i,line in enumerate(urllib2.urlopen(\"http://bioinformatics.hitsz.edu.cn/iPromoter-2L/static/result/162.217.60.184_1629412700.42.txt\")):\n",
    "    if \"The length of the query sequence\" in str(line):\n",
    "        prom_name = \"prom\"+str(line.decode(\"utf-8\")).split(\"prom\")[1].split(\" )\")[0]\n",
    "    elif \"non-promoter sequence\" in str(line):\n",
    "        n=1\n",
    "    elif n == 1:\n",
    "        n=0\n",
    "        iProm2L_results[prom_name] = str(line).split(\":\")[1][:-3]\n",
    "print(len(iProm2L_results))\n",
    "\n",
    "#ipro70 analysis:\n",
    "with open('./altered_proms/promIV_prime/ipro70_results', 'r') as fin:\n",
    "    ipro70_results = {}\n",
    "    for i, line in enumerate(fin):\n",
    "        if i%3 == 0:\n",
    "            prom_name = line.replace(\" \",\"_\").strip(\"\\n\")\n",
    "        if i%3 == 1:\n",
    "            prom_seq = line\n",
    "        if i%3 == 2 and \"Non-promoter\" in line:\n",
    "            ipro70_results[prom_name] = [prom_seq.strip(\"\\n\"),float(line.split(\"score: \")[1].strip(\"\\n\"))]\n",
    "print(len(ipro70_results))\n",
    "\n",
    "#CNN prom analysis            \n",
    "with open('./altered_proms/promIV_prime/CNN_prom_results', 'r') as fin:\n",
    "    CNN_prom = []\n",
    "    for i, line in enumerate(fin):\n",
    "        if i%2 == 0:\n",
    "            prom_name = line[1:].replace(\" \",\"_\").strip(\"\\n\")\n",
    "        if i%2 == 1 and \"Not a Promoter\" in line:\n",
    "            CNN_prom.append(prom_name)\n",
    "print(len(CNN_prom))\n",
    "\n",
    "common_proms_CNN_ipro70 = {prom_name:prom_seq for prom_name, prom_seq in ipro70_results.items() if prom_name in CNN_prom}\n",
    "prom_seq_lowscore = min(common_proms_CNN_ipro70.items(), key=lambda x: x[1][1])\n",
    "\n",
    "print(prom_seq_lowscore)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"atgcatgcatgcatgc\".find(\"tgca\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After building the .ape file of m13 with altered proms, I have to confirm every CDS encodes for the same amino acid:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "il y a 59 nt différents par rapport à la version originale\n",
      "{'II': 19, 'X': 6, 'V': 0, 'VII': 0, 'IX': 0, 'VIII': 0, 'III': 9, 'VI': 5, 'I': 15, 'IV': 5}\n"
     ]
    }
   ],
   "source": [
    "#confirm no problem occur during the construction:\n",
    "path_to_m13_modified_genome = \"../ref_files_m13/m13_alt/alt_prom_v1.gb\"\n",
    "i=True\n",
    "all_gene_m13_modified = {\"II\":\"atgattgacatgctagttttacgattaccgttcatcgattctcttgtttgctccagactctcaggcaatgacctgatagcctttgtagatctctcaaaaatagctaccctctccggcattaatttatcagctagaacggttgaatatcatattgatggtgatttgactgtctccggcctTTCTCACCCTTTTGAGTCGTTACCGACGCACTACTCAGGCATTGCATTCAAGATCTATGAGGGTTCTAAAAATTTTTATCCttgcgttgaaataaaggcttctcccgcaaaagtattacagggtcataatgtttttggtacaaccgatttagctttatgctctgaggctttattgcttaattttgctaattctttgccttgcctgtatgatttattggatgttaatgctactactattagtagaattgatgccaccttttcagctcgcgccccaaatgaaaatatagctaaacaggttattgaccatttgcgaaatgtatctaatggtcaaactaaatctactcgttcgcagaattgggaatcaactgttatatggaatgaaacttccagacaccgtactttagttgcatatttaaaacatgttgagctacagcattatattcagcaattaagctctaagccatccgcaaaaatgacctcttatcaaaaggagcaattaaaggtactctctaatcctgacctgttggagtttgcttccggtctggttcgctttgaagctcgaattaaaacgcgatatttgAAGTCTTTCGGGCTTCCTCTTAACCTCTTTGATGCTATCCGCTTTGCTTCTGACTACAACAGTCAGGGTAAAGACCTGATTtttgatttatggtcattctcgttttctgaactgtttaaagcatttgagggggattcaatgaatatttatgacgattccgcagtattggacgctatccagtctaaacattttactattaccccctctggcaaaacttcttttgcaaaagcctctcgctattttggtttttatcgtcgtctggtaaacgagggttatgatagtgttgctcttactatgcctcgtaattccttttggcgttatgtatctgcattagttgaatgtggtattcctaaatctcaactgatgaatctttctacctgtaataatgttGTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAATGAGCCCGTTCTTAAAATCGCAtaa\".upper(),\n",
    "                        \"X\":\"atgaatatttatgacgattccgcagtattggacgctatccagtctaaacattttactattaccccctctggcaaaacttcttttgcaaaagcctctcgctattttggtttttatcgtcgtctggtaaacgagggttatgatagtgttgctcttactatgcctcgtaattccttttggcgttatgtatctgcattagttgaatgtggtattcctaaatctcaactgatgaatctttctacctgtaataatgttGTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAATGAGCCCGTTCTTAAAATCGCAtaa\".upper(),\n",
    "                        \"V\":\"atgattaaagttgaaattaaaccatctcaagcccaatttactactcgttctggtgtttctcgtcagggcaagccttattcactgaatgagcagctttgttacgttgatttgggtaatgaatatccggttcttgtcaagattactcttgatgaaggtcagccagcctatgcgcctggtctgtacaccgttcatctgtcctctttcaaagttggtcagttcggttcccttatgattgaccgtctgcgcctcgttccggctaagtaa\".upper(),\n",
    "                        \"VII\":\"atggagcaggtcgcggatttcgacacaatttatcaggcgatgatacaaatctccgttgtactttgtttcgcgcttggtataatcgctgggggtcaaagatga\".upper(),\n",
    "                        \"IX\":\"atgagtgttttagtgtattcttttgcctctttcgttttaggttggtgccttcgtagtggcattacgtattttacccgtttaatggaaacttcctcatga\".upper(),\n",
    "                        \"VIII\":\"atgaaaaagtctttagtcctcaaagcctctgtagccgttgctaccctcgttccgatgctgtctttcgctgctgagggtgacgatcccgcaaaagcggcctttaactccctgcaagcctcagcgaccgaatatatcggttatgcgtgggcgatggttgttgtcattgtcggcgcaactatcggtatcaagctgtttaagaaattcacctcgaaagcaagctga\".upper(),\n",
    "                        \"III\":\"gtgaaaaaattattattcgcaattcctttagttgttcctttctattctcactccgctgaaactgttgaaagttgtttagcaaaatcccatacagaaaattcatttactaacgtctggaaagacgacaaaactttagatcgttacgctaactatgagggctgtctgtggaatgctacaggcgttgtagtttgtactggtgacgaaactcagtgttacggtacatgggttcctattgggcttgctatccctgaaaatgagggtggtggctctgagggtggcggttctgagggtggcggttctgagggtggcggtactaaacctcctgagtacggtgatacacctattccgggctatacttatatcaaccctctcgacggcacttatccgcctggtactgagcaaaaccccgctaatcctaatccttctcttgaggagtctcagcctcttaatactttcatgtttcagaataataggttccgaaataggcagggggcattaactgtttatacgggcactgttactcaaggcactgaccccgttaaaacttattaccagtacactcctgtatcatcaaaagccatgtatgacgcttactggaacggtaaattcagagactgcgctttccattctggctttaatgaggatttatttgtttgtgaatatcaaggccaatcgtctgacctgcctcaacctcctgtcaatgctggcggcggctctggtggtggttctggtggcggctctgagggtggtggctctgagggtggcggttctgagggtggcggctctgagggaggcggttccggtggtggctctggttccggtgattttgattatgaaaagatggcaaacgctaataagggggctatgaccgaaaatgccgatgaaaacgcgctacagtctgacgctaaaggcaaacttgattctgtcgctactgattacggtgctgctatcgatggtttcattggtgacgtttccggccttgctaatggtaatggtgctactggtgattttgctggctctaattcccaaatggctcaagtcggtgacggtgataattcacctttaatgaataatttccgtcaatatttaccttccctccctcaatcggttgaatgtcgcccttTTGTCTTTGGCGCTGGGAAACCGTACGAGTTTTCTATTGATTGTGACAAGATCAACCTCTTTCGTGGTGTCTTTGCGTTTCttttatatgttgccacctttatgtatgtattttctacgtttgctaacatactgcgtaataaggagtcttaa\".upper(),\n",
    "                        \"VI\":\"atgccagttcttttgggtattccgttattattgcgtttcctcggtttccttctggtaactttgttcggctatctgcttacttttcttaaaaagggcttcggtaagatagctattgctatttcattgtttcttgctcttattattgggcttaactcaattcttgtgggttatctctctgatattagcgctcaattaccctctgactTTGTTCAGGGTGTTCAGTTAATTCTCCCGTCTAATGCGCTTCCCTGTTTTTACGTGATATTGTCTGTAAAGGCTGCTATTTtcatttttgacgttaaacaaaaaatcgtttcttatttggattgggataaataa\".upper(),\n",
    "                        \"I\":\"atggctgtttatttTGTAACTGGCAAGCTTGGCTCGGGCAAGACGCTCGTTAGCGTAGGCAAGATCCAGGATAAGATCGTAGCTGGGTGCAAAATagcaactaatcttgatttaaggcttcaaaacctcccgcaagtcgggaggttcgctaaaacgcctcgcgttcttagaataccggataagccttctatatctgatttgcttgctattgggcgcggtaatgattcctacgatgaaaataaaaacggcttgcttgttctcgatgagtgcggtacttggtttaatacccgttcttggaatgataaggaaagacagccgattattgattggtttctacatgctcgtaaattaggatgggatattatttttcttgttcaggacttatctattgttgataaacaggcgcgttctgcattagctgaacatgttgtttattgtcgtcgtctggacagaattactttaccttttgtcggtactttatattctcttattactggctcgaaaatgcctctgcctaaattacatgttggcgttgttaaatatggcgattctcaattaagccctactgttgagcgttggctttatactggtaagaatttgtataacgcatatgatactaaacaggctttttctagtaattatgattccggtgtttattcttatttaacgccttatttatcacacggtcggtatttcaaaccattaaatttaggtcagaagatgaaattaactaaaatatatttgaaaaagttttctcgcgttctttgtcttgcgattggatttgcatcagcatttacatatagttatataacccaacctaagccggaggttaaaaaggtagtctctcagACCTATGATTTTGATAAATTCACTATCGATTCTTCTCAGCGTCTTAATCTCAGCTACCGCTACGTTTTCAAGGATTCTAAGggaaaattaattaatagcgacgatttacagaagcaaggttattcactcacatatattgatttatgtactgtttccattaaaaaaggtaattcaaatgaaattgttaaatgtaattaa\".upper(),\n",
    "                        \"IV\":\"atgaaattgttaaatgtaattaattttgttttcttgatgtttgtttcatcatcttcttttgctcaggtaattgaaatgaataattcgcctctgcgcgattttgtaacttggtattcaaagcaatcaggcgaatccgttattgtttctcccgatgtaaaaggtactgttactgtatattcatctgacgttaaacctgaaaatctacgcaatttctttatttctgttttacgtgcaaataattttgatatggtaggttctaacccttccattattcagaagtataatccaaacaatcaggattatattgatgaattgccatcatctgataatcaggaatatgatgataattccgctccttctggtggtttctttgttccgcaaaatgataatgttactcaaacTTTTAAAATTAATAACGTTCGGGCCAAGGATTTAATACGAGTAGTCGAGCTCTTTGTAAAGTCTAATACTTCTAAATCCTCaaatgtattatctattgacggctctaatctattagttgttagtgctcctaaagatattttagataaccttcctcaattcctttcaactgttgatttgccaactgaccagatattgattgagggtttgatatttgaggttcagcaaggtgatgctttagatttttcatttgctgctggctctcagcgtggcactgttgcaggcggtgttaatactgaccgcctcacctctgttttatcttctgctggtggttcgttcggtatttttaatggcgatgttttagggctatcagttcgcgcattaaagactaatagccattcaaaaatattgtctgtgccacgtattcttacgctttcaggtcagaagggttctatctctgttggccagaatgtcccttttattactggtcgtgtgactggtgaatctgccaatgtaaataatccatttcagacgattgagcgtcaaaatgtaggtatttccatgagcgtttttcctgttgcaatggctggcggtaatattgttctggatattaccagcaaggccgatagtttgagttcttctactcaggcaagtgatgttattactaatcaaagaagtattgctacaacggttaatttgcgtgatggacagactcttttactcggtggcctcactgattataaaaacacttctcaggattctggcgtaccgttcctgtctaaaatccctttaatcggcctcctgtttagctcccgctctgattctaacgaggaaagcacgttatacgtgctcgtcaaagcaaccatagtacgcgccctgtag\".upper(),\n",
    "                        }\n",
    "\n",
    "#get with the initial kevin neil file:\n",
    "all_gene_m13_native = {\"II\":\"atgattgacatgctagttttacgattaccgttcatcgattctcttgtttgctccagactctcaggcaatgacctgatagcctttgtagatctctcaaaaatagctaccctctccggcattaatttatcagctagaacggttgaatatcatattgatggtgatttgactgtctccggcctttctcacccttttgaatctttacctacacattactcaggcattgcatttaaaatatatgagggttctaaaaatttttatccttgcgttgaaataaaggcttctcccgcaaaagtattacagggtcataatgtttttggtacaaccgatttagctttatgctctgaggctttattgcttaattttgctaattctttgccttgcctgtatgatttattggatgttaatgctactactattagtagaattgatgccaccttttcagctcgcgccccaaatgaaaatatagctaaacaggttattgaccatttgcgaaatgtatctaatggtcaaactaaatctactcgttcgcagaattgggaatcaactgttatatggaatgaaacttccagacaccgtactttagttgcatatttaaaacatgttgagctacagcattatattcagcaattaagctctaagccatccgcaaaaatgacctcttatcaaaaggagcaattaaaggtactctctaatcctgacctgttggagtttgcttccggtctggttcgctttgaagctcgaattaaaacgcgatatttgaagtctttcgggcttcctcttaatctttttgatgcaatccgctttgcttctgactataatagtcagggtaaagacctgatttttgatttatggtcattctcgttttctgaactgtttaaagcatttgagggggattcaatgaatatttatgacgattccgcagtattggacgctatccagtctaaacattttactattaccccctctggcaaaacttcttttgcaaaagcctctcgctattttggtttttatcgtcgtctggtaaacgagggttatgatagtgttgctcttactatgcctcgtaattccttttggcgttatgtatctgcattagttgaatgtggtattcctaaatctcaactgatgaatctttctacctgtaataatgttgttccgttagttcgttttattaacgtagatttttcttcccaacgtcctgactggtataatgagccagttcttaaaatcgcataa\".upper(),\n",
    "                        \"X\":\"atgaatatttatgacgattccgcagtattggacgctatccagtctaaacattttactattaccccctctggcaaaacttcttttgcaaaagcctctcgctattttggtttttatcgtcgtctggtaaacgagggttatgatagtgttgctcttactatgcctcgtaattccttttggcgttatgtatctgcattagttgaatgtggtattcctaaatctcaactgatgaatctttctacctgtaataatgttgttccgttagttcgttttattaacgtagatttttcttcccaacgtcctgactggtataatgagccagttcttaaaatcgcataa\".upper(),\n",
    "                        \"V\":\"atgattaaagttgaaattaaaccatctcaagcccaatttactactcgttctggtgtttctcgtcagggcaagccttattcactgaatgagcagctttgttacgttgatttgggtaatgaatatccggttcttgtcaagattactcttgatgaaggtcagccagcctatgcgcctggtctgtacaccgttcatctgtcctctttcaaagttggtcagttcggttcccttatgattgaccgtctgcgcctcgttccggctaagtaa\".upper(),\n",
    "                        \"VII\":\"atggagcaggtcgcggatttcgacacaatttatcaggcgatgatacaaatctccgttgtactttgtttcgcgcttggtataatcgctgggggtcaaagatga\".upper(),\n",
    "                        \"IX\":\"atgagtgttttagtgtattcttttgcctctttcgttttaggttggtgccttcgtagtggcattacgtattttacccgtttaatggaaacttcctcatga\".upper(),\n",
    "                        \"VIII\":\"atgaaaaagtctttagtcctcaaagcctctgtagccgttgctaccctcgttccgatgctgtctttcgctgctgagggtgacgatcccgcaaaagcggcctttaactccctgcaagcctcagcgaccgaatatatcggttatgcgtgggcgatggttgttgtcattgtcggcgcaactatcggtatcaagctgtttaagaaattcacctcgaaagcaagctga\".upper(),\n",
    "                        \"III\":\"gtgaaaaaattattattcgcaattcctttagttgttcctttctattctcactccgctgaaactgttgaaagttgtttagcaaaatcccatacagaaaattcatttactaacgtctggaaagacgacaaaactttagatcgttacgctaactatgagggctgtctgtggaatgctacaggcgttgtagtttgtactggtgacgaaactcagtgttacggtacatgggttcctattgggcttgctatccctgaaaatgagggtggtggctctgagggtggcggttctgagggtggcggttctgagggtggcggtactaaacctcctgagtacggtgatacacctattccgggctatacttatatcaaccctctcgacggcacttatccgcctggtactgagcaaaaccccgctaatcctaatccttctcttgaggagtctcagcctcttaatactttcatgtttcagaataataggttccgaaataggcagggggcattaactgtttatacgggcactgttactcaaggcactgaccccgttaaaacttattaccagtacactcctgtatcatcaaaagccatgtatgacgcttactggaacggtaaattcagagactgcgctttccattctggctttaatgaggatttatttgtttgtgaatatcaaggccaatcgtctgacctgcctcaacctcctgtcaatgctggcggcggctctggtggtggttctggtggcggctctgagggtggtggctctgagggtggcggttctgagggtggcggctctgagggaggcggttccggtggtggctctggttccggtgattttgattatgaaaagatggcaaacgctaataagggggctatgaccgaaaatgccgatgaaaacgcgctacagtctgacgctaaaggcaaacttgattctgtcgctactgattacggtgctgctatcgatggtttcattggtgacgtttccggccttgctaatggtaatggtgctactggtgattttgctggctctaattcccaaatggctcaagtcggtgacggtgataattcacctttaatgaataatttccgtcaatatttaccttccctccctcaatcggttgaatgtcgcccttttgtctttggcgctggtaaaccatatgaattttctattgattgtgacaaaataaacttattccgtggtgtctttgcgtttcttttatatgttgccacctttatgtatgtattttctacgtttgctaacatactgcgtaataaggagtcttaa\".upper(),\n",
    "                        \"VI\":\"atgccagttcttttgggtattccgttattattgcgtttcctcggtttccttctggtaactttgttcggctatctgcttacttttcttaaaaagggcttcggtaagatagctattgctatttcattgtttcttgctcttattattgggcttaactcaattcttgtgggttatctctctgatattagcgctcaattaccctctgactttgttcagggtgttcagttaattctcccgtctaatgcgcttccctgtttttatgttattctctctgtaaaggctgctattttcatttttgacgttaaacaaaaaatcgtttcttatttggattgggataaataa\".upper(),\n",
    "                        \"I\":\"atggctgtttattttgtaactggcaaattaggctctggaaagacgctcgttagcgttggtaagattcaggataaaattgtagctgggtgcaaaatagcaactaatcttgatttaaggcttcaaaacctcccgcaagtcgggaggttcgctaaaacgcctcgcgttcttagaataccggataagccttctatatctgatttgcttgctattgggcgcggtaatgattcctacgatgaaaataaaaacggcttgcttgttctcgatgagtgcggtacttggtttaatacccgttcttggaatgataaggaaagacagccgattattgattggtttctacatgctcgtaaattaggatgggatattatttttcttgttcaggacttatctattgttgataaacaggcgcgttctgcattagctgaacatgttgtttattgtcgtcgtctggacagaattactttaccttttgtcggtactttatattctcttattactggctcgaaaatgcctctgcctaaattacatgttggcgttgttaaatatggcgattctcaattaagccctactgttgagcgttggctttatactggtaagaatttgtataacgcatatgatactaaacaggctttttctagtaattatgattccggtgtttattcttatttaacgccttatttatcacacggtcggtatttcaaaccattaaatttaggtcagaagatgaaattaactaaaatatatttgaaaaagttttctcgcgttctttgtcttgcgattggatttgcatcagcatttacatatagttatataacccaacctaagccggaggttaaaaaggtagtctctcagacctatgattttgataaattcactattgactcttctcagcgtcttaatctaagctatcgctatgttttcaaggattctaagggaaaattaattaatagcgacgatttacagaagcaaggttattcactcacatatattgatttatgtactgtttccattaaaaaaggtaattcaaatgaaattgttaaatgtaattaa\".upper(),\n",
    "                        \"IV\":\"atgaaattgttaaatgtaattaattttgttttcttgatgtttgtttcatcatcttcttttgctcaggtaattgaaatgaataattcgcctctgcgcgattttgtaacttggtattcaaagcaatcaggcgaatccgttattgtttctcccgatgtaaaaggtactgttactgtatattcatctgacgttaaacctgaaaatctacgcaatttctttatttctgttttacgtgcaaataattttgatatggtaggttctaacccttccattattcagaagtataatccaaacaatcaggattatattgatgaattgccatcatctgataatcaggaatatgatgataattccgctccttctggtggtttctttgttccgcaaaatgataatgttactcaaacttttaaaattaataacgttcgggcaaaggatttaatacgagttgtcgaattgtttgtaaagtctaatacttctaaatcctcaaatgtattatctattgacggctctaatctattagttgttagtgctcctaaagatattttagataaccttcctcaattcctttcaactgttgatttgccaactgaccagatattgattgagggtttgatatttgaggttcagcaaggtgatgctttagatttttcatttgctgctggctctcagcgtggcactgttgcaggcggtgttaatactgaccgcctcacctctgttttatcttctgctggtggttcgttcggtatttttaatggcgatgttttagggctatcagttcgcgcattaaagactaatagccattcaaaaatattgtctgtgccacgtattcttacgctttcaggtcagaagggttctatctctgttggccagaatgtcccttttattactggtcgtgtgactggtgaatctgccaatgtaaataatccatttcagacgattgagcgtcaaaatgtaggtatttccatgagcgtttttcctgttgcaatggctggcggtaatattgttctggatattaccagcaaggccgatagtttgagttcttctactcaggcaagtgatgttattactaatcaaagaagtattgctacaacggttaatttgcgtgatggacagactcttttactcggtggcctcactgattataaaaacacttctcaggattctggcgtaccgttcctgtctaaaatccctttaatcggcctcctgtttagctcccgctctgattctaacgaggaaagcacgttatacgtgctcgtcaaagcaaccatagtacgcgccctgtag\".upper(),\n",
    "                        }\n",
    "\n",
    "# traceback_dic ={}\n",
    "# for record in SeqIO.parse(path_to_m13_modified_genome, \"genbank\"):\n",
    "#     for feat in record.features:\n",
    "#         if feat.type == \"gene\" and str(feat.qualifiers[\"gene\"][0]) != \"II\":\n",
    "#             print(feat)\n",
    "#             if feat.qualifiers[\"gene\"][0] not in traceback_dic:\n",
    "#                 traceback_dic[str(feat.qualifiers[\"gene\"][0])] = feat.location.end\n",
    "#             else:\n",
    "#                 if traceback_dic[str(feat.qualifiers[\"gene\"][0])] > feat.location.end:\n",
    "#                     continue\n",
    "#                 else:\n",
    "#                     traceback_dic[str(feat.qualifiers[\"gene\"][0])] = feat.location.end\n",
    "#             if feat.qualifiers[\"gene\"][0] in all_gene_m13_modified:\n",
    "#                     traceback_dic[str(feat.qualifiers[\"gene\"][0])] = feat.location.end\n",
    "#                     all_gene_m13_modified[str(feat.qualifiers[\"gene\"][0])] = str(record.seq[str(record.seq).find(all_gene_m13_modified[str(feat.qualifiers[\"gene\"][0])]):feat.location.end])\n",
    "#             else:\n",
    "#                 all_gene_m13_modified[str(feat.qualifiers[\"gene\"][0])] = str(record.seq)[feat.location.start:feat.location.end]\n",
    "#print(all_gene_m13_modified)\n",
    "\n",
    "diff_dic = {}\n",
    "for gene,seq in all_gene_m13_modified.items():\n",
    "    if not check_if_same_prot(seq,all_gene_m13_native[str(gene)]):\n",
    "        print(gene)\n",
    "    n=0\n",
    "    for i,nt in enumerate(seq):\n",
    "        if nt != all_gene_m13_native[str(gene)][i]:\n",
    "            n+=1 \n",
    "    diff_dic[gene] = n\n",
    "    \n",
    "print(f\"il y a {sum(diff_dic.values())} nt différents par rapport à la version originale\")        \n",
    "print(diff_dic)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'X': 1, 'VIII': 1, 'II_prime': 2, 'IV': 3, 'V': 4, 'VI': 5, 'I': 6, 'I_prime': 7, 'IV_prime': 8}\n"
     ]
    }
   ],
   "source": [
    "wt_proms = [\"aagtctttcgggcttcctcttAATcTttttGaTGcaAtccgctttgcttctGacTATAATagtcagggtaaagacctgatt\",\n",
    "\"ttcgacacaatttatcaggcgatgatacAAaTctccGTTGtactttgtttcgcgctTggTATAATcgctgggggtcaaaga\",\n",
    "\"ttctcacccttttgAATcTttacCTacACAttactcaggcattgcattTAaAATatatgagggttctaaaaatttttatcc\",\n",
    "\"acctatgattttgatAAAtTcactaTTGACtcttctcagcgtcttaatcTAagcTatcgctATgttttcaaggattctaag\",\n",
    "\"gttccgttagttcgttttattAAcgtagatTTttCttcccaacgtcctgacTggTATAATgagcCAgttcttaaaatcgca\",\n",
    "\"ttgtctttggcgctggTaaacCaTatgAattttctattgattgTgacAaAATaaacttATtccgtggtgtctttgcgtttc\",\n",
    "\"ttgttcagggtgttcagttaattcTcccgtctaatgCgcttccctgtttttaTgtTATtcTctctgtaAaggctgctattt\",\n",
    "\"tgtaactggcaAATtaggctCTgGAaAgacgctcgttagcgtTggTAagATtcaggataAAATtgtagctgggtgcaaaat\",\n",
    "\"ttttaaaattAATaacgttCggGcaAaggatttaatacgagTtgTcgAATtgtttgtaAagtctaatacttctaaatcctc\"]\n",
    "\n",
    "alt_proms = [\"AAGTCTTTCGGGCTTCCTCTTAACCTCTTTGATGCTATCCGCTTTGCTTCTGACTACAACAGTCAGGGTAAAGACCTGATT\",\n",
    "\"ttcgacacaatttatcaggcgatgatacAAaTctccGTTGtactttgtttcgcgctTggTATAATcgctgggggtcaaaga\",\n",
    "\"TTCTCACCCTTTTGAGTCGTTACCGACGCACTACTCAGGCATTGCATTCAAGATCTATGAGGGTTCTAAAAATTTTTATCC\",\n",
    "\"ACCTATGATTTTGATAAATTCACTATCGATTCTTCTCAGCGTCTTAATCTCAGCTACCGCTACGTTTTCAAGGATTCTAAG\",\n",
    "\"GTTCCGTTAGTTCGTTTTATTAACGTAGATTTCAGCTCCCAACGTCCTGACTGGTACAATGAGCCCGTTCTTAAAATCGCA\",\n",
    "\"TTGTCTTTGGCGCTGGGAAACCGTACGAGTTTTCTATTGATTGTGACAAGATCAACCTCTTTCGTGGTGTCTTTGCGTTTC\",\n",
    "\"TTGTTCAGGGTGTTCAGTTAATTCTCCCGTCTAATGCGCTTCCCTGTTTTTACGTGATATTGTCTGTAAAGGCTGCTATTT\",\n",
    "\"TGTAACTGGCAAGCTTGGCTCGGGCAAGACGCTCGTTAGCGTAGGCAAGATCCAGGATAAGATCGTAGCTGGGTGCAAAAT\",\n",
    "\"TTTTAAAATTAATAACGTTCGGGCCAAGGATTTAATACGAGTAGTCGAGCTCTTTGTAAAGTCTAATACTTCTAAATCCTC\"]\n",
    "\n",
    "all_prom = [\"X\",\"VIII\",\"II_prime\",\"IV\",\"V\",\"VI\",\"I\",\"I_prime\",\"IV_prime\"]\n",
    "\n",
    "prom_dif = {} \n",
    "\n",
    "for z,(i,j) in enumerate(zip(wt_proms,alt_proms)):\n",
    "    n=0\n",
    "    for nt in i:\n",
    "        if i != j:\n",
    "            n += 1\n",
    "    prom_dif[all_prom[z]] = n\n",
    "    \n",
    "print(prom_dif)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'X': {'ATGAATATTTATGACGATTCCGCAGTATTGGACGCTATCCAGTCTAAACATTTTACTATTACCCCCTCTGGCAAAACTTCTTTTGCAAAAGCCTCTCGCTATTTTGGTTTTTATCGTCGTCTGGTAAACGAGGGTTATGATAGTGTTGCTCTTACTATGCCTCGTAATTCCTTTTGGCGTTATGTATCTGCATTAGTTGAATGTGGTATTCCTAAATCTCAACTGATGAATCTTTCTACCTGTAATAATGTT': (495, 747)}, 'V': {'ATGATTAAAGTTGAAATTAAACCATCTCAAGCCCAATTTACTACTCGTTCTGGTGTTTCTCGTCAGGGCAAGCCTTATTCACTGAATGAGCAGCTTTGTTACGTTGATTTGGGTAATGAATATCCGGTTCTTGTCAAGATTACTCTTGATGAAGGTCAGCCAGCCTATGCGCCTGGTCTGTACACCGTTCATCTGTCCTCTTTCAAAGTTGGTCAGTTCGGTTCCCTTATGATTGACCGTCTGCGCCTCGTTC': (842, 1095)}, 'VII': {'ATGGAGCAGGTCGCGGATTTCGACACAATTTATCAGGCGATGATACAAATCTCCGTTGTACTTTGTTTCGCGCTTGGTATAATCGCT': (1107, 1194), 'ATGA': (1205, 1209)}, 'IX': {'ATGAGTGTTTTAGTGTATTCTTTTGCCTCTTTCGTTTTAGGTTGGTGCCTTCGTAGTGGCATTACGTATTTTACCCGTTTA': (1205, 1286), 'ATGA': (1300, 1304)}, 'VIII': {'ATGAAAAAGTCTTTAGTCCTCAAAGCCTCTGTAGCCGTTGCTACCCTCGTTCCGATGCTGTCTTTCGCTGCTGAGGGTGACGATCCCGCAAAAGCGGCCTTTAACTCCCTGCAAGCCTCAGCGACCGAATATATCGGTTATGCGTGGGCGATGGTTGTTGTCATTGTCGGCGCAACTATCGGTATCAAGCTGTTTAAGAAATTCACCTCGAAAGCAAGCTGA': (1300, 1522)}, 'III': {'GTGAAAAAATTATTATTCGCAATTCCTTTAGTTGTTCCTTTCTATTCTCACTCCGCTGAAACTGTTGAAAGTTGTTTAGCAAAATCCCATACAGAAAATTCATTTACTAACGTCTGGAAAGACGACAAAACTTTAGATCGTTACGCTAACTATGAGGGCTGTCTGTGGAATGCTACAGGCGTTGTAGTTTGTACTGGTGACGAAACTCAGTGTTACGGTACATGGGTTCCTATTGGGCTTGCTATCCCTGAAAATGAGGGTGGTGGCTCTGAGGGTGGCGGTTCTGAGGGTGGCGGTTCTGAGGGTGGCGGTACTAAACCTCCTGAGTACGGTGATACACCTATTCCGGGCTATACTTATATCAACCCTCTCGACGGCACTTATCCGCCTGGTACTGAGCAAAACCCCGCTAATCCTAATCCTTCTCTTGAGGAGTCTCAGCCTCTTAATACTTTCATGTTTCAGAATAATAGGTTCCGAAATAGGCAGGGGGCATTAACTGTTTATACGGGCACTGTTACTCAAGGCACTGACCCCGTTAAAACTTATTACCAGTACACTCCTGTATCATCAAAAGCCATGTATGACGCTTACTGGAACGGTAAATTCAGAGACTGCGCTTTCCATTCTGGCTTTAATGAGGATTTATTTGTTTGTGAATATCAAGGCCAATCGTCTGACCTGCCTCAACCTCCTGTCAATGCTGGCGGCGGCTCTGGTGGTGGTTCTGGTGGCGGCTCTGAGGGTGGTGGCTCTGAGGGTGGCGGTTCTGAGGGTGGCGGCTCTGAGGGAGGCGGTTCCGGTGGTGGCTCTGGTTCCGGTGATTTTGATTATGAAAAGATGGCAAACGCTAATAAGGGGGCTATGACCGAAAATGCCGATGAAAACGCGCTACAGTCTGACGCTAAAGGCAAACTTGATTCTGTCGCTACTGATTACGGTGCTGCTATCGATGGTTTCATTGGTGACGTTTCCGGCCTTGCTAATGGTAATGGTGCTACTGGTGATTTTGCTGGCTCTAATTCCCAAATGGCTCAAGTCGGTGACGGTGATAATTCACCTTTAATGAATAATTTCCGTCAATATTTACCTTCCCTCCCTCAATCGGTTGAATGTCGCCCTT': (1581, 2704), 'TTTTATATGTTGCCACCTTTATGTATGTATTTTCTACGTTTGCTAACATACTGCGTAA': (2785, 2843)}, 'VI': {'ATGCCAGTTCTTTTGGGTATTCCGTTATTATTGCGTTTCCTCGGTTTCCTTCTGGTAACTTTGTTCGGCTATCTGCTTACTTTTCTTAAAAAGGGCTTCGGTAAGATAGCTATTGCTATTTCATTGTTTCTTGCTCTTATTATTGGGCTTAACTCAATTCTTGTGGGTTATCTCTCTGATATTAGCGCTCAATTACCCTCTGACT': (2858, 3063), 'TCATTTTTGACGTTAAACAAAAAATCGTTTCTTATTTGGA': (3144, 3184)}, 'I': {'ATGGCTGTTTATTT': (3198, 3212), 'AGCAACTAATCTTGATTTAAGGCT': (3293, 3317), 'CGCTAAAACGCCTCGCGTTCTTAGAATACCGGATAAGCCTTCTATATCTGATTTGCTTGCTATTGGGCGCGGTAATGATTCCTACGATGAAAATAAAAACGGCTTGCTTGTTCTCGATGAGTGCGGTACTTGGTTTAATACCCGTTCTTGGAATGATAAGGAAAGACAGCCGATTATTGATTGGTTTCTACATGCTCGTAAATTAGGATGGGATATTATTTTTCTTGTTCAGGACTTATCTATTGTTGATAAACAGGCGCGTTCTGCATTAGCTGAACATGTTGTTTATTGTCGTCGTCTGGACAGAATTACTTTACCTTTTGTCGGTACTTTATATTCTCTTATTACTGGCTCGAAAATGCCTCTGCCTAAATTACATGTTGGCGTTGTTAAATATGGCGATTCTCAATTAAGCCCTACTGTTGAGCGTTGGCTTTATACTGGTAAGAATTTGTATAACGCATATGATACTAAACAGGCTTTTTCTAGTAATTATGATTCCGGTGTTTATTCTTATTTAACGCCTTATTTATCACACGGTCGGTATTTCAAACCATTAAATTTAGGTCAGAAGATGAAATTAACTAAAATATATTTGAAAAAGTTTTCTCGCGTTCTTTGTCTTGCGATTGGATTTGCATCAGCATTTACATATAGTTATATAACCCAACCTAAGCCGGAGGTTAAAAAGGTAGTCTCTCAG': (3344, 4047), 'GGAAAATTAATTAATAGCGACGATTTACAGAAGCAAGGTTATTCACTCACATATATTGATTTATGTACTGTTTCCATTA': (4128, 4207), 'ATGAAATTGTTAAATGTAATTAA': (4222, 4245)}, 'XI': {'ATGAAATTAACTAAAATATATTTGAAAAAGTTTTCTCGCGTTCTTTGTCTTGCGATTGGATTTGCATCAGCATTTACATATAGTTATATAACCCAACCTAAGCCGGAGGTTAAAAAGGTAGTCTCTCAG': (3918, 4047), 'GGAAAATTAATTAATAGCGACGATTTACAGAAGCAAGGTTATTCACTCACATATATTGATTTATGTACTGTTTCCATTA': (4128, 4207), 'ATGAAATTGTTAAATGTAATTAA': (4222, 4245)}, 'IV': {'ATGAAATTGTTAAATGTAATTAATTTTGTTTTCTTGATGTTTGTTTCATCATCTTCTTTTGCTCAGGTAATTGAAATGAATAATTCGCCTCTGCGCGATTTTGTAACTTGGTATTCAAAGCAATCAGGCGAATCCGTTATTGTTTCTCCCGATGTAAAAGGTACTGTTACTGTATATTCATCTGACGTTAAACCTGAAAATCTACGCAATTTCTTTATTTCTGTTTTACGTGCAAATAATTTTGATATGGTAGGTTCTAACCCTTCCATTATTCAGAAGTATAATCCAAACAATCAGGATTATATTGATGAATTGCCATCATCTGATAATCAGGAATATGATGATAATTCCGCTCCTTCTGGTGGTTTCTTTGTTCCGCAAAATGATAATGTTACTCAAAC': (4222, 4623), 'AAATGTATTATCTATTGACGGCTCTAATCTATTAGTTGTTAGTGCTCCTAAAGATATTTTAGATAACCTTCCTCAATTCCTTTCAACTGTTGATTTGCCAACTGACCAGATATTGATTGAGGGTTTGATATTTGAGGTTCAGCAAGGTGATGCTTTAGATTTTTCATTTGCTGCTGGCTCTCAGCGTGGCACTGTTGCAGGCGGTGTTAATACTGACCGCCTCACCTCTGTTTTATCTTCTGCTGGTGGTTCGTTCGGTATTTTTAATGGCGATGTTTTAGGGCTATCAGTTCGCGCATTAAAGACTAATAGCCATTCAAAAATATTGTCTGTGCCACGTATTCTTACGCTTTCAGGTCAGAAGGGTTCTATCTCTGTTGGCCAGAATGTCCCTTTTATTACTGGTCGTGTGACTGGTGAATCTGCCAATGTAAATAATCCATTTCAGACGATTGAGCGTCAAAATGTAGGTATTTCCATGAGCGTTTTTCCTGTTGCAATGGCTGGCGGTAATATTGTTCTGGATATTACCAGCAAGGCCGATAGTTTGAGTTCTTCTACTCAGGCAAGTGATGTTATTACTAATCAAAGAAGTATTGCTACAACGGTTAATTTGCGTGATGGACAGACTCTTTTACTCGGTGGCCTCACTGATTATAAAAACACTTCTCAGGATTCTGGCGTACCGTTCCTGTCTAAAATCCCTTTAATCGGCCTCCTGTTTAGCTCCCGCTCTGATTCTAACGAGGAAAGCACGTTATACGTGCTCGTCAAAGCAACCATAGTACGCGCCCTGT': (4704, 5501)}, 'kanR': {'TTAGAAAAACTCATCGAGCATCAAATGAAACTGCAATTTATTCATATCAGGATTATCAATACCATATTTTTGAAAAAGCCGTTTCTGTAATGAAGGAGAAAACTCACCGAGGCAGTTCCATAGGATGGCAAGATCCTGGTATCGGTCTGCGATTCCGACTCGTCCAACATCAATACAACCTATTAATTTCCCCTCGTCAAAAATAAGGTTATCAAGTGAGAAATCACCATGAGTGACGACTGAATCCGGTGAGAATGGCAAAAGCTTATGCATTTCTTTCCAGACTTGTTCAACAGGCCAGCCATTACGCTCGTCATCAAAATCACTCGCATCAACCAAACCGTTATTCATTCGTGATTGCGCCTGAGCGAGACGAAATACGCGATCGCTGTTAAAAGGACAATTACAAACAGGAATCGAATGCAACCGGCGCAGGAACACTGCCAGCGCATCAACAATATTTTCACCTGAATCAGGATATTCTTCTAATACCTGGAATGCTGTTTTCCCGGGGATCGCAGTGGTGAGTAACCATGCATCATCAGGAGTACGGATAAAATGCTTGATGGTCGGAAGAGGCATAAATTCCGTCAGCCAGTTTAGTCTGACCATCTCATCTGTAACATCATTGGCAACGCTACCTTTGCCATGTTTCAGAAACAACTCTGGCGCATCGGGCTTCCCATACAATCGATAGATTGTCGCACCTGATTGCCCGACATTATCGCGAGCCCATTTATACCCATATAAATCAGCATCCATGTTGGAATTTAATCGCGGCCTCGAGCAAGACGTTTCCCGTTGAATATGGCTCAT': (5871, 6687)}, 'II': {'ATGCTAGTTTTACGATTACCGTTCATCGATTCTCTTGTTTGCTCCAGACTCTCAGGCAATGACCTGATAGCCTTTGTAGATCTCTCAAAAATAGCTACCCTCTCCGGCATTAATTTATCAGCTAGAACGGTTGAATATCATATTGATGGTGATTTGACTGTCTCCGGCCT': (6893, 7063), 'TTGCGTTGAAATAAAGGCTTCTCCCGCAAAAGTATTACAGGGTCATAATGTTTTTGGTACAACCGATTTAGCTTTATGCTCTGAGGCTTTATTGCTTAATTTTGCTAATTCTTTGCCTTGCCTGTATGATTTATTGGATGTT': (7144, 7286), 'AATGCTACTACTATTAGTAGAATTGATGCCACCTTTTCAGCTCGCGCCCCAAATGAAAATATAGCTAAACAGGTTATTGACCATTTGCGAAATGTATCTAATGGTCAAACTAAATCTACTCGTTCGCAGAATTGGGAATCAACTGTTATATGGAATGAAACTTCCAGACACCGTACTTTAGTTGCATATTTAAAACATGTTGAGCTACAGCATTATATTCAGCAATTAAGCTCTAAGCCATCCGCAAAAATGACCTCTTATCAAAAGGAGCAATTAAAGGTACTCTCTAATCCTGACCTGTTGGAGTTTGCTTCCGGTCTGGTTCGCTTTGAAGCTCGAATTAAAACGCGATATTTG': (0, 357), 'TTTGATTTATGGTCATTCTCGTTTTCTGAACTGTTTAAAGCA': (438, 480), 'ATGAATATTTATGACGATTCCGCAGTATTGGACGCTATCCAGTCTAAACATTTTACTATTACCCCCTCTGGCAAAACTTCTTTTGCAAAAGCCTCTCGCTATTTTGGTTTTTATCGTCGTCTGGTAAACGAGGGTTATGATAGTGTTGCTCTTACTATGCCTCGTAATTCCTTTTGGCGTTATGTATCTGCATTAGTTGAATGTGGTATTCCTAAATCTCAACTGATGAATCTTTCTACCTGTAATAATGTT': (495, 747)}} {'X': {'ATGAATATTTATGACGATTCCGCAGTATTGGACGCTATCCAGTCTAAACATTTTACTATTACCCCCTCTGGCAAAACTTCTTTTGCAAAAGCCTCTCGCTATTTTGGTTTTTATCGTCGTCTGGTAAACGAGGGTTATGATAGTGTTGCTCTTACTATGCCTCGTAATTCCTTTTGGCGTTATGTATCTGCATTAGTTGAATGTGGTATTCCTAAATCTCAACTGATGAATCTTTCTACCTGTAATAATGTT': (495, 747)}, 'V': {'ATGATTAAAGTTGAAATTAAACCATCTCAAGCCCAATTTACTACTCGTTCTGGTGTTTCTCGTCAGGGCAAGCCTTATTCACTGAATGAGCAGCTTTGTTACGTTGATTTGGGTAATGAATATCCGGTTCTTGTCAAGATTACTCTTGATGAAGGTCAGCCAGCCTATGCGCCTGGTCTGTACACCGTTCATCTGTCCTCTTTCAAAGTTGGTCAGTTCGGTTCCCTTATGATTGACCGTCTGCGCCTCGTTC': (842, 1095)}, 'VII': {'ATGGAGCAGGTCGCGGATTTCGACACAATTTATCAGGCGATGATACAAATCTCCGTTGTACTTTGTTTCGCGCTTGGTATAATCGCT': (1107, 1194), 'ATGA': (1205, 1209)}, 'IX': {'ATGAGTGTTTTAGTGTATTCTTTTGCCTCTTTCGTTTTAGGTTGGTGCCTTCGTAGTGGCATTACGTATTTTACCCGTTTA': (1205, 1286), 'ATGA': (1300, 1304)}, 'VIII': {'ATGAAAAAGTCTTTAGTCCTCAAAGCCTCTGTAGCCGTTGCTACCCTCGTTCCGATGCTGTCTTTCGCTGCTGAGGGTGACGATCCCGCAAAAGCGGCCTTTAACTCCCTGCAAGCCTCAGCGACCGAATATATCGGTTATGCGTGGGCGATGGTTGTTGTCATTGTCGGCGCAACTATCGGTATCAAGCTGTTTAAGAAATTCACCTCGAAAGCAAGCTGA': (1300, 1522)}, 'III': {'GTGAAAAAATTATTATTCGCAATTCCTTTAGTTGTTCCTTTCTATTCTCACTCCGCTGAAACTGTTGAAAGTTGTTTAGCAAAATCCCATACAGAAAATTCATTTACTAACGTCTGGAAAGACGACAAAACTTTAGATCGTTACGCTAACTATGAGGGCTGTCTGTGGAATGCTACAGGCGTTGTAGTTTGTACTGGTGACGAAACTCAGTGTTACGGTACATGGGTTCCTATTGGGCTTGCTATCCCTGAAAATGAGGGTGGTGGCTCTGAGGGTGGCGGTTCTGAGGGTGGCGGTTCTGAGGGTGGCGGTACTAAACCTCCTGAGTACGGTGATACACCTATTCCGGGCTATACTTATATCAACCCTCTCGACGGCACTTATCCGCCTGGTACTGAGCAAAACCCCGCTAATCCTAATCCTTCTCTTGAGGAGTCTCAGCCTCTTAATACTTTCATGTTTCAGAATAATAGGTTCCGAAATAGGCAGGGGGCATTAACTGTTTATACGGGCACTGTTACTCAAGGCACTGACCCCGTTAAAACTTATTACCAGTACACTCCTGTATCATCAAAAGCCATGTATGACGCTTACTGGAACGGTAAATTCAGAGACTGCGCTTTCCATTCTGGCTTTAATGAGGATTTATTTGTTTGTGAATATCAAGGCCAATCGTCTGACCTGCCTCAACCTCCTGTCAATGCTGGCGGCGGCTCTGGTGGTGGTTCTGGTGGCGGCTCTGAGGGTGGTGGCTCTGAGGGTGGCGGTTCTGAGGGTGGCGGCTCTGAGGGAGGCGGTTCCGGTGGTGGCTCTGGTTCCGGTGATTTTGATTATGAAAAGATGGCAAACGCTAATAAGGGGGCTATGACCGAAAATGCCGATGAAAACGCGCTACAGTCTGACGCTAAAGGCAAACTTGATTCTGTCGCTACTGATTACGGTGCTGCTATCGATGGTTTCATTGGTGACGTTTCCGGCCTTGCTAATGGTAATGGTGCTACTGGTGATTTTGCTGGCTCTAATTCCCAAATGGCTCAAGTCGGTGACGGTGATAATTCACCTTTAATGAATAATTTCCGTCAATATTTACCTTCCCTCCCTCAATCGGTTGAATGTCGCCCTT': (1581, 2704), 'TTTTATATGTTGCCACCTTTATGTATGTATTTTCTACGTTTGCTAACATACTGCGTAA': (2785, 2843)}, 'VI': {'ATGCCAGTTCTTTTGGGTATTCCGTTATTATTGCGTTTCCTCGGTTTCCTTCTGGTAACTTTGTTCGGCTATCTGCTTACTTTTCTTAAAAAGGGCTTCGGTAAGATAGCTATTGCTATTTCATTGTTTCTTGCTCTTATTATTGGGCTTAACTCAATTCTTGTGGGTTATCTCTCTGATATTAGCGCTCAATTACCCTCTGACT': (2858, 3063), 'TCATTTTTGACGTTAAACAAAAAATCGTTTCTTATTTGGA': (3144, 3184)}, 'I': {'ATGGCTGTTTATTT': (3198, 3212), 'AGCAACTAATCTTGATTTAAGGCT': (3293, 3317), 'CGCTAAAACGCCTCGCGTTCTTAGAATACCGGATAAGCCTTCTATATCTGATTTGCTTGCTATTGGGCGCGGTAATGATTCCTACGATGAAAATAAAAACGGCTTGCTTGTTCTCGATGAGTGCGGTACTTGGTTTAATACCCGTTCTTGGAATGATAAGGAAAGACAGCCGATTATTGATTGGTTTCTACATGCTCGTAAATTAGGATGGGATATTATTTTTCTTGTTCAGGACTTATCTATTGTTGATAAACAGGCGCGTTCTGCATTAGCTGAACATGTTGTTTATTGTCGTCGTCTGGACAGAATTACTTTACCTTTTGTCGGTACTTTATATTCTCTTATTACTGGCTCGAAAATGCCTCTGCCTAAATTACATGTTGGCGTTGTTAAATATGGCGATTCTCAATTAAGCCCTACTGTTGAGCGTTGGCTTTATACTGGTAAGAATTTGTATAACGCATATGATACTAAACAGGCTTTTTCTAGTAATTATGATTCCGGTGTTTATTCTTATTTAACGCCTTATTTATCACACGGTCGGTATTTCAAACCATTAAATTTAGGTCAGAAGATGAAATTAACTAAAATATATTTGAAAAAGTTTTCTCGCGTTCTTTGTCTTGCGATTGGATTTGCATCAGCATTTACATATAGTTATATAACCCAACCTAAGCCGGAGGTTAAAAAGGTAGTCTCTCAG': (3344, 4047), 'GGAAAATTAATTAATAGCGACGATTTACAGAAGCAAGGTTATTCACTCACATATATTGATTTATGTACTGTTTCCATTA': (4128, 4207), 'ATGAAATTGTTAAATGTAATTAA': (4222, 4245)}, 'XI': {'ATGAAATTAACTAAAATATATTTGAAAAAGTTTTCTCGCGTTCTTTGTCTTGCGATTGGATTTGCATCAGCATTTACATATAGTTATATAACCCAACCTAAGCCGGAGGTTAAAAAGGTAGTCTCTCAG': (3918, 4047), 'GGAAAATTAATTAATAGCGACGATTTACAGAAGCAAGGTTATTCACTCACATATATTGATTTATGTACTGTTTCCATTA': (4128, 4207), 'ATGAAATTGTTAAATGTAATTAA': (4222, 4245)}, 'IV': {'ATGAAATTGTTAAATGTAATTAATTTTGTTTTCTTGATGTTTGTTTCATCATCTTCTTTTGCTCAGGTAATTGAAATGAATAATTCGCCTCTGCGCGATTTTGTAACTTGGTATTCAAAGCAATCAGGCGAATCCGTTATTGTTTCTCCCGATGTAAAAGGTACTGTTACTGTATATTCATCTGACGTTAAACCTGAAAATCTACGCAATTTCTTTATTTCTGTTTTACGTGCAAATAATTTTGATATGGTAGGTTCTAACCCTTCCATTATTCAGAAGTATAATCCAAACAATCAGGATTATATTGATGAATTGCCATCATCTGATAATCAGGAATATGATGATAATTCCGCTCCTTCTGGTGGTTTCTTTGTTCCGCAAAATGATAATGTTACTCAAAC': (4222, 4623), 'AAATGTATTATCTATTGACGGCTCTAATCTATTAGTTGTTAGTGCTCCTAAAGATATTTTAGATAACCTTCCTCAATTCCTTTCAACTGTTGATTTGCCAACTGACCAGATATTGATTGAGGGTTTGATATTTGAGGTTCAGCAAGGTGATGCTTTAGATTTTTCATTTGCTGCTGGCTCTCAGCGTGGCACTGTTGCAGGCGGTGTTAATACTGACCGCCTCACCTCTGTTTTATCTTCTGCTGGTGGTTCGTTCGGTATTTTTAATGGCGATGTTTTAGGGCTATCAGTTCGCGCATTAAAGACTAATAGCCATTCAAAAATATTGTCTGTGCCACGTATTCTTACGCTTTCAGGTCAGAAGGGTTCTATCTCTGTTGGCCAGAATGTCCCTTTTATTACTGGTCGTGTGACTGGTGAATCTGCCAATGTAAATAATCCATTTCAGACGATTGAGCGTCAAAATGTAGGTATTTCCATGAGCGTTTTTCCTGTTGCAATGGCTGGCGGTAATATTGTTCTGGATATTACCAGCAAGGCCGATAGTTTGAGTTCTTCTACTCAGGCAAGTGATGTTATTACTAATCAAAGAAGTATTGCTACAACGGTTAATTTGCGTGATGGACAGACTCTTTTACTCGGTGGCCTCACTGATTATAAAAACACTTCTCAGGATTCTGGCGTACCGTTCCTGTCTAAAATCCCTTTAATCGGCCTCCTGTTTAGCTCCCGCTCTGATTCTAACGAGGAAAGCACGTTATACGTGCTCGTCAAAGCAACCATAGTACGCGCCCTGT': (4704, 5501)}, 'kanR': {'TTAGAAAAACTCATCGAGCATCAAATGAAACTGCAATTTATTCATATCAGGATTATCAATACCATATTTTTGAAAAAGCCGTTTCTGTAATGAAGGAGAAAACTCACCGAGGCAGTTCCATAGGATGGCAAGATCCTGGTATCGGTCTGCGATTCCGACTCGTCCAACATCAATACAACCTATTAATTTCCCCTCGTCAAAAATAAGGTTATCAAGTGAGAAATCACCATGAGTGACGACTGAATCCGGTGAGAATGGCAAAAGCTTATGCATTTCTTTCCAGACTTGTTCAACAGGCCAGCCATTACGCTCGTCATCAAAATCACTCGCATCAACCAAACCGTTATTCATTCGTGATTGCGCCTGAGCGAGACGAAATACGCGATCGCTGTTAAAAGGACAATTACAAACAGGAATCGAATGCAACCGGCGCAGGAACACTGCCAGCGCATCAACAATATTTTCACCTGAATCAGGATATTCTTCTAATACCTGGAATGCTGTTTTCCCGGGGATCGCAGTGGTGAGTAACCATGCATCATCAGGAGTACGGATAAAATGCTTGATGGTCGGAAGAGGCATAAATTCCGTCAGCCAGTTTAGTCTGACCATCTCATCTGTAACATCATTGGCAACGCTACCTTTGCCATGTTTCAGAAACAACTCTGGCGCATCGGGCTTCCCATACAATCGATAGATTGTCGCACCTGATTGCCCGACATTATCGCGAGCCCATTTATACCCATATAAATCAGCATCCATGTTGGAATTTAATCGCGGCCTCGAGCAAGACGTTTCCCGTTGAATATGGCTCAT': (5871, 6687)}, 'II': {'ATGCTAGTTTTACGATTACCGTTCATCGATTCTCTTGTTTGCTCCAGACTCTCAGGCAATGACCTGATAGCCTTTGTAGATCTCTCAAAAATAGCTACCCTCTCCGGCATTAATTTATCAGCTAGAACGGTTGAATATCATATTGATGGTGATTTGACTGTCTCCGGCCT': (6893, 7063), 'TTGCGTTGAAATAAAGGCTTCTCCCGCAAAAGTATTACAGGGTCATAATGTTTTTGGTACAACCGATTTAGCTTTATGCTCTGAGGCTTTATTGCTTAATTTTGCTAATTCTTTGCCTTGCCTGTATGATTTATTGGATGTT': (7144, 7286), 'AATGCTACTACTATTAGTAGAATTGATGCCACCTTTTCAGCTCGCGCCCCAAATGAAAATATAGCTAAACAGGTTATTGACCATTTGCGAAATGTATCTAATGGTCAAACTAAATCTACTCGTTCGCAGAATTGGGAATCAACTGTTATATGGAATGAAACTTCCAGACACCGTACTTTAGTTGCATATTTAAAACATGTTGAGCTACAGCATTATATTCAGCAATTAAGCTCTAAGCCATCCGCAAAAATGACCTCTTATCAAAAGGAGCAATTAAAGGTACTCTCTAATCCTGACCTGTTGGAGTTTGCTTCCGGTCTGGTTCGCTTTGAAGCTCGAATTAAAACGCGATATTTG': (0, 357), 'TTTGATTTATGGTCATTCTCGTTTTCTGAACTGTTTAAAGCA': (438, 480), 'ATGAATATTTATGACGATTCCGCAGTATTGGACGCTATCCAGTCTAAACATTTTACTATTACCCCCTCTGGCAAAACTTCTTTTGCAAAAGCCTCTCGCTATTTTGGTTTTTATCGTCGTCTGGTAAACGAGGGTTATGATAGTGTTGCTCTTACTATGCCTCGTAATTCCTTTTGGCGTTATGTATCTGCATTAGTTGAATGTGGTATTCCTAAATCTCAACTGATGAATCTTTCTACCTGTAATAATGTT': (495, 747)}}\n",
      "{'X': {'ATGAATATTTATGACGATTCCGCAGTATTGGACGCTATCCAGTCTAAACATTTTACTATTACCCCCTCTGGCAAAACTTCTTTTGCAAAAGCCTCTCGCTATTTTGGTTTTTATCGTCGTCTGGTAAACGAGGGTTATGATAGTGTTGCTCTTACTATGCCTCGTAATTCCTTTTGGCGTTATGTATCTGCATTAGTTGAATGTGGTATTCCTAAATCTCAACTGATGAATCTTTCTACCTGTAATAATGTT': (495, 747)}, 'V': {'ATGATTAAAGTTGAAATTAAACCATCTCAAGCCCAATTTACTACTCGTTCTGGTGTTTCTCGTCAGGGCAAGCCTTATTCACTGAATGAGCAGCTTTGTTACGTTGATTTGGGTAATGAATATCCGGTTCTTGTCAAGATTACTCTTGATGAAGGTCAGCCAGCCTATGCGCCTGGTCTGTACACCGTTCATCTGTCCTCTTTCAAAGTTGGTCAGTTCGGTTCCCTTATGATTGACCGTCTGCGCCTCGTTC': (842, 1095)}, 'VII': {'ATGGAGCAGGTCGCGGATTTCGACACAATTTATCAGGCGATGATACAAATCTCCGTTGTACTTTGTTTCGCGCTTGGTATAATCGCT': (1107, 1194), 'ATGA': (1205, 1209)}, 'IX': {'ATGAGTGTTTTAGTGTATTCTTTTGCCTCTTTCGTTTTAGGTTGGTGCCTTCGTAGTGGCATTACGTATTTTACCCGTTTA': (1205, 1286), 'ATGA': (1300, 1304)}, 'VIII': {'ATGAAAAAGTCTTTAGTCCTCAAAGCCTCTGTAGCCGTTGCTACCCTCGTTCCGATGCTGTCTTTCGCTGCTGAGGGTGACGATCCCGCAAAAGCGGCCTTTAACTCCCTGCAAGCCTCAGCGACCGAATATATCGGTTATGCGTGGGCGATGGTTGTTGTCATTGTCGGCGCAACTATCGGTATCAAGCTGTTTAAGAAATTCACCTCGAAAGCAAGCTGA': (1300, 1522)}, 'III': {'GTGAAAAAATTATTATTCGCAATTCCTTTAGTTGTTCCTTTCTATTCTCACTCCGCTGAAACTGTTGAAAGTTGTTTAGCAAAATCCCATACAGAAAATTCATTTACTAACGTCTGGAAAGACGACAAAACTTTAGATCGTTACGCTAACTATGAGGGCTGTCTGTGGAATGCTACAGGCGTTGTAGTTTGTACTGGTGACGAAACTCAGTGTTACGGTACATGGGTTCCTATTGGGCTTGCTATCCCTGAAAATGAGGGTGGTGGCTCTGAGGGTGGCGGTTCTGAGGGTGGCGGTTCTGAGGGTGGCGGTACTAAACCTCCTGAGTACGGTGATACACCTATTCCGGGCTATACTTATATCAACCCTCTCGACGGCACTTATCCGCCTGGTACTGAGCAAAACCCCGCTAATCCTAATCCTTCTCTTGAGGAGTCTCAGCCTCTTAATACTTTCATGTTTCAGAATAATAGGTTCCGAAATAGGCAGGGGGCATTAACTGTTTATACGGGCACTGTTACTCAAGGCACTGACCCCGTTAAAACTTATTACCAGTACACTCCTGTATCATCAAAAGCCATGTATGACGCTTACTGGAACGGTAAATTCAGAGACTGCGCTTTCCATTCTGGCTTTAATGAGGATTTATTTGTTTGTGAATATCAAGGCCAATCGTCTGACCTGCCTCAACCTCCTGTCAATGCTGGCGGCGGCTCTGGTGGTGGTTCTGGTGGCGGCTCTGAGGGTGGTGGCTCTGAGGGTGGCGGTTCTGAGGGTGGCGGCTCTGAGGGAGGCGGTTCCGGTGGTGGCTCTGGTTCCGGTGATTTTGATTATGAAAAGATGGCAAACGCTAATAAGGGGGCTATGACCGAAAATGCCGATGAAAACGCGCTACAGTCTGACGCTAAAGGCAAACTTGATTCTGTCGCTACTGATTACGGTGCTGCTATCGATGGTTTCATTGGTGACGTTTCCGGCCTTGCTAATGGTAATGGTGCTACTGGTGATTTTGCTGGCTCTAATTCCCAAATGGCTCAAGTCGGTGACGGTGATAATTCACCTTTAATGAATAATTTCCGTCAATATTTACCTTCCCTCCCTCAATCGGTTGAATGTCGCCCTT': (1581, 2704), 'TTTTATATGTTGCCACCTTTATGTATGTATTTTCTACGTTTGCTAACATACTGCGTAA': (2785, 2843)}, 'VI': {'ATGCCAGTTCTTTTGGGTATTCCGTTATTATTGCGTTTCCTCGGTTTCCTTCTGGTAACTTTGTTCGGCTATCTGCTTACTTTTCTTAAAAAGGGCTTCGGTAAGATAGCTATTGCTATTTCATTGTTTCTTGCTCTTATTATTGGGCTTAACTCAATTCTTGTGGGTTATCTCTCTGATATTAGCGCTCAATTACCCTCTGACT': (2858, 3063), 'TCATTTTTGACGTTAAACAAAAAATCGTTTCTTATTTGGA': (3144, 3184)}, 'I': {'ATGGCTGTTTATTT': (3198, 3212), 'AGCAACTAATCTTGATTTAAGGCT': (3293, 3317), 'CGCTAAAACGCCTCGCGTTCTTAGAATACCGGATAAGCCTTCTATATCTGATTTGCTTGCTATTGGGCGCGGTAATGATTCCTACGATGAAAATAAAAACGGCTTGCTTGTTCTCGATGAGTGCGGTACTTGGTTTAATACCCGTTCTTGGAATGATAAGGAAAGACAGCCGATTATTGATTGGTTTCTACATGCTCGTAAATTAGGATGGGATATTATTTTTCTTGTTCAGGACTTATCTATTGTTGATAAACAGGCGCGTTCTGCATTAGCTGAACATGTTGTTTATTGTCGTCGTCTGGACAGAATTACTTTACCTTTTGTCGGTACTTTATATTCTCTTATTACTGGCTCGAAAATGCCTCTGCCTAAATTACATGTTGGCGTTGTTAAATATGGCGATTCTCAATTAAGCCCTACTGTTGAGCGTTGGCTTTATACTGGTAAGAATTTGTATAACGCATATGATACTAAACAGGCTTTTTCTAGTAATTATGATTCCGGTGTTTATTCTTATTTAACGCCTTATTTATCACACGGTCGGTATTTCAAACCATTAAATTTAGGTCAGAAGATGAAATTAACTAAAATATATTTGAAAAAGTTTTCTCGCGTTCTTTGTCTTGCGATTGGATTTGCATCAGCATTTACATATAGTTATATAACCCAACCTAAGCCGGAGGTTAAAAAGGTAGTCTCTCAG': (3344, 4047), 'GGAAAATTAATTAATAGCGACGATTTACAGAAGCAAGGTTATTCACTCACATATATTGATTTATGTACTGTTTCCATTA': (4128, 4207), 'ATGAAATTGTTAAATGTAATTAA': (4222, 4245)}, 'XI': {'ATGAAATTAACTAAAATATATTTGAAAAAGTTTTCTCGCGTTCTTTGTCTTGCGATTGGATTTGCATCAGCATTTACATATAGTTATATAACCCAACCTAAGCCGGAGGTTAAAAAGGTAGTCTCTCAG': (3918, 4047), 'GGAAAATTAATTAATAGCGACGATTTACAGAAGCAAGGTTATTCACTCACATATATTGATTTATGTACTGTTTCCATTA': (4128, 4207), 'ATGAAATTGTTAAATGTAATTAA': (4222, 4245)}, 'IV': {'ATGAAATTGTTAAATGTAATTAATTTTGTTTTCTTGATGTTTGTTTCATCATCTTCTTTTGCTCAGGTAATTGAAATGAATAATTCGCCTCTGCGCGATTTTGTAACTTGGTATTCAAAGCAATCAGGCGAATCCGTTATTGTTTCTCCCGATGTAAAAGGTACTGTTACTGTATATTCATCTGACGTTAAACCTGAAAATCTACGCAATTTCTTTATTTCTGTTTTACGTGCAAATAATTTTGATATGGTAGGTTCTAACCCTTCCATTATTCAGAAGTATAATCCAAACAATCAGGATTATATTGATGAATTGCCATCATCTGATAATCAGGAATATGATGATAATTCCGCTCCTTCTGGTGGTTTCTTTGTTCCGCAAAATGATAATGTTACTCAAAC': (4222, 4623), 'AAATGTATTATCTATTGACGGCTCTAATCTATTAGTTGTTAGTGCTCCTAAAGATATTTTAGATAACCTTCCTCAATTCCTTTCAACTGTTGATTTGCCAACTGACCAGATATTGATTGAGGGTTTGATATTTGAGGTTCAGCAAGGTGATGCTTTAGATTTTTCATTTGCTGCTGGCTCTCAGCGTGGCACTGTTGCAGGCGGTGTTAATACTGACCGCCTCACCTCTGTTTTATCTTCTGCTGGTGGTTCGTTCGGTATTTTTAATGGCGATGTTTTAGGGCTATCAGTTCGCGCATTAAAGACTAATAGCCATTCAAAAATATTGTCTGTGCCACGTATTCTTACGCTTTCAGGTCAGAAGGGTTCTATCTCTGTTGGCCAGAATGTCCCTTTTATTACTGGTCGTGTGACTGGTGAATCTGCCAATGTAAATAATCCATTTCAGACGATTGAGCGTCAAAATGTAGGTATTTCCATGAGCGTTTTTCCTGTTGCAATGGCTGGCGGTAATATTGTTCTGGATATTACCAGCAAGGCCGATAGTTTGAGTTCTTCTACTCAGGCAAGTGATGTTATTACTAATCAAAGAAGTATTGCTACAACGGTTAATTTGCGTGATGGACAGACTCTTTTACTCGGTGGCCTCACTGATTATAAAAACACTTCTCAGGATTCTGGCGTACCGTTCCTGTCTAAAATCCCTTTAATCGGCCTCCTGTTTAGCTCCCGCTCTGATTCTAACGAGGAAAGCACGTTATACGTGCTCGTCAAAGCAACCATAGTACGCGCCCTGT': (4704, 5501)}, 'kanR': {'TTAGAAAAACTCATCGAGCATCAAATGAAACTGCAATTTATTCATATCAGGATTATCAATACCATATTTTTGAAAAAGCCGTTTCTGTAATGAAGGAGAAAACTCACCGAGGCAGTTCCATAGGATGGCAAGATCCTGGTATCGGTCTGCGATTCCGACTCGTCCAACATCAATACAACCTATTAATTTCCCCTCGTCAAAAATAAGGTTATCAAGTGAGAAATCACCATGAGTGACGACTGAATCCGGTGAGAATGGCAAAAGCTTATGCATTTCTTTCCAGACTTGTTCAACAGGCCAGCCATTACGCTCGTCATCAAAATCACTCGCATCAACCAAACCGTTATTCATTCGTGATTGCGCCTGAGCGAGACGAAATACGCGATCGCTGTTAAAAGGACAATTACAAACAGGAATCGAATGCAACCGGCGCAGGAACACTGCCAGCGCATCAACAATATTTTCACCTGAATCAGGATATTCTTCTAATACCTGGAATGCTGTTTTCCCGGGGATCGCAGTGGTGAGTAACCATGCATCATCAGGAGTACGGATAAAATGCTTGATGGTCGGAAGAGGCATAAATTCCGTCAGCCAGTTTAGTCTGACCATCTCATCTGTAACATCATTGGCAACGCTACCTTTGCCATGTTTCAGAAACAACTCTGGCGCATCGGGCTTCCCATACAATCGATAGATTGTCGCACCTGATTGCCCGACATTATCGCGAGCCCATTTATACCCATATAAATCAGCATCCATGTTGGAATTTAATCGCGGCCTCGAGCAAGACGTTTCCCGTTGAATATGGCTCAT': (5871, 6687)}, 'II': {}}\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'rev_ecoli_codon_usage' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-34124057de76>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    197\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mstart_part\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0mend_part\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    198\u001b[0m             \u001b[0min_frame_part\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdna_seq\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstart_codon\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mend_codon\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 199\u001b[0;31m             \u001b[0mrecoded_part\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrecode_perc_change\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m50\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0min_frame_part\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    200\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mrecoded_part\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m\"\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    201\u001b[0m                 \u001b[0mrecoded_genome_parts\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgene_name\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mrecoded_part\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mstart_codon\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mend_codon\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-2-34124057de76>\u001b[0m in \u001b[0;36mrecode_perc_change\u001b[0;34m(perc, cds)\u001b[0m\n\u001b[1;32m     12\u001b[0m             \u001b[0;31m#if no TGG (cannot be changed), no start codon (cannot be changed) or no stop codon:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m             \u001b[0mother_codons\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m             \u001b[0;32mfor\u001b[0m \u001b[0maa\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcodon_freq\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrev_ecoli_codon_usage\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mcodon_in_cds\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcodon_freq\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m                     \u001b[0;32mfor\u001b[0m \u001b[0mcodon\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mfreq\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcodon_freq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'rev_ecoli_codon_usage' is not defined"
     ]
    }
   ],
   "source": [
    "#collect all infos from the altprom file:\n",
    "\n",
    "def recode_perc_change(perc,cds):\n",
    "    #recode the cds by changing the % provided as input\n",
    "    #output: recoded cds (string)\n",
    "    new_cds = \"\"\n",
    "    for pos in range(int(len(cds)/3)):    #25% de chance de le changer\n",
    "        codon_in_cds = cds[pos*3:pos*3+3].upper()\n",
    "        #set seed at each position in the CDS\n",
    "        random.seed(pos)\n",
    "        if pick_this_pos(perc) and codon_in_cds !=\"ATG\" and codon_in_cds!=\"TGG\" and pos*3+3 != len(cds):\n",
    "            #if no TGG (cannot be changed), no start codon (cannot be changed) or no stop codon:\n",
    "            other_codons = {}\n",
    "            for aa, codon_freq in rev_ecoli_codon_usage.items():\n",
    "                if codon_in_cds in codon_freq:\n",
    "                    for codon,freq in codon_freq.items():\n",
    "                        if codon_in_cds != codon:\n",
    "                            other_codons[codon] = freq\n",
    "            new_codon_dict = {codon:freq/sum(list(codon_freq.values())) for codon,freq in other_codons.items()}\n",
    "            replacement_codon = random.choices(population=list(new_codon_dict.keys()),weights=list(new_codon_dict.values()),k=1)\n",
    "            new_cds += ''.join(replacement_codon).upper()\n",
    "        else:\n",
    "            new_cds += codon_in_cds\n",
    "    #new_cds is recoded\n",
    "    return new_cds\n",
    "\n",
    "def merge_intervals(intervals):\n",
    "    #get list of tuple as input.\n",
    "    sorted_by_lower_bound = sorted(intervals, key=lambda tup: tup[0])\n",
    "    merged = []\n",
    "\n",
    "    for higher in sorted_by_lower_bound:\n",
    "        if not merged:\n",
    "            merged.append(higher)\n",
    "        else:\n",
    "            lower = merged[-1]\n",
    "            # test for intersection between lower and higher:\n",
    "            # we know via sorting that lower[0] <= higher[0]\n",
    "            if higher[0] <= lower[1]:\n",
    "                upper_bound = max(lower[1], higher[1])\n",
    "                merged[-1] = (lower[0], upper_bound)  # replace by merged interval\n",
    "            else:\n",
    "                merged.append(higher)\n",
    "    return merged\n",
    "\n",
    "def merge_contiguous_regions(list_of_tuples):\n",
    "    #the regions do not overlap, but have common borders. This function merge common borders into one forbidden region.\n",
    "    #input: list of tuples\n",
    "    #output: list of tuples merged\n",
    "    merge_list = []\n",
    "    max_range = [-1]\n",
    "    max_list_of_tuples = max([max(i) for i in list_of_tuples])\n",
    "    min_range = [max_list_of_tuples+1]\n",
    "    for nb,ref in enumerate(list_of_tuples):\n",
    "        if nb+1 < len(list_of_tuples):    #to make sure we do not compared with something outside the list!\n",
    "            comp = list_of_tuples[nb+1]\n",
    "            if comp[0] <= ref[1] and comp[1] >= ref[0]: #if there is any overlap:\n",
    "                min_temp = min([ref[0],comp[0]])\n",
    "                max_temp = max([ref[1],comp[1]])\n",
    "                min_range.append(min_temp)\n",
    "                max_range.append(max_temp)\n",
    "                max_range = [max(max_range)]\n",
    "                min_range = [min(min_range)]\n",
    "                if nb+1 == len(list_of_tuples)-1:\n",
    "                    merge_list.append((min_range[0],max_range[0]))\n",
    "            else:\n",
    "                if nb > 0:\n",
    "                    if max_range == [-1] and min_range == [max_list_of_tuples+1]:\n",
    "                        merge_list.append((ref[0],ref[1]))\n",
    "                    else:\n",
    "                        merge_list.append((min_range[0],max_range[0]))\n",
    "                    if nb+1 == len(list_of_tuples)-1:  #if last\n",
    "                         merge_list.append((comp[0],comp[1]))\n",
    "                    max_range = [-1]\n",
    "                    min_range = [max_list_of_tuples+1]\n",
    "                else:\n",
    "                    merge_list.append((ref[0],ref[1]))\n",
    "    return merge_list\n",
    "def pick_this_pos(percent):\n",
    "    #return true if the random number (between 0-100) is lower than the given percent.\n",
    "    return random.randrange(100) < percent\n",
    "\n",
    "def same_ORF(interval, ref_pos_gene):\n",
    "    #this function tells if coordinates of a part of a cds to be recoded overlaps with an actual gene (other than himself)\n",
    "    #if the pos[0]-interval[0] < 3, on l'exclu de toute maniere\n",
    "    return bool(ref_pos_gene[0]-interval[0]%3 == 0)\n",
    "\n",
    "def overlap_situation(cds_parts,CDSs):\n",
    "    no_overlap_to_recoded = {}\n",
    "    for gene_name, parts in cds_parts.items():\n",
    "        no_overlap_to_recoded[gene_name] = {}\n",
    "        for gene_seq, interval in parts.items():\n",
    "            for comp_gene_name,pos in CDSs.items():\n",
    "                if interval[0] < interval[1]:#in the case of a gene that starts at tne end of the plasmid and goes through position 0 in the construct\n",
    "                    if pos[0] <= interval[1] and interval[0] <= pos[1]: #if overlap\n",
    "                        if gene_name == comp_gene_name: #if we are looking at interval in same gene\n",
    "                            no_overlap_to_recoded[gene_name][gene_seq] = interval\n",
    "                        else: #if comp_gene_name != gene_name, meaning an interval (seq to recode) overlaps with an actual CDSs in the construct other than himself\n",
    "                            if same_ORF(interval,pos):#if this genome section is in the same orf the others, append it. Otherwise, exclude it.\n",
    "                                no_overlap_to_recoded[gene_name][gene_seq] = interval\n",
    "                else:\n",
    "                    print(gene_name)\n",
    "    return no_overlap_to_recoded\n",
    "                    \n",
    "def find_forbidden_regions(path_to_gb_file):\n",
    "    #we have to include promoter sequence, terminator, rbs and the oriR in the restricted regions dic:\n",
    "    #data structure : {start:end, start1:end1, start2:end2, start3:end3...}\n",
    "    forbidden_regions = []\n",
    "    CDSs = {}\n",
    "    #CDSs in the form of :{I:(0,1000), II:(980,1500),...}: it is the position of every gene!\n",
    "    for gb_record in SeqIO.parse(open(path_to_gb_file, \"r\"), \"genbank\"):\n",
    "        dna_seq = gb_record.seq\n",
    "        for feat in gb_record.features:\n",
    "            if \"prom\" in str(feat.qualifiers) or \"rbs\" in str(feat.qualifiers) or \"term\" in str(feat.qualifiers):\n",
    "                initial_loc_forbid = str(feat.location)\n",
    "                start = int(initial_loc_forbid.split(\":\")[0][1:])\n",
    "                end = int(initial_loc_forbid.split(\":\")[1].split(\"]\")[0])\n",
    "                forbidden_regions.append((start,end))\n",
    "            if feat.type == \"CDS\" and \"join\":\n",
    "                if \"join\" not in str(feat.location):\n",
    "                    #join is here for gene II, because it is a circular construction (overlaps with the beginning and the end of the file)\n",
    "                    initial_loc_cds = str(feat.location)\n",
    "                    start = int(initial_loc_cds.split(\":\")[0][1:])\n",
    "                    end = int(initial_loc_cds.split(\":\")[1].split(\"]\")[0])\n",
    "                    CDSs[feat.qualifiers[\"label\"][0]] = (start,end)\n",
    "                else:\n",
    "                    initial_loc_cds_start = str(feat.location)[5:].split(\",\")[0]\n",
    "                    initial_loc_cds_end = str(feat.location).split(\",\")[1]\n",
    "                    start = int(initial_loc_cds_start.split(\":\")[0][1:])\n",
    "                    end = int(initial_loc_cds_end.split(\":\")[1].split(\"]\")[0])\n",
    "                    CDSs[feat.qualifiers[\"label\"][0]] = (start,end)\n",
    "    \n",
    "    return forbidden_regions,CDSs,dna_seq\n",
    "    \n",
    "def get_changing_cds_parts(coor,mforbiden_regions,dna_seq):\n",
    "    #coor is absolute coordinates of the fasta file\n",
    "    #returns all dna_string that could be recoded (excludes forbidden sequences (term_seq,prom_seq, rbs, rep origin...) and absolute pos in a dict \n",
    "    #{seq1:(start1,end1), seq2:(start2,end2),...}\n",
    "    cds_parts = {}\n",
    "    notpickreg = []\n",
    "    for i,regions in enumerate(mforbiden_regions):\n",
    "        #find all changeable part of CDSs, or exclude forbiden sequences:\n",
    "        if regions[0] > coor[0] and regions[0] < coor[1]: #Normal case: if, elif else\n",
    "            if not notpickreg: #if notpickreg empty\n",
    "                cds_parts[str(dna_seq[coor[0]:regions[0]])] = (coor[0],regions[0])\n",
    "                notpickreg = [regions[0],regions[1]]\n",
    "            else:\n",
    "                cds_parts[str(dna_seq[notpickreg[1]:regions[0]])] = (notpickreg[1],regions[0])\n",
    "                notpickreg = [regions[0],regions[1]]\n",
    "        elif regions[0] <= coor[0] and regions[1] >= coor[0]:\n",
    "            notpickreg = [regions[0],regions[1]]\n",
    "            if regions[1] >= coor[1]:\n",
    "                return []\n",
    "        else:\n",
    "            if regions[0] > coor[1]:  #if region[0] exceeds the limit of the cds\n",
    "                if not notpickreg: #meaning if no forbidden region in cds:\n",
    "                    cds_parts[str(dna_seq[coor[0]:coor[1]])] = (coor[0],coor[1])\n",
    "                else:\n",
    "                    if notpickreg[1] <= coor[1]:\n",
    "                        cds_parts[str(dna_seq[notpickreg[1]:coor[1]])] = (notpickreg[1],coor[1])\n",
    "                break\n",
    "        if regions == forbiden_regions[-1]: #last forbidden region, exceptionnal case:\n",
    "            if not notpickreg:\n",
    "                cds_parts[str(dna_seq[coor[0]:regions[0]])] = (coor[0],regions[0])\n",
    "            else:\n",
    "                cds_parts[str(dna_seq[notpickreg[1]:coor[1]])] = (notpickreg[1],coor[1])\n",
    "    return cds_parts\n",
    "\n",
    "forbiden_regions,CDSs,dna_seq = find_forbidden_regions(path_to_gb_file = \"../ref_files_m13/m13_alt/NEB_m13_kan.gb\")\n",
    "#add overlaping region in forbiden_regions:\n",
    "merge_forbiden_regions = merge_contiguous_regions(forbiden_regions)\n",
    "max_len_dna = len(str(dna_seq))\n",
    "cds_parts = {}\n",
    "for gene_name, coor in CDSs.items():\n",
    "    gene = dna_seq[coor[0]:coor[1]]\n",
    "    if coor[0] < coor[1]:\n",
    "        cds_parts[gene_name] = get_changing_cds_parts(coor,merge_forbiden_regions,dna_seq=dna_seq)\n",
    "    else:\n",
    "        cds_parts_end = get_changing_cds_parts((coor[0],len(dna_seq)),merge_forbiden_regions,dna_seq=dna_seq)\n",
    "        cds_parts_start = get_changing_cds_parts((0, coor[1]),merge_forbiden_regions,dna_seq=dna_seq)\n",
    "        cds_parts[gene_name] = {**cds_parts_end, **cds_parts_start}\n",
    "        \n",
    "gene_coor = list(CDSs.values())\n",
    "print(cds_parts,cds_parts)\n",
    "cds_part_no_prob_overlap = overlap_situation(cds_parts,CDSs)\n",
    "print(cds_part_no_prob_overlap)\n",
    "\n",
    "recoded_genome_parts = {}\n",
    "for gene_name, parts in cds_part_no_prob_overlap.items():\n",
    "    recoded_genome_parts[gene_name] = {}\n",
    "    for part,coor in parts.items():\n",
    "        #Search for orf coordinates and recoded the cds:\n",
    "        start_part = coor[0]\n",
    "        end_part = coor[1]\n",
    "        start_codon = start_part+(start_part%3)\n",
    "        end_codon = end_part-(end_part%3)\n",
    "        if start_part < end_part:\n",
    "            in_frame_part = dna_seq[start_codon:end_codon]\n",
    "            recoded_part = str(recode_perc_change(50,in_frame_part))\n",
    "            if recoded_part != \"\":\n",
    "                recoded_genome_parts[gene_name][recoded_part] = (start_codon,end_codon)\n",
    "        else:\n",
    "            print(\"minou\")\n",
    "            in_frame_part_end = dna_seq[start_codon:max_len_dna]\n",
    "            in_frame_part_start = dna_seq[0:end_codon]\n",
    "            recoded_part_end = str(recode_perc_change(50,in_frame_part_end))\n",
    "            recoded_part_start = str(recode_perc_change(50,in_frame_part_start))\n",
    "            if recoded_part != \"\":\n",
    "                recoded_genome_parts[gene_name][in_frame_part_end] = (start_codon,max_len_dna)\n",
    "                recoded_genome_parts[gene_name][recoded_part_start] = (0,end_codon)       \n",
    "print(recoded_genome_parts)\n",
    "\n",
    "#construct the final m13 vector:\n",
    "# all_recoded_start_coor = [str(list(recoded_parts.values())).split(\",\")[0][2:] for recoded_parts in recoded_genome_parts.values()]\n",
    "# all_recoded_start_coor = [(recoded_parts.values()) for recoded_parts in recoded_genome_parts.values()]\n",
    "# print(all_recoded_start_coor)\n",
    "# all_recoded_end_coor = [coor[1] for seq,coor in parts.items() for recoded_parts in recoded_genome_parts.values()]\n",
    "# flat_recoded_cds = {coor:seq for seq,coor in parts.items() for parts in recoded_genome_parts.values()}\n",
    "# print(flat_recoded_cds)\n",
    "\n",
    "# max_len_dna = len(str(dna_seq))\n",
    "# m13_WTZ = \"\"\n",
    "# last_pos = 0\n",
    "# for i in range(max_len_dna):\n",
    "#     if i in all_recoded_start_coor:\n",
    "#         m13_WTZ += dna_seq[last_pos:i]\n",
    "#         last_pos = i\n",
    "#     elif i in all_recoded_end_coor:\n",
    "#         m13_WTZ += dna_seq[last_pos:i]\n",
    "#         last_pos = i\n",
    "#     else:\n",
    "#         continue\n",
    "# print(len(m13_WTZ))\n",
    "        \n",
    "        \n",
    "#     print(\"old\",cds_parts, \"\\n\")    \n",
    "#     #le gene en chevauche-t-il un autre? Si oui, couper la partie correspondante dans la liste de gene\n",
    "#     cds_parts = overlap_situation(coor,cds_parts) \n",
    "#     print(\"new\",cds_parts,\"\\n\")\n",
    "#     print(old_cds_parts==cds_parts)    \n",
    "#     cds_recoded = {}\n",
    "#     traceback = []\n",
    "\n",
    "        \n",
    "                 \n",
    "        \n",
    "        \n",
    "        #when forbiden_regions[0] < coor[0]\n",
    "    #find the reading frame in each part of changeable CDSs\n",
    "    \n",
    "    #change the reading frame in cds\n",
    "    \n",
    "    #merge all substring into a new file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(1, 2), (4, 5), (7, 9), (10, 17), (19, 25)]"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge_contiguous_regions([(1,2),(4,5),(7,9),(10,15),(11,17),(19,25)])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
